// @allianz/ngx-ndbx/esm2015/base/error/error.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: error/error.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, InjectionToken, Inject, Optional } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
/** @type {?} */
let nextId = 0;
/**
 * Represents the default options for the error notification that can be configured
 * using the `ERROR_DEFAULT_OPTIONS` injection token.
 * @record
 */
export function ErrorDefaultOptions() { }
if (false) {
    /**
     * Defines the style type of the error notification.
     * @type {?|undefined}
     */
    ErrorDefaultOptions.prototype.appearance;
}
/** @type {?} */
export const ERROR_DEFAULT_OPTIONS = new InjectionToken('ERROR_DEFAULT_OPTIONS');
export class NxErrorComponent {
    /**
     * @param {?} _changeDetectorRef
     * @param {?} _defaultOptions
     */
    constructor(_changeDetectorRef, _defaultOptions) {
        this._changeDetectorRef = _changeDetectorRef;
        this._defaultOptions = _defaultOptions;
        this._showIcon = true;
        this._appearance = 'message';
        this._id = `nx-error-${nextId++}`;
        if (this._defaultOptions && this._defaultOptions.appearance) {
            this.appearance = this._defaultOptions.appearance;
        }
    }
    /**
     * Whether an icon should be displayed.
     * @param {?} value
     * @return {?}
     */
    set showIcon(value) {
        this._showIcon = coerceBooleanProperty(value);
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get showIcon() {
        return this._showIcon;
    }
    /**
     * Id of the nx-error.
     *
     * If not set, the selectable card gets an incremented value by default.
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        if (value && value !== this._id) {
            this._id = value;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get id() {
        return this._id;
    }
    /**
     * Whether the error should have message or text styling.
     *
     * Default is 'message'.
     * @param {?} value
     * @return {?}
     */
    set appearance(value) {
        if (value !== this.appearance) {
            this._appearance = value;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get appearance() {
        return this._appearance;
    }
}
NxErrorComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-error',
                template: "<!-- b2b (expert) error notification -->\n<ng-container *ngIf=\"appearance === 'text'\">\n  <nx-icon *ngIf=\"showIcon\" class=\"nx-error__icon\" [name]=\"'exclamation-triangle'\">\n  </nx-icon>\n  <div class=\"nx-error__content-wrapper\">\n    <div class=\"nx-error__content\">\n      <ng-container [ngTemplateOutlet]=\"errorContent\"></ng-container>\n    </div>\n  </div>\n</ng-container>\n\n<!-- b2c error notification -->\n<nx-message *ngIf=\"appearance === 'message'\" [showIcon]=\"showIcon\" nxContext=\"error\">\n  <ng-container [ngTemplateOutlet]=\"errorContent\"></ng-container>\n</nx-message>\n\n<ng-template #errorContent>\n  <ng-content></ng-content>\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[attr.role]': '"alert"',
                    '[attr.id]': 'id',
                    '[class.nx-error--message]': 'appearance == "message"'
                },
                styles: [":host{font-size:16px;font-size:var(--base-error-simple-font-size,16px);line-height:20px;line-height:var(--base-error-simple-line-height,20px);font-weight:400;font-weight:var(--base-error-simple-font-weight,400);letter-spacing:0;letter-spacing:var(--base-error-simple-letter-spacing,0);color:#dc3149;color:var(--base-error-color,#dc3149);width:100%;display:block;text-align:left}:host .nx-error__icon{font-size:20px;font-size:var(--base-error-simple-icon-size,20px);line-height:20px;line-height:var(--base-error-simple-line-height,20px);margin-right:8px;float:left}@media screen and (-ms-high-contrast:active){:host .nx-error__icon{color:windowText}}:host.nx-error--message{margin:12px 0}:host.nx-error--message nx-message{margin:0}"]
            }] }
];
/** @nocollapse */
NxErrorComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [ERROR_DEFAULT_OPTIONS,] }] }
];
NxErrorComponent.propDecorators = {
    showIcon: [{ type: Input }],
    id: [{ type: Input }],
    appearance: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._showIcon;
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._appearance;
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._id;
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    NxErrorComponent.prototype._defaultOptions;
}
