// @allianz/ngx-ndbx/esm2015/button/button-base.js
/**
 * @fileoverview added by tsickle
 * Generated from: button/button-base.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HostBinding } from '@angular/core';
/** @type {?} */
const DEFAULT_SIZE = 'medium';
/** @type {?} */
const DEFAULT_TYPE = 'primary';
/**
 * \@docs-private
 */
export class NxButtonBase {
    /**
     * @param {?} _changeDetectorRef
     * @param {?} _elementRef
     */
    constructor(_changeDetectorRef, _elementRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._elementRef = _elementRef;
        /**
         * \@docs-private
         */
        this.type = DEFAULT_TYPE;
        /**
         * \@docs-private
         */
        this.size = DEFAULT_SIZE;
        this.negative = false;
        this.block = false;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get isPrimary() { return this.type === 'primary'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isSecondary() { return this.type === 'secondary'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isTertiary() { return this.type === 'tertiary'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isCta() { return this.type === 'cta'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isEmphasis() { return this.type === 'emphasis'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isLarge() { return this.size === 'large'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isMedium() { return this.size === 'medium'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isSmallMedium() { return this.size === 'small-medium'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isSmall() { return this.size === 'small'; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isBlock() { return this.block; }
    /**
     * \@docs-private
     * @return {?}
     */
    get isNegative() { return this.negative; }
    /**
     * @param {?} value
     * @return {?}
     */
    set classNames(value) {
        if (this._classNames === value) {
            return;
        }
        this._classNames = value;
        const [type = null] = this._classNames.match(/primary|secondary|tertiary|cta|emphasis/) || [DEFAULT_TYPE];
        this.type = (/** @type {?} */ (type));
        const [size = null] = this._classNames.match(/small-medium|small|medium|large/) || [DEFAULT_SIZE];
        this.size = (/** @type {?} */ (size));
        this.negative = /negative/.test(this._classNames);
        this.block = /block/.test(this._classNames);
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get classNames() {
        return this._classNames;
    }
    /**
     * \@docs-private
     * getter used for the modal component as a quickfix
     * since button got changed from directive to component the reference used by the ngOpenModelOnClick directive
     * for nxButtons is a reference to component instance instead of an element reference. As a workaround we need a
     * way to reach the elementRef of the component until the modal gets refactored.
     * @return {?}
     */
    get elementRef() {
        return this._elementRef;
    }
}
NxButtonBase.propDecorators = {
    isPrimary: [{ type: HostBinding, args: ['class.nx-button--primary',] }],
    isSecondary: [{ type: HostBinding, args: ['class.nx-button--secondary',] }],
    isTertiary: [{ type: HostBinding, args: ['class.nx-button--tertiary',] }],
    isCta: [{ type: HostBinding, args: ['class.nx-button--cta',] }],
    isEmphasis: [{ type: HostBinding, args: ['class.nx-button--emphasis',] }],
    isLarge: [{ type: HostBinding, args: ['class.nx-button--large',] }],
    isMedium: [{ type: HostBinding, args: ['class.nx-button--medium',] }],
    isSmallMedium: [{ type: HostBinding, args: ['class.nx-button--small-medium',] }],
    isSmall: [{ type: HostBinding, args: ['class.nx-button--small',] }],
    isBlock: [{ type: HostBinding, args: ['class.nx-button--block',] }],
    isNegative: [{ type: HostBinding, args: ['class.nx-button--negative',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._classNames;
    /**
     * \@docs-private
     * @type {?}
     */
    NxButtonBase.prototype.type;
    /**
     * \@docs-private
     * @type {?}
     */
    NxButtonBase.prototype.size;
    /** @type {?} */
    NxButtonBase.prototype.negative;
    /** @type {?} */
    NxButtonBase.prototype.block;
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._elementRef;
}
