// @allianz/ngx-ndbx/esm2015/copytext/copytext.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: copytext/copytext.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
/** @type {?} */
const DEFAULT_TYPE = 'normal';
export class NxCopytextComponent {
    constructor() {
        /**
         * \@docs-private
         */
        this.type = DEFAULT_TYPE;
        /**
         * \@docs-private
         */
        this.negative = false;
    }
    /**
     * Sets the type and whether the negative set of styling should be used.
     *
     * You can combine the values of type NxCopytextType and 'negative'.
     *
     * Default type: 'normal'
     * @param {?} value
     * @return {?}
     */
    set classNames(value) {
        if (this._classNames === value) {
            return;
        }
        this._classNames = value;
        const [type = null] = this._classNames.match(/small|medium|normal|large/) || [DEFAULT_TYPE];
        this.type = (/** @type {?} */ (type));
        this.negative = !!this._classNames.match(/negative/);
    }
    /**
     * @return {?}
     */
    get classNames() {
        return this._classNames;
    }
}
NxCopytextComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'p[nxCopytext]',
                template: `<ng-content></ng-content>`,
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[class.nx-copy]': 'true',
                    '[class.nx-copy--small]': 'type === "small"',
                    '[class.nx-copy--medium]': 'type === "medium"',
                    '[class.nx-copy--normal]': 'type === "normal"',
                    '[class.nx-copy--large]': 'type === "large"',
                    '[class.nx-copy--negative]': 'negative',
                },
                styles: [":host{margin:0 0 32px;font-size:16px;font-size:var(--paragraph-03-font-size,16px);line-height:24px;line-height:var(--paragraph-03-line-height,24px);font-weight:400;font-weight:var(--paragraph-03-font-weight,400);letter-spacing:0;letter-spacing:var(--paragraph-03-letter-spacing,0)}:host.nx-copy.nx-copy{font-weight:400}:host(.nx-copy--negative){color:#fff}:host(.nx-copy--small){font-size:12px;font-size:var(--paragraph-05-font-size,12px);line-height:16px;line-height:var(--paragraph-05-line-height,16px);font-weight:600;font-weight:var(--paragraph-05-font-weight,600);letter-spacing:.5px;letter-spacing:var(--paragraph-05-letter-spacing,.5px)}:host(.nx-copy--medium){font-size:14px;font-size:var(--paragraph-04-font-size,14px);line-height:20px;line-height:var(--paragraph-04-line-height,20px);font-weight:600;font-weight:var(--paragraph-04-font-weight,600);letter-spacing:.2px;letter-spacing:var(--paragraph-04-letter-spacing,.2px)}:host(.nx-copy--large){font-size:18px;font-size:var(--paragraph-02-font-size,18px);line-height:24px;line-height:var(--paragraph-02-line-height,24px);font-weight:400;font-weight:var(--paragraph-02-font-weight,400);letter-spacing:.2px;letter-spacing:var(--paragraph-02-letter-spacing,.2px)}"]
            }] }
];
NxCopytextComponent.propDecorators = {
    classNames: [{ type: Input, args: ['nxCopytext',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxCopytextComponent.prototype._classNames;
    /**
     * \@docs-private
     * @type {?}
     */
    NxCopytextComponent.prototype.type;
    /**
     * \@docs-private
     * @type {?}
     */
    NxCopytextComponent.prototype.negative;
}
