// @allianz/ngx-ndbx/esm5/moment-date-adapter/adapter/index.js
/**
 * @fileoverview added by tsickle
 * Generated from: adapter/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NxDateAdapter, NX_DATE_LOCALE, NX_DATE_FORMATS, NX_DATE_LOCALE_PROVIDER } from '@allianz/ngx-ndbx/datefield';
import { NgModule } from '@angular/core';
import { NxMomentDateAdapter } from './moment-date-adapter';
import { NX_MOMENT_DATE_FORMATS } from './moment-date-formats';
export { NxMomentDateAdapter } from './moment-date-adapter';
export { NX_MOMENT_DATE_FORMATS } from './moment-date-formats';
var MomentDateModule = /** @class */ (function () {
    function MomentDateModule() {
    }
    MomentDateModule.decorators = [
        { type: NgModule, args: [{
                    providers: [
                        NX_DATE_LOCALE_PROVIDER,
                        { provide: NxDateAdapter, useClass: NxMomentDateAdapter, deps: [NX_DATE_LOCALE] }
                    ],
                },] }
    ];
    return MomentDateModule;
}());
export { MomentDateModule };
var ɵ0 = NX_MOMENT_DATE_FORMATS;
var NxMomentDateModule = /** @class */ (function () {
    function NxMomentDateModule() {
    }
    NxMomentDateModule.decorators = [
        { type: NgModule, args: [{
                    imports: [MomentDateModule],
                    providers: [{ provide: NX_DATE_FORMATS, useValue: ɵ0 }],
                },] }
    ];
    return NxMomentDateModule;
}());
export { NxMomentDateModule };
export { ɵ0 };
