// @allianz/ngx-ndbx/esm5/message/message-toast/message-toast-config.js
/**
 * @fileoverview added by tsickle
 * Generated from: message/message-toast/message-toast-config.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Configuration used when opening a message toast.
 */
var /**
 * Configuration used when opening a message toast.
 */
NxMessageToastConfig = /** @class */ (function () {
    function NxMessageToastConfig() {
        /**
         * The politeness level for the LiveAnnouncer announcement.
         *
         *  Default value: 'polite'.
         */
        this.politeness = 'polite';
        /**
         * Message to be announced by the LiveAnnouncer. When opening a toast message without a custom
         * component or template, the announcement message will default to the specified message.
         *
         * Default value: ''.
         */
        this.announcementMessage = '';
        /**
         * The length of time in milliseconds to wait before automatically dismissing the message toast.
         *
         * Default value: 3000.
         */
        this.duration = 3000;
        /**
         * Context of the message toast.
         *
         * Default value: 'info'.
         */
        this.context = 'info';
    }
    return NxMessageToastConfig;
}());
/**
 * Configuration used when opening a message toast.
 */
export { NxMessageToastConfig };
if (false) {
    /**
     * The politeness level for the LiveAnnouncer announcement.
     *
     *  Default value: 'polite'.
     * @type {?}
     */
    NxMessageToastConfig.prototype.politeness;
    /**
     * Message to be announced by the LiveAnnouncer. When opening a toast message without a custom
     * component or template, the announcement message will default to the specified message.
     *
     * Default value: ''.
     * @type {?}
     */
    NxMessageToastConfig.prototype.announcementMessage;
    /**
     * The length of time in milliseconds to wait before automatically dismissing the message toast.
     *
     * Default value: 3000.
     * @type {?}
     */
    NxMessageToastConfig.prototype.duration;
    /**
     * Context of the message toast.
     *
     * Default value: 'info'.
     * @type {?}
     */
    NxMessageToastConfig.prototype.context;
}
/**
 * \@docs-private
 * Needed so that the user text data can be injected in the message toastcomponent
 */
var /**
 * \@docs-private
 * Needed so that the user text data can be injected in the message toastcomponent
 */
NxMessageToastData = /** @class */ (function () {
    function NxMessageToastData(data) {
        this.data = data;
    }
    return NxMessageToastData;
}());
/**
 * \@docs-private
 * Needed so that the user text data can be injected in the message toastcomponent
 */
export { NxMessageToastData };
if (false) {
    /** @type {?} */
    NxMessageToastData.prototype.data;
}
