// @allianz/ngx-ndbx/esm2015/tabs/tab-body.js
/**
 * @fileoverview added by tsickle
 * Generated from: tabs/tab-body.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ChangeDetectionStrategy, Component, Input, ViewChild, ViewContainerRef } from '@angular/core';
import { NxTabComponent } from './tab';
import { NxTabGroupBase } from './tab-group-base';
/**
 * \@docs-private
 */
export class NxTabBodyComponent {
    /**
     * @param {?} _tabGroup
     */
    constructor(_tabGroup) {
        this._tabGroup = _tabGroup;
        this._active = false;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set active(value) {
        this._active = coerceBooleanProperty(value);
        if (this._active) {
            this.attach();
        }
        else {
            this.detach();
        }
    }
    /**
     * @return {?}
     */
    get active() {
        return this._active;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._appearanceSubscription = ((/** @type {?} */ (this._tabGroup)))._appearanceChange.subscribe((/**
         * @return {?}
         */
        () => {
            this.detach();
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._appearanceSubscription.unsubscribe();
    }
    /**
     * @return {?}
     */
    attach() {
        this._outlet.insert(this.tab.contentViewRef);
    }
    /**
     * @return {?}
     */
    detach() {
        /** @type {?} */
        const index = this._outlet.indexOf(this.tab.contentViewRef);
        if (index !== -1) {
            this._outlet.detach(index);
        }
    }
}
NxTabBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-tab-body',
                template: "<ng-container #outlet></ng-container>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{display:block}"]
            }] }
];
/** @nocollapse */
NxTabBodyComponent.ctorParameters = () => [
    { type: NxTabGroupBase }
];
NxTabBodyComponent.propDecorators = {
    _outlet: [{ type: ViewChild, args: ['outlet', { static: true, read: ViewContainerRef },] }],
    tab: [{ type: Input }],
    active: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxTabBodyComponent.prototype._appearanceSubscription;
    /** @type {?} */
    NxTabBodyComponent.prototype._outlet;
    /** @type {?} */
    NxTabBodyComponent.prototype.tab;
    /**
     * @type {?}
     * @private
     */
    NxTabBodyComponent.prototype._active;
    /**
     * @type {?}
     * @private
     */
    NxTabBodyComponent.prototype._tabGroup;
}
