// @allianz/ngx-ndbx/esm2015/tabs/tab-header-outlet.js
/**
 * @fileoverview added by tsickle
 * Generated from: tabs/tab-header-outlet.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EmbeddedViewRef, Input, ViewChild, ViewContainerRef } from '@angular/core';
import { NxTabGroupBase } from './tab-group-base';
/**
 * \@docs-private
 */
export class NxTabHeaderOutletComponent {
    /**
     * @param {?} _tabGroup
     */
    constructor(_tabGroup) {
        this._tabGroup = _tabGroup;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._appearanceSubscription = ((/** @type {?} */ (this._tabGroup)))._appearanceChange.subscribe((/**
         * @return {?}
         */
        () => {
            this.detach();
        }));
        this.attach();
    }
    /**
     * @return {?}
     */
    attach() {
        this._outlet.insert(this.content);
    }
    /**
     * @return {?}
     */
    detach() {
        /** @type {?} */
        const index = this._outlet.indexOf(this.content);
        if (index !== -1) {
            this._outlet.detach(index);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._appearanceSubscription.unsubscribe();
    }
}
NxTabHeaderOutletComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-tab-header-outlet',
                template: '<ng-container #outlet></ng-container>'
            }] }
];
/** @nocollapse */
NxTabHeaderOutletComponent.ctorParameters = () => [
    { type: NxTabGroupBase }
];
NxTabHeaderOutletComponent.propDecorators = {
    _outlet: [{ type: ViewChild, args: ['outlet', { static: true, read: ViewContainerRef },] }],
    content: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NxTabHeaderOutletComponent.prototype._outlet;
    /** @type {?} */
    NxTabHeaderOutletComponent.prototype.content;
    /**
     * @type {?}
     * @private
     */
    NxTabHeaderOutletComponent.prototype._appearanceSubscription;
    /**
     * @type {?}
     * @private
     */
    NxTabHeaderOutletComponent.prototype._tabGroup;
}
