// @allianz/ngx-ndbx/esm2015/popover/popover.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: popover/popover.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ContentChild, EventEmitter, Output, TemplateRef, ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
import { NxPopoverContentDirective } from './popover-content';
import { ENTER, SPACE } from '@angular/cdk/keycodes';
export class NxPopoverComponent {
    constructor() {
        /**
         * Event emitted when the popover is closed.
         */
        this.closed = new EventEmitter();
        /**
         * \@docs-private
         */
        this.closeButtonClick = new Subject();
        /**
         * \@docs-private
         */
        this.showCloseButton = false;
        /**
         * \@docs-private
         */
        this.arrowStyle = {};
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.closed.complete();
    }
    // emit to notify the popover trigger directive that the close button was clicked
    /**
     * \@docs-private
     * @return {?}
     */
    emitCloseButtonClick() {
        this.closeButtonClick.next();
    }
    /**
     * \@docs-private
     * @param {?} $event
     * @return {?}
     */
    _onCloseKeyup($event) {
        if ($event && ($event.keyCode === ENTER || $event.keyCode === SPACE)) {
            this.emitCloseButtonClick();
        }
        $event.preventDefault();
    }
    /**
     * \@docs-private
     * @return {?}
     */
    emitClosedEvent() {
        this.closed.emit();
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get classList() {
        if (this.direction) {
            return [`nx-popover--${this.direction}`];
        }
    }
    /**
     * Prevent the popover from closing when the user clicks on the popover content.
     * @param {?} event
     * @return {?}
     */
    _onClick(event) {
        event.stopPropagation();
    }
}
NxPopoverComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-popover',
                template: "<ng-template>\n  <div [ngClass]=\"classList\" role=\"tooltip\" aria-hidden=\"false\" (click)=\"_onClick($event)\">\n    <div class=\"nx-popover__content\" tabindex=\"0\">\n        <div class=\"nx-popover__arrow\" [ngStyle]=\"arrowStyle\">\n          </div>\n      <span\n          aria-label=\"close\"\n          tabindex=\"0\"\n          role=\"button\"\n          class=\"nx-popover__close-icon\"\n          (click)=\"emitCloseButtonClick()\"\n          (keyup)=\"_onCloseKeyup($event)\"\n          *ngIf=\"showCloseButton\">\n        <nx-icon aria-hidden=\"true\" name=\"close\"></nx-icon>\n        </span>\n\n      <span [id]=\"id\">\n        <ng-content></ng-content>\n        <ng-container [ngTemplateOutlet]=\"_lazyContent?._template\"></ng-container>\n      </span>\n    </div>\n  </div>\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'nxPopover',
                styles: [".nx-popover__content{font-size:16px;font-size:var(--popover-content-font-size,16px);line-height:24px;line-height:var(--popover-content-line-height,24px);font-weight:400;font-weight:var(--popover-content-font-weight,400);letter-spacing:0;letter-spacing:var(--popover-content-letter-spacing,0);background-color:#fff;background-color:var(--popover-background-color,#fff);color:#414141;color:var(--popover-text-color,#414141);border:1px solid #d9d9d9;border:1px solid var(--popover-border-color,#d9d9d9);border-radius:4px;box-shadow:0 2px 4px rgba(65,65,65,.5);padding:23px;display:inline-block}.nx-popover__content:focus{outline:0}.nx-popover__content::-moz-focus-inner{border:0}.nx-popover__close-icon{display:flex;position:absolute;top:4px;right:4px;cursor:pointer;font-size:16px;font-size:var(--popover-close-icon-size,16px);color:#d9d9d9;color:var(--popover-close-icon-color,#d9d9d9);border-radius:2px}.nx-popover__close-icon:focus{outline:0}.nx-popover__close-icon::-moz-focus-inner{border:0}::ng-deep [data-whatinput=keyboard] .nx-popover__close-icon:focus,::ng-deep [data-whatinput=keyboard] .nx-popover__content:focus{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6)}@media screen and (-ms-high-contrast:active){::ng-deep [data-whatinput=keyboard] .nx-popover__close-icon:focus,::ng-deep [data-whatinput=keyboard] .nx-popover__content:focus{box-shadow:0 0 0 2px background,0 0 0 6px windowText}}.nx-popover__arrow{width:12px;height:12px;background-color:inherit;position:absolute;transform:translate(-50%,0);border-top:1px solid #d9d9d9;border-top:1px solid var(--popover-border-color,#d9d9d9);border-left:1px solid #d9d9d9;border-left:1px solid var(--popover-border-color,#d9d9d9)}.nx-popover--top .nx-popover__arrow{transform:translate(-50%,0) rotate(225deg);bottom:-5px;box-shadow:2px 2px 1px 1px #fff,-2px -2px 4px rgba(65,65,65,.5);box-shadow:2px 2px 1px 1px var(--popover-background-color,#fff),-2px -2px 4px rgba(65,65,65,.5)}.nx-popover--right .nx-popover__arrow{transform:translate(0,-50%) rotate(-45deg);left:-5px}.nx-popover--bottom .nx-popover__arrow{transform:translate(-50%,0) rotate(45deg);top:-5px}.nx-popover--left .nx-popover__arrow{transform:translate(0,-50%) rotate(135deg);right:-5px}"]
            }] }
];
/** @nocollapse */
NxPopoverComponent.ctorParameters = () => [];
NxPopoverComponent.propDecorators = {
    templateRef: [{ type: ViewChild, args: [TemplateRef, { static: false },] }],
    _lazyContent: [{ type: ContentChild, args: [NxPopoverContentDirective, { static: false },] }],
    closed: [{ type: Output, args: ['nxClosed',] }]
};
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.templateRef;
    /**
     * Content that will be rendered lazily.
     * @type {?}
     */
    NxPopoverComponent.prototype._lazyContent;
    /**
     * Event emitted when the popover is closed.
     * @type {?}
     */
    NxPopoverComponent.prototype.closed;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.closeButtonClick;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.id;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.direction;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.showCloseButton;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPopoverComponent.prototype.arrowStyle;
}
