// @allianz/ngx-ndbx/esm5/table/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { NxTableModule } from './table.module';
export { NxTableCellComponent } from './table-cell.component';
export { NxHeaderCellDirective } from './header-cell.directive';
export { NxTableRowComponent } from './table-row.component';
export { NxTableComponent } from './table.component';
export { NxToggleButtonComponent } from './expandable/toggle-button.component';
export { NxExpandableTableCellComponent } from './expandable/expandable-table-cell.component';
export { NxExpandableTableRowComponent } from './expandable/expandable-table-row.component';
export { NxExpandableTableDirective } from './expandable/expandable-table.directive';
export { NxSortHeaderComponent } from './sort-header/sort-header.component';
export { SortEvent, NxSortDirective } from './sort-header/sort.directive';
export { NxSortHeaderIntl } from './sort-header/sort-header-intl';
