// @allianz/ngx-ndbx/esm5/table/table-row.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: table-row.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input, ChangeDetectorRef, Output, ElementRef } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { EventEmitter } from '@angular/core';
import { SPACE } from '@angular/cdk/keycodes';
/**
 * This is a table row.
 */
var NxTableRowComponent = /** @class */ (function () {
    function NxTableRowComponent(_changeDetectorRef, _elementRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._elementRef = _elementRef;
        this._selectable = false;
        this._selected = false;
        this.select = new EventEmitter();
    }
    Object.defineProperty(NxTableRowComponent.prototype, "selectable", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selectable;
        },
        /** Whether this table row is selectable */
        set: /**
         * Whether this table row is selectable
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._selectable = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxTableRowComponent.prototype, "selected", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selected;
        },
        /** Whether this table row is selected */
        set: /**
         * Whether this table row is selected
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._selected = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} $event
     * @return {?}
     */
    NxTableRowComponent.prototype._onSelect = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (!this._selectable || this.isActionEvent($event)) {
            return;
        }
        if ($event.keyCode === SPACE) {
            $event.preventDefault();
        }
        this.select.emit();
    };
    /**
     * Checks if the event would trigger an action.
     * Return `true` if a button, link, input or label are clicked.
     */
    /**
     * Checks if the event would trigger an action.
     * Return `true` if a button, link, input or label are clicked.
     * @private
     * @param {?} $event
     * @return {?}
     */
    NxTableRowComponent.prototype.isActionEvent = /**
     * Checks if the event would trigger an action.
     * Return `true` if a button, link, input or label are clicked.
     * @private
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var parent = $event.target;
        while (parent && parent !== this._elementRef.nativeElement) {
            if (['A', 'INPUT', 'BUTTON'].indexOf(parent.tagName) >= 0) {
                return true;
            }
            else if (parent.tagName === 'LABEL' && parent.getAttribute('for')) {
                return true;
            }
            parent = parent.parentElement;
        }
        return false;
    };
    NxTableRowComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'tr[nxTableRow]',
                    host: {
                        'class': 'nx-table-row',
                        '[class.is-selected]': 'selected',
                        '[class.nx-table-row--selectable]': 'selectable',
                        '[attr.aria-selected]': 'selected',
                        '(click)': '_onSelect($event)',
                        '(keydown.space)': '_onSelect($event)'
                    },
                    template: "<ng-content></ng-content>",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    NxTableRowComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: ElementRef }
    ]; };
    NxTableRowComponent.propDecorators = {
        selectable: [{ type: Input }],
        selected: [{ type: Input }],
        select: [{ type: Output }]
    };
    return NxTableRowComponent;
}());
export { NxTableRowComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxTableRowComponent.prototype._selectable;
    /**
     * @type {?}
     * @private
     */
    NxTableRowComponent.prototype._selected;
    /** @type {?} */
    NxTableRowComponent.prototype.select;
    /**
     * @type {?}
     * @protected
     */
    NxTableRowComponent.prototype._changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    NxTableRowComponent.prototype._elementRef;
}
