// @allianz/ngx-ndbx/esm2015/mask/iban-mask.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: iban-mask.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Inject, forwardRef } from '@angular/core';
import * as IBAN from 'iban';
import { NG_VALIDATORS } from '@angular/forms';
import { NxMaskDirective } from './mask.directive';
/** @type {?} */
export const NX_IBAN_MASK_VALIDATORS = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => NxIbanMaskDirective)),
    multi: true
};
/**
 * To use the `NxIbanMaskDirective`, you have to install the **peer dependency** `iban.js`.
 */
export class NxIbanMaskDirective {
    /**
     * @param {?} _elementRef
     * @param {?} maskDirective
     */
    constructor(_elementRef, maskDirective) {
        this._elementRef = _elementRef;
        this.maskDirective = maskDirective;
        this._afterInputHook = (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            /** @type {?} */
            const input = (/** @type {?} */ (event.target));
            this._setCountryCode(input.value.substr(0, 2));
        });
        this._beforePasteHook = (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            // change the country code here if necessary
            /** @type {?} */
            const input = (/** @type {?} */ (event.target));
            /** @type {?} */
            const pastedData = (event.clipboardData || ((/** @type {?} */ (window))).clipboardData).getData('text');
            /** @type {?} */
            const enteredCountryCode = (this.maskDirective.elementRefValue.substr(0, input.selectionStart)
                + this.maskDirective.getMaskedString(pastedData, input.selectionStart)).substr(0, 2);
            this._setCountryCode(enteredCountryCode);
        });
        this.maskDirective.registerAfterInputHook(this._afterInputHook);
        this.maskDirective.registerBeforePasteHook(this._beforePasteHook);
        this.maskDirective.cvaModelChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        (value) => {
            /** @type {?} */
            const enteredCountryCode = this.maskDirective.getMaskedString(value).substr(0, 2);
            this._setCountryCode(enteredCountryCode);
        }));
    }
    /**
     * @private
     * @param {?} code
     * @return {?}
     */
    _setCountryCode(code) {
        code = code.toUpperCase();
        if (code.length === 2 && this._countryCode !== code) {
            if (this._countryCodeExists(code)) {
                this._countryCode = code;
                this.maskDirective.setMask(this._getMask(this._countryCode));
            }
            else {
                this._countryCode = null;
                this.maskDirective.setMask('SS');
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // set only first two letters as I don't know a country yet
        this.maskDirective.mask = 'SS';
        this.maskDirective.convertTo = 'upper';
    }
    /**
     * @private
     * @param {?} countryCode
     * @return {?}
     */
    _getMask(countryCode) {
        // the countrySpecs of a country contain: countryCode ("DE"), length (22), structure ("F08F10")
        // and an example belonging to each country
        /** @type {?} */
        const countrySpecs = IBAN['countries'][countryCode];
        // 'SS' for country code + '00' for IBAN checksum
        /** @type {?} */
        let mask = 'SS00';
        // split up after every third character
        /** @type {?} */
        const characterDefs = countrySpecs['structure'].match(/.{1,3}/g);
        characterDefs.forEach((/**
         * @param {?} charDef
         * @return {?}
         */
        charDef => {
            /** @type {?} */
            const character = charDef[0];
            /** @type {?} */
            const count = Number(charDef.substring(1, 3));
            switch (character) {
                // [0-9]
                case 'F':
                    mask = mask + '0'.repeat(count);
                    break;
                // [0-9A-Za-z]
                case 'A':
                    mask = mask + 'A'.repeat(count);
                    break;
                // [A-Z]
                // 'S' in nxMask does accept also [a-z].
                // There is no option for only accepting capital letters at the moment.
                case 'U':
                    mask = mask + 'S'.repeat(count);
                    break;
            }
        }));
        // insert whitespaces after every 4 characters
        mask = mask.match(/.{1,4}/g).join(' ');
        return mask;
    }
    /**
     * @private
     * @param {?} countryCode
     * @return {?}
     */
    _countryCodeExists(countryCode) {
        return (!!IBAN['countries'][countryCode]);
    }
    /**
     * @private
     * @return {?}
     */
    _validateFn() {
        /** @type {?} */
        const enteredCountryCode = this._elementRef.nativeElement.value.substr(0, 2);
        if (enteredCountryCode.length === 2 && !this._countryCodeExists(enteredCountryCode)) {
            // immediately show error to user
            this.maskDirective._touch();
            return { nxIbanInvalidCountryError: 'no valid country code' };
        }
        if (!IBAN.isValid(this.maskDirective.getUnmaskedValue())) {
            return { nxIbanParseError: 'no valid iban' };
        }
        return null;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    validate() {
        return this.maskDirective.validateMask ? this._validateFn() : null;
    }
}
NxIbanMaskDirective.decorators = [
    { type: Directive, args: [{
                selector: 'input[nxIbanMask]',
                exportAs: 'nxIbanMaskDirective',
                providers: [
                    NX_IBAN_MASK_VALIDATORS
                ]
            },] }
];
/** @nocollapse */
NxIbanMaskDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NxMaskDirective, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => NxMaskDirective)),] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxIbanMaskDirective.prototype._countryCode;
    /**
     * @type {?}
     * @private
     */
    NxIbanMaskDirective.prototype._afterInputHook;
    /**
     * @type {?}
     * @private
     */
    NxIbanMaskDirective.prototype._beforePasteHook;
    /**
     * @type {?}
     * @private
     */
    NxIbanMaskDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    NxIbanMaskDirective.prototype.maskDirective;
}
