// @allianz/ngx-ndbx/esm5/badge/badge.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: badge.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
var NxBadgeComponent = /** @class */ (function () {
    function NxBadgeComponent(_changeDetRef) {
        this._changeDetRef = _changeDetRef;
        this._vibrant = false;
    }
    Object.defineProperty(NxBadgeComponent.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () {
            return this._type;
        },
        /** Sets the class name for the badge element.  */
        set: /**
         * Sets the class name for the badge element.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this._type) {
                this._type = value;
                this._changeDetRef.markForCheck();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxBadgeComponent.prototype, "vibrant", {
        get: /**
         * @return {?}
         */
        function () {
            return this._vibrant;
        },
        /** Change badge style to vibrant. */
        set: /**
         * Change badge style to vibrant.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var newValue = coerceBooleanProperty(value);
            if (value !== this._vibrant) {
                this._vibrant = newValue;
                this._changeDetRef.markForCheck();
            }
        },
        enumerable: true,
        configurable: true
    });
    NxBadgeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-badge',
                    template: '<ng-content></ng-content>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.nx-badge--active]': 'type === "active"',
                        '[class.nx-badge--positive]': 'type === "positive"',
                        '[class.nx-badge--critical]': 'type === "critical"',
                        '[class.nx-badge--negative]': 'type === "negative"',
                        '[class.nx-badge--vibrant]': 'vibrant'
                    },
                    styles: [":host{font-size:16px;font-size:var(--badge-font-size,16px);line-height:24px;line-height:var(--badge-line-height,24px);font-weight:400;font-weight:var(--badge-font-weight,400);letter-spacing:0;letter-spacing:var(--badge-letter-spacing,0);display:inline-block;padding:0 8px;white-space:nowrap;border-radius:12px;min-width:24px;color:#414141;color:var(--badge-color,#414141);background-color:#d9d9d9;background-color:var(--badge-background-color,#d9d9d9)}@media screen and (-ms-high-contrast:active){:host{border:1px solid windowText}}:host.nx-badge--active{background-color:#cfe9ee;background-color:var(--badge-active-background-color,#cfe9ee);color:#414141;color:var(--badge-active-color,#414141)}:host.nx-badge--positive{background-color:#e3ebaf;background-color:var(--badge-positive-background-color,#e3ebaf);color:#414141;color:var(--badge-positive-color,#414141)}:host.nx-badge--critical{background-color:#f7c7c3;background-color:var(--badge-critical-background-color,#f7c7c3);color:#414141;color:var(--badge-critical-color,#414141)}:host.nx-badge--negative{background-color:#ffe8b0;background-color:var(--badge-negative-background-color,#ffe8b0);color:#414141;color:var(--badge-negative-color,#414141)}:host.nx-badge--vibrant{color:#fff;color:var(--badge-vibrant-color,#fff);background-color:#414141;background-color:var(--badge-vibrant-background-color,#414141)}:host.nx-badge--vibrant.nx-badge--active{background-color:#007ab3;background-color:var(--badge-vibrant-active-background-color,#007ab3);color:#fff;color:var(--badge-vibrant-active-color,#fff)}:host.nx-badge--vibrant.nx-badge--positive{background-color:#1e8927;background-color:var(--badge-vibrant-positive-background-color,#1e8927);color:#fff;color:var(--badge-vibrant-positive-color,#fff)}:host.nx-badge--vibrant.nx-badge--critical{background-color:#dc3149;background-color:var(--badge-vibrant-critical-background-color,#dc3149);color:#fff;color:var(--badge-vibrant-critical-color,#fff)}:host.nx-badge--vibrant.nx-badge--negative{background-color:#efbe25;background-color:var(--badge-vibrant-negative-background-color,#efbe25);color:#414141;color:var(--badge-vibrant-negative-color,#414141)}"]
                }] }
    ];
    /** @nocollapse */
    NxBadgeComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxBadgeComponent.propDecorators = {
        type: [{ type: Input }],
        vibrant: [{ type: Input }]
    };
    return NxBadgeComponent;
}());
export { NxBadgeComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxBadgeComponent.prototype._type;
    /**
     * @type {?}
     * @private
     */
    NxBadgeComponent.prototype._vibrant;
    /**
     * @type {?}
     * @private
     */
    NxBadgeComponent.prototype._changeDetRef;
}
