// @allianz/ngx-ndbx/esm5/button/button-base.js
/**
 * @fileoverview added by tsickle
 * Generated from: button/button-base.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { HostBinding } from '@angular/core';
/** @type {?} */
var DEFAULT_SIZE = 'medium';
/** @type {?} */
var DEFAULT_TYPE = 'primary';
/**
 * \@docs-private
 */
var NxButtonBase = /** @class */ (function () {
    function NxButtonBase(_changeDetectorRef, _elementRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._elementRef = _elementRef;
        /**
         * \@docs-private
         */
        this.type = DEFAULT_TYPE;
        /**
         * \@docs-private
         */
        this.size = DEFAULT_SIZE;
        this.negative = false;
        this.block = false;
    }
    Object.defineProperty(NxButtonBase.prototype, "isPrimary", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.type === 'primary'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isSecondary", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.type === 'secondary'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isTertiary", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.type === 'tertiary'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isCta", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.type === 'cta'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isEmphasis", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.type === 'emphasis'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isLarge", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.size === 'large'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isMedium", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.size === 'medium'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isSmallMedium", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.size === 'small-medium'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isSmall", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.size === 'small'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isBlock", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.block; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "isNegative", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.negative; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "classNames", {
        get: /**
         * @return {?}
         */
        function () {
            return this._classNames;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this._classNames === value) {
                return;
            }
            this._classNames = value;
            var _a = tslib_1.__read(this._classNames.match(/primary|secondary|tertiary|cta|emphasis/) || [DEFAULT_TYPE], 1), _b = _a[0], type = _b === void 0 ? null : _b;
            this.type = (/** @type {?} */ (type));
            var _c = tslib_1.__read(this._classNames.match(/small-medium|small|medium|large/) || [DEFAULT_SIZE], 1), _d = _c[0], size = _d === void 0 ? null : _d;
            this.size = (/** @type {?} */ (size));
            this.negative = /negative/.test(this._classNames);
            this.block = /block/.test(this._classNames);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxButtonBase.prototype, "elementRef", {
        /**
         * @docs-private
         * getter used for the modal component as a quickfix
         * since button got changed from directive to component the reference used by the ngOpenModelOnClick directive
         * for nxButtons is a reference to component instance instead of an element reference. As a workaround we need a
         * way to reach the elementRef of the component until the modal gets refactored.
         */
        get: /**
         * \@docs-private
         * getter used for the modal component as a quickfix
         * since button got changed from directive to component the reference used by the ngOpenModelOnClick directive
         * for nxButtons is a reference to component instance instead of an element reference. As a workaround we need a
         * way to reach the elementRef of the component until the modal gets refactored.
         * @return {?}
         */
        function () {
            return this._elementRef;
        },
        enumerable: true,
        configurable: true
    });
    NxButtonBase.propDecorators = {
        isPrimary: [{ type: HostBinding, args: ['class.nx-button--primary',] }],
        isSecondary: [{ type: HostBinding, args: ['class.nx-button--secondary',] }],
        isTertiary: [{ type: HostBinding, args: ['class.nx-button--tertiary',] }],
        isCta: [{ type: HostBinding, args: ['class.nx-button--cta',] }],
        isEmphasis: [{ type: HostBinding, args: ['class.nx-button--emphasis',] }],
        isLarge: [{ type: HostBinding, args: ['class.nx-button--large',] }],
        isMedium: [{ type: HostBinding, args: ['class.nx-button--medium',] }],
        isSmallMedium: [{ type: HostBinding, args: ['class.nx-button--small-medium',] }],
        isSmall: [{ type: HostBinding, args: ['class.nx-button--small',] }],
        isBlock: [{ type: HostBinding, args: ['class.nx-button--block',] }],
        isNegative: [{ type: HostBinding, args: ['class.nx-button--negative',] }]
    };
    return NxButtonBase;
}());
export { NxButtonBase };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._classNames;
    /**
     * \@docs-private
     * @type {?}
     */
    NxButtonBase.prototype.type;
    /**
     * \@docs-private
     * @type {?}
     */
    NxButtonBase.prototype.size;
    /** @type {?} */
    NxButtonBase.prototype.negative;
    /** @type {?} */
    NxButtonBase.prototype.block;
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    NxButtonBase.prototype._elementRef;
}
