// @allianz/ngx-ndbx/esm5/breadcrumb/breadcrumb-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: breadcrumb-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Renderer2, ElementRef, ChangeDetectionStrategy } from '@angular/core';
var NxBreadcrumbItemComponent = /** @class */ (function () {
    function NxBreadcrumbItemComponent(_renderer, _elemRef) {
        this._renderer = _renderer;
        this._elemRef = _elemRef;
    }
    /**@docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxBreadcrumbItemComponent.prototype.setAsLast = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this._renderer.setAttribute(this._elemRef.nativeElement, 'aria-current', 'page');
    };
    /**@docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxBreadcrumbItemComponent.prototype.resetAriaLabel = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this._renderer.removeAttribute(this._elemRef.nativeElement, 'aria-current');
    };
    // prevent location change when the user clicks on the chevron arrow
    // prevent location change when the user clicks on the chevron arrow
    /**
     * @param {?} event
     * @return {?}
     */
    NxBreadcrumbItemComponent.prototype._onIconClick = 
    // prevent location change when the user clicks on the chevron arrow
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
        event.preventDefault();
    };
    NxBreadcrumbItemComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'a[nxBreadcrumbItem]',
                    template: "\n<span class=\"nx-breadcrumb-item__text\"><ng-content></ng-content></span>\n<nx-icon\n    (click)=\"_onIconClick($event)\"\n    class=\"nx-breadcrumb-item__chevron\"\n    aria-hidden=\"true\"\n    name=\"chevron-right\"\n    size=\"xs\">\n</nx-icon>",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        'class': 'nx-breadcrumb-item',
                    },
                    styles: [":host{text-decoration:none;display:flex;align-items:center;cursor:auto;color:#414141;color:var(--breadcrumb-color,#414141);font-size:16px;font-size:var(--breadcrumb-font-size,16px);line-height:24px;line-height:var(--breadcrumb-line-height,24px);font-weight:400;font-weight:var(--breadcrumb-font-weight,400);letter-spacing:0;letter-spacing:var(--breadcrumb-letter-spacing,0)}.nx-breadcrumb-item__chevron{margin-left:4px}.nx-breadcrumb-item__text{cursor:pointer}:host-context([data-whatinput=keyboard]):focus .nx-breadcrumb-item__text{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6)}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]):focus .nx-breadcrumb-item__text{box-shadow:0 0 0 2px background,0 0 0 6px windowText}}"]
                }] }
    ];
    /** @nocollapse */
    NxBreadcrumbItemComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    return NxBreadcrumbItemComponent;
}());
export { NxBreadcrumbItemComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxBreadcrumbItemComponent.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    NxBreadcrumbItemComponent.prototype._elemRef;
}
