// @allianz/ngx-ndbx/esm5/breadcrumb/breadcrumb.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: breadcrumb.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, QueryList, Input, ChangeDetectorRef, ChangeDetectionStrategy } from '@angular/core';
import { NxBreadcrumbItemComponent } from './breadcrumb-item.component';
import { startWith, takeUntil, filter } from 'rxjs/operators';
import { Subject } from 'rxjs';
var NxBreadcrumbComponent = /** @class */ (function () {
    function NxBreadcrumbComponent(_cdr) {
        this._cdr = _cdr;
        this._destroyed = new Subject();
        this._negative = false;
    }
    Object.defineProperty(NxBreadcrumbComponent.prototype, "negative", {
        get: /**
         * @return {?}
         */
        function () {
            return this._negative;
        },
        /** Whether the component uses the negative styling. */
        set: /**
         * Whether the component uses the negative styling.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._negative = value;
            this._cdr.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxBreadcrumbComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.breadcrumbItems.length === 0) {
            console.warn('A breadcrumb needs NxBreadcrumbItemComponent children wrapped in <li>!');
        }
        this.breadcrumbItems.changes
            .pipe(startWith(this.breadcrumbItems), filter((/**
         * @param {?} items
         * @return {?}
         */
        function (items) { return items.length !== 0; })), takeUntil(this._destroyed))
            .subscribe((/**
         * @param {?} items
         * @return {?}
         */
        function (items) {
            _this.breadcrumbItems.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.resetAriaLabel(); }));
            _this.breadcrumbItems.last.setAsLast();
        }));
    };
    /**
     * @return {?}
     */
    NxBreadcrumbComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._destroyed.next();
        this._destroyed.complete();
    };
    NxBreadcrumbComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'ol[nxBreadcrumb]',
                    template: "<ng-content></ng-content>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.is-negative]': 'negative'
                    },
                    styles: [":host{display:flex;list-style:none;height:24px;padding:0}::ng-deep li:last-child ::ng-deep .nx-breadcrumb-item{font-weight:600;cursor:auto}::ng-deep li:last-child ::ng-deep .nx-breadcrumb-item__chevron{display:none}::ng-deep li+li ::ng-deep .nx-breadcrumb-item{margin-left:4px}:host(.is-negative) ::ng-deep .nx-breadcrumb-item{color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    NxBreadcrumbComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxBreadcrumbComponent.propDecorators = {
        negative: [{ type: Input }],
        breadcrumbItems: [{ type: ContentChildren, args: [NxBreadcrumbItemComponent, { descendants: true },] }]
    };
    return NxBreadcrumbComponent;
}());
export { NxBreadcrumbComponent };
if (false) {
    /** @type {?} */
    NxBreadcrumbComponent.prototype._destroyed;
    /**
     * @type {?}
     * @private
     */
    NxBreadcrumbComponent.prototype._negative;
    /**
     * \@docs-private
     * @type {?}
     */
    NxBreadcrumbComponent.prototype.breadcrumbItems;
    /**
     * @type {?}
     * @private
     */
    NxBreadcrumbComponent.prototype._cdr;
}
