// @allianz/ngx-ndbx/esm5/tabs/tab-header-outlet.js
/**
 * @fileoverview added by tsickle
 * Generated from: tabs/tab-header-outlet.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EmbeddedViewRef, Input, ViewChild, ViewContainerRef } from '@angular/core';
import { NxTabGroupBase } from './tab-group-base';
/**
 * \@docs-private
 */
var NxTabHeaderOutletComponent = /** @class */ (function () {
    function NxTabHeaderOutletComponent(_tabGroup) {
        this._tabGroup = _tabGroup;
    }
    /**
     * @return {?}
     */
    NxTabHeaderOutletComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._appearanceSubscription = ((/** @type {?} */ (this._tabGroup)))._appearanceChange.subscribe((/**
         * @return {?}
         */
        function () {
            _this.detach();
        }));
        this.attach();
    };
    /**
     * @return {?}
     */
    NxTabHeaderOutletComponent.prototype.attach = /**
     * @return {?}
     */
    function () {
        this._outlet.insert(this.content);
    };
    /**
     * @return {?}
     */
    NxTabHeaderOutletComponent.prototype.detach = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var index = this._outlet.indexOf(this.content);
        if (index !== -1) {
            this._outlet.detach(index);
        }
    };
    /**
     * @return {?}
     */
    NxTabHeaderOutletComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._appearanceSubscription.unsubscribe();
    };
    NxTabHeaderOutletComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-tab-header-outlet',
                    template: '<ng-container #outlet></ng-container>'
                }] }
    ];
    /** @nocollapse */
    NxTabHeaderOutletComponent.ctorParameters = function () { return [
        { type: NxTabGroupBase }
    ]; };
    NxTabHeaderOutletComponent.propDecorators = {
        _outlet: [{ type: ViewChild, args: ['outlet', { static: true, read: ViewContainerRef },] }],
        content: [{ type: Input }]
    };
    return NxTabHeaderOutletComponent;
}());
export { NxTabHeaderOutletComponent };
if (false) {
    /** @type {?} */
    NxTabHeaderOutletComponent.prototype._outlet;
    /** @type {?} */
    NxTabHeaderOutletComponent.prototype.content;
    /**
     * @type {?}
     * @private
     */
    NxTabHeaderOutletComponent.prototype._appearanceSubscription;
    /**
     * @type {?}
     * @private
     */
    NxTabHeaderOutletComponent.prototype._tabGroup;
}
