// @allianz/ngx-ndbx/esm5/number-stepper/auto-resize.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: number-stepper/auto-resize.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { getFontShorthand } from '@allianz/ngx-ndbx/utils';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ChangeDetectorRef, Directive, ElementRef, HostBinding, Input, Renderer2, } from '@angular/core';
/**
 * \@docs-private
 */
var NxAutoResizeDirective = /** @class */ (function () {
    function NxAutoResizeDirective(_element, _renderer, _cdr) {
        this._element = _element;
        this._renderer = _renderer;
        this._cdr = _cdr;
        this._resize = true;
        this.updateInputWidth = this.updateInputWidth.bind(this);
    }
    Object.defineProperty(NxAutoResizeDirective.prototype, "resize", {
        get: /**
         * @return {?}
         */
        function () {
            return this._resize;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._resize = coerceBooleanProperty(value);
            if (this._resize) {
                this._addEventListener();
                this.updateInputWidth();
            }
            else {
                this._removeEventListener();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxAutoResizeDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.resize) {
            this._addEventListener();
        }
    };
    /**
     * @return {?}
     */
    NxAutoResizeDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._removeEventListener();
    };
    /**
     * @return {?}
     */
    NxAutoResizeDirective.prototype.updateInputWidth = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var measureCanvas = this._renderer.createElement('canvas');
        /** @type {?} */
        var ctx = measureCanvas.getContext('2d');
        /** @type {?} */
        var styles = window.getComputedStyle(this._element.nativeElement);
        ctx.font = getFontShorthand(styles);
        /** @type {?} */
        var metrics = ctx.measureText(this._element.nativeElement.value);
        /** @type {?} */
        var padding = this.sumStyles(styles.paddingLeft, styles.paddingRight);
        /** @type {?} */
        var border = this.sumStyles(styles.borderLeftWidth, styles.borderRightWidth);
        // the pixels are needed, because despite the correct calculation the last pixels of a number are always cut
        /** @type {?} */
        var newWidth = metrics.width + padding + border + 16;
        // Limit to own given minimal width
        /** @type {?} */
        var parsed = parseFloat(styles.minWidth);
        this.width = Math.max(Number.isNaN(parsed) ? 0 : parsed, newWidth);
        // needed when the outer component is onPush
        this._cdr.markForCheck();
    };
    /**
     * @return {?}
     */
    NxAutoResizeDirective.prototype._addEventListener = /**
     * @return {?}
     */
    function () {
        this._element.nativeElement.addEventListener('input', this.updateInputWidth, true);
        this._element.nativeElement.addEventListener('change', this.updateInputWidth, true);
    };
    /**
     * @return {?}
     */
    NxAutoResizeDirective.prototype._removeEventListener = /**
     * @return {?}
     */
    function () {
        this._element.nativeElement.removeEventListener('input', this.updateInputWidth, true);
        this._element.nativeElement.removeEventListener('change', this.updateInputWidth, true);
    };
    /**
     * @param {?} left
     * @param {?} right
     * @return {?}
     */
    NxAutoResizeDirective.prototype.sumStyles = /**
     * @param {?} left
     * @param {?} right
     * @return {?}
     */
    function (left, right) {
        left = parseInt(left, 10);
        right = parseInt(right, 10);
        left = Number.isNaN(left) ? 0 : left;
        right = Number.isNaN(right) ? 0 : right;
        return left + right;
    };
    NxAutoResizeDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'input[nxAutoResize]'
                },] }
    ];
    /** @nocollapse */
    NxAutoResizeDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ChangeDetectorRef }
    ]; };
    NxAutoResizeDirective.propDecorators = {
        width: [{ type: HostBinding, args: ['style.width.px',] }],
        resize: [{ type: Input, args: ['nxAutoResize',] }]
    };
    return NxAutoResizeDirective;
}());
export { NxAutoResizeDirective };
if (false) {
    /** @type {?} */
    NxAutoResizeDirective.prototype.width;
    /**
     * @type {?}
     * @private
     */
    NxAutoResizeDirective.prototype._resize;
    /**
     * @type {?}
     * @private
     */
    NxAutoResizeDirective.prototype._element;
    /**
     * @type {?}
     * @private
     */
    NxAutoResizeDirective.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    NxAutoResizeDirective.prototype._cdr;
}
