// @allianz/ngx-ndbx/esm5/footer/footer.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: footer.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Directive } from '@angular/core';
var NxFooterCopyrightDirective = /** @class */ (function () {
    function NxFooterCopyrightDirective() {
    }
    NxFooterCopyrightDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'nx-footer-copyright',
                    exportAs: 'NxFooterCopyright',
                    host: {
                        'class': 'nx-footer__copyright'
                    }
                },] }
    ];
    return NxFooterCopyrightDirective;
}());
export { NxFooterCopyrightDirective };
var NxFooterNavigationDirective = /** @class */ (function () {
    function NxFooterNavigationDirective() {
    }
    NxFooterNavigationDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'nx-footer-navigation',
                    exportAs: 'NxFooterNavigation',
                    host: {
                        'class': 'nx-footer__navigation',
                        'role': 'list'
                    }
                },] }
    ];
    return NxFooterNavigationDirective;
}());
export { NxFooterNavigationDirective };
var NxFooterLinkDirective = /** @class */ (function () {
    function NxFooterLinkDirective() {
    }
    NxFooterLinkDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'nx-footer-link',
                    exportAs: 'NxFooterLink',
                    host: {
                        'class': 'nx-footer__link',
                        'role': 'listitem'
                    }
                },] }
    ];
    return NxFooterLinkDirective;
}());
export { NxFooterLinkDirective };
var NxFooterComponent = /** @class */ (function () {
    function NxFooterComponent() {
    }
    NxFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-footer, [nx-footer]',
                    template: '<ng-content></ng-content>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        'class': 'nx-footer'
                    },
                    styles: [":host(.nx-footer){font-size:16px;font-size:var(--footer-font-size,16px);line-height:24px;line-height:var(--footer-line-height,24px);font-weight:400;font-weight:var(--footer-font-weight,400);letter-spacing:0;letter-spacing:var(--footer-letter-spacing,0);color:#767676;color:var(--footer-color,#767676);display:flex;flex-direction:row;background:0 0;padding:72px 32px 24px}::ng-deep a{text-decoration:none}::ng-deep .nx-footer__copyright{font-weight:700;font-weight:var(--footer-copyright-font-weight,700);align-self:flex-end;font-size:inherit;line-height:inherit;color:inherit}@media (max-width:703px){:host(.nx-footer){flex-direction:column-reverse;padding:72px 16px 40px}::ng-deep .nx-footer__copyright{-ms-grid-row-align:center;align-self:center;text-align:center}}::ng-deep .nx-footer__navigation{display:flex;align-self:flex-end}@media (max-width:703px){::ng-deep .nx-footer__navigation{text-align:center;-ms-grid-row-align:center;align-self:center;flex-direction:column}::ng-deep .nx-footer__navigation .nx-footer__link{padding-left:0;padding-bottom:16px}}::ng-deep .nx-footer__link{font-size:inherit;color:inherit;line-height:inherit;padding-left:32px}:host-context([data-whatinput=keyboard]) ::ng-deep a:focus{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6);border-radius:4px}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]) ::ng-deep a:focus{box-shadow:0 0 0 2px background,0 0 0 6px windowText}}"]
                }] }
    ];
    return NxFooterComponent;
}());
export { NxFooterComponent };
