// @allianz/ngx-ndbx/esm5/input/password-toggle.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: password-toggle.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Component, ChangeDetectorRef, ChangeDetectionStrategy } from '@angular/core';
import { ENTER, SPACE } from '@angular/cdk/keycodes';
/** @type {?} */
var visibilityIcons = {
    show: 'password-show-o',
    hide: 'password-hide-o'
};
var NxPasswordToggleComponent = /** @class */ (function () {
    function NxPasswordToggleComponent(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        /**
         * \@docs-private
         */
        this._currentIcon = visibilityIcons['show'];
        /**
         * \@docs-private
         */
        this._pressed = false;
        this._ariaLabel = 'Show password';
    }
    Object.defineProperty(NxPasswordToggleComponent.prototype, "ariaLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return this._ariaLabel;
        },
        /** Sets the aria-label needed for accessibility.
         * Notice that this `aria-label` should describe the initial action according to the status of the visibility.
         * E.g if you have an `input[type=password]` at the beginning then the
         * password will be hidden and the correct aria-label would be 'Show password.'
         */
        set: /**
         * Sets the aria-label needed for accessibility.
         * Notice that this `aria-label` should describe the initial action according to the status of the visibility.
         * E.g if you have an `input[type=password]` at the beginning then the
         * password will be hidden and the correct aria-label would be 'Show password.'
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this._ariaLabel) {
                this._ariaLabel = value;
                this._changeDetectorRef.markForCheck();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxPasswordToggleComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (!this.control) {
            console.warn('You need to pass an input as a control to the password toggle.');
        }
        else {
            // show the right icon according to the initial type of the input
            this._currentIcon = this.control.type === 'password' ? visibilityIcons['show'] : visibilityIcons['hide'];
        }
    };
    /** Toggles the type of the input. */
    /**
     * Toggles the type of the input.
     * @return {?}
     */
    NxPasswordToggleComponent.prototype.toggleInputType = /**
     * Toggles the type of the input.
     * @return {?}
     */
    function () {
        if (this.control) {
            this.control.type = this.control.type === 'password' ? 'text' : 'password';
            this._pressed = !this._pressed;
            this.toggleIcon();
            this._changeDetectorRef.markForCheck();
        }
    };
    /**@docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxPasswordToggleComponent.prototype.toggleIcon = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this._currentIcon = this._currentIcon === visibilityIcons['show'] ? visibilityIcons['hide'] : visibilityIcons['show'];
    };
    /**@docs-private */
    /**
     * \@docs-private
     * @param {?} $event
     * @return {?}
     */
    NxPasswordToggleComponent.prototype._onKeydown = /**
     * \@docs-private
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if ($event && ($event.keyCode === ENTER || $event.keyCode === SPACE)) {
            this.toggleInputType();
        }
    };
    Object.defineProperty(NxPasswordToggleComponent.prototype, "tabindex", {
        /**@docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            if (this.control) {
                return this.control.disabled ? -1 : 0;
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    NxPasswordToggleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-password-toggle',
                    template: "<nx-icon\n      aria-hidden=\"true\"\n      [name]=\"_currentIcon\">\n    </nx-icon>",
                    host: {
                        '[attr.aria-label]': 'ariaLabel',
                        '[attr.tabindex]': 'tabindex',
                        '[attr.aria-pressed]': '_pressed',
                        'role': 'button',
                        '(click)': 'toggleInputType()',
                        '(keydown)': '_onKeydown($event)'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host{position:relative;display:inline-block;height:24px;width:24px;cursor:pointer}:host:focus{outline:0}:host:hover{color:#006192;color:var(--hover-primary,#006192)}:host-context([data-whatinput=keyboard]):focus ::ng-deep nx-icon{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6);border-radius:4px}@media screen and (-ms-high-contrast:active){:host{color:buttonText}:host-context([data-whatinput=keyboard]):focus ::ng-deep nx-icon{box-shadow:0 0 0 2px background,0 0 0 6px windowText}}:host-context(.is-disabled){pointer-events:none}"]
                }] }
    ];
    /** @nocollapse */
    NxPasswordToggleComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxPasswordToggleComponent.propDecorators = {
        control: [{ type: Input }],
        ariaLabel: [{ type: Input }]
    };
    return NxPasswordToggleComponent;
}());
export { NxPasswordToggleComponent };
if (false) {
    /**
     * Input element using the toggle functionality.
     * @type {?}
     */
    NxPasswordToggleComponent.prototype.control;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPasswordToggleComponent.prototype._currentIcon;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPasswordToggleComponent.prototype._pressed;
    /**
     * @type {?}
     * @private
     */
    NxPasswordToggleComponent.prototype._ariaLabel;
    /**
     * @type {?}
     * @private
     */
    NxPasswordToggleComponent.prototype._changeDetectorRef;
}
