// @allianz/ngx-ndbx/esm5/pagination/pagination.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: pagination/pagination.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Inject, Input, Optional, Output, } from '@angular/core';
import { DefaultPaginationTexts, NX_PAGINATION_TEXTS } from './pagination-texts';
import { NxPaginationUtils } from './pagination-utils';
/**
 * \@docs-private
 * @record
 */
export function Page() { }
if (false) {
    /** @type {?} */
    Page.prototype.label;
    /** @type {?} */
    Page.prototype.value;
    /** @type {?} */
    Page.prototype.class;
}
var NxPaginationComponent = /** @class */ (function () {
    function NxPaginationComponent(paginationTexts, paginationUtilsService, _changeDetectorRef) {
        this.paginationUtilsService = paginationUtilsService;
        this._changeDetectorRef = _changeDetectorRef;
        this._type = 'simple';
        /**
         * \@docs-private
         */
        this.totalNumberPages = 0;
        /**
         * An event emitted when the previous page button is clicked.
         */
        this.nxGoPrev = new EventEmitter();
        /**
         * An event emitted when the next page button is clicked
         */
        this.nxGoNext = new EventEmitter();
        /**
         * An event emitted when a page number is clicked.
         * Provides the number of the page as parameter.
         */
        this.nxGoPage = new EventEmitter();
        this.paginationTexts = paginationTexts || DefaultPaginationTexts;
    }
    Object.defineProperty(NxPaginationComponent.prototype, "page", {
        get: /**
         * @return {?}
         */
        function () { return this._page; },
        /** Sets the current page. */
        set: /**
         * Sets the current page.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._page = value;
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxPaginationComponent.prototype, "count", {
        get: /**
         * @return {?}
         */
        function () { return this._count; },
        /** Number of total items over all pages. */
        set: /**
         * Number of total items over all pages.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._count = value;
            this.totalNumberPages = this.calculateTotalPages();
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxPaginationComponent.prototype, "perPage", {
        get: /**
         * @return {?}
         */
        function () { return this._perPage; },
        /** Sets the number of items you want to show per page. */
        set: /**
         * Sets the number of items you want to show per page.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._perPage = value;
            this.totalNumberPages = this.calculateTotalPages();
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxPaginationComponent.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () { return this._type; },
        /**
         * Determines the type of pagination.
         *
         * Values: simple | advanced, default: simple.
         */
        set: /**
         * Determines the type of pagination.
         *
         * Values: simple | advanced, default: simple.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._type = value;
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxPaginationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.totalNumberPages = this.calculateTotalPages();
    };
    /**
     * @return {?}
     */
    NxPaginationComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        if (this.type === 'advanced' && (!this.paginationTexts.last || !this.paginationTexts.first)) {
            console.warn('Please define aria labels for the last and first arrows.');
        }
    };
    /** Returns the number of the first page. */
    /**
     * Returns the number of the first page.
     * @return {?}
     */
    NxPaginationComponent.prototype.getMin = /**
     * Returns the number of the first page.
     * @return {?}
     */
    function () {
        return this.totalNumberPages > 0 ? 1 : 0;
    };
    /** Returns the number of the last page. */
    /**
     * Returns the number of the last page.
     * @return {?}
     */
    NxPaginationComponent.prototype.getMax = /**
     * Returns the number of the last page.
     * @return {?}
     */
    function () {
        /** @type {?} */
        var max = this._perPage * this._page;
        if (max > this._count) {
            max = this._count;
        }
        return max;
    };
    /** Returns the total number of pages */
    /**
     * Returns the total number of pages
     * @return {?}
     */
    NxPaginationComponent.prototype.calculateTotalPages = /**
     * Returns the total number of pages
     * @return {?}
     */
    function () {
        return Math.ceil(this._count / this._perPage) || 0;
    };
    /** Directs to the page with number n. */
    /**
     * Directs to the page with number n.
     * @param {?} n
     * @return {?}
     */
    NxPaginationComponent.prototype.onPage = /**
     * Directs to the page with number n.
     * @param {?} n
     * @return {?}
     */
    function (n) {
        this.nxGoPage.emit(n);
    };
    /** Directs to the previous page. */
    /**
     * Directs to the previous page.
     * @return {?}
     */
    NxPaginationComponent.prototype.onPrev = /**
     * Directs to the previous page.
     * @return {?}
     */
    function () {
        if (!this._isPaginationPreviousDisabled()) {
            this.nxGoPrev.emit();
        }
    };
    /** Directs to the next page. */
    /**
     * Directs to the next page.
     * @return {?}
     */
    NxPaginationComponent.prototype.onNext = /**
     * Directs to the next page.
     * @return {?}
     */
    function () {
        if (!this._isPaginationNextDisabled()) {
            this.nxGoNext.emit();
        }
    };
    /** Directs to the first page. */
    /**
     * Directs to the first page.
     * @return {?}
     */
    NxPaginationComponent.prototype.onFirst = /**
     * Directs to the first page.
     * @return {?}
     */
    function () {
        if (!this._isPaginationPreviousDisabled()) {
            this.onPage(1);
        }
    };
    /** Directs to the last page. */
    /**
     * Directs to the last page.
     * @return {?}
     */
    NxPaginationComponent.prototype.onLast = /**
     * Directs to the last page.
     * @return {?}
     */
    function () {
        if (!this._isPaginationNextDisabled()) {
            this.onPage(this.totalNumberPages);
        }
    };
    /** Returns if the current page is the last page. */
    /**
     * Returns if the current page is the last page.
     * @return {?}
     */
    NxPaginationComponent.prototype.lastPage = /**
     * Returns if the current page is the last page.
     * @return {?}
     */
    function () {
        return this._perPage * this._page >= this._count;
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxPaginationComponent.prototype.getPages = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        return this.paginationUtilsService.getPages(this._page, this.totalNumberPages);
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxPaginationComponent.prototype.getMobilePages = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        return this.paginationUtilsService.getMobilePages(this._page, this.totalNumberPages);
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @param {?} page
     * @return {?}
     */
    NxPaginationComponent.prototype.getPaginationItemClasses = /**
     * \@docs-private
     * @param {?} page
     * @return {?}
     */
    function (page) {
        /** @type {?} */
        var classes = {
            'is-ellipsis': page.label === '...',
            'nx-pagination__item--expanded-view': page.class === 'expanded-view'
        };
        return classes;
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @param {?} page
     * @return {?}
     */
    NxPaginationComponent.prototype.getPaginationNumberClasses = /**
     * \@docs-private
     * @param {?} page
     * @return {?}
     */
    function (page) {
        /** @type {?} */
        var classes = {
            'is-active': page.value === this.page,
            'nx-pagination__ellipsis': page.label === '...',
            'nx-pagination__link': page.label !== '...'
        };
        return classes;
    };
    /** Returns true, if `nxCount` is greater than 0, else false. */
    /**
     * Returns true, if `nxCount` is greater than 0, else false.
     * @return {?}
     */
    NxPaginationComponent.prototype.isPaginationVisible = /**
     * Returns true, if `nxCount` is greater than 0, else false.
     * @return {?}
     */
    function () {
        return this.count > 0;
    };
    /** Returns true, if `nxCount` is greater than 0 and the type of pagination is 'simple', else false. */
    /**
     * Returns true, if `nxCount` is greater than 0 and the type of pagination is 'simple', else false.
     * @return {?}
     */
    NxPaginationComponent.prototype.isPaginationCompactVisible = /**
     * Returns true, if `nxCount` is greater than 0 and the type of pagination is 'simple', else false.
     * @return {?}
     */
    function () {
        return this.type.includes('simple') && this.count > 0;
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxPaginationComponent.prototype.isPaginationContainerVisible = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        return this.type.includes('advanced');
    };
    /**
     * @return {?}
     */
    NxPaginationComponent.prototype._isPaginationPreviousDisabled = /**
     * @return {?}
     */
    function () {
        return this.page === this.getMin();
    };
    /**
     * @return {?}
     */
    NxPaginationComponent.prototype._isPaginationNextDisabled = /**
     * @return {?}
     */
    function () {
        return this.page === this.totalNumberPages;
    };
    NxPaginationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-pagination',
                    template: "<nav *ngIf=\"isPaginationVisible()\"  aria-label=\"Page navigation\" class=\"nx-pagination\" >\n  <ul *ngIf=\"isPaginationContainerVisible()\"  [attr.aria-label]=\"paginationTexts.ariaLabel\" class=\"nx-pagination__container\">\n    <li class=\"nx-pagination__item\">\n      <button\n        [attr.aria-label]=\"paginationTexts.first\"\n        [class.is-disabled]=\"_isPaginationPreviousDisabled()\"\n        [tabindex]=\"_isPaginationPreviousDisabled() ? -1 : 0\"\n        class=\"nx-pagination__link nx-pagination__link--first\"\n        (click)=\"onFirst()\">\n        <nx-icon class=\"nx-pagination__arrow\" name=\"arrow-first\"></nx-icon>\n      </button>\n    </li>\n    <li class=\"nx-pagination__item nx-pagination__item-previous\">\n      <button\n        [attr.aria-label]=\"paginationTexts.previous\"\n        [class.is-disabled]=\"_isPaginationPreviousDisabled()\"\n        [tabindex]=\"_isPaginationPreviousDisabled() ? -1 : 0\"\n        class=\"nx-pagination__link nx-pagination__link--previous\"\n        (click)=\"onPrev()\">\n        <nx-icon class=\"nx-pagination__arrow\" name=\"arrow-left\"></nx-icon>\n      </button>\n    </li>\n    <li *ngFor=\"let pageNum of getPages()\" class=\"nx-pagination__item nx-pagination__item--desktop\" [ngClass]=\"getPaginationItemClasses(pageNum)\">\n      <button class=\"nx-pagination--number\"\n        *ngIf=\"pageNum.label !== '...'\"\n        (click)=\"onPage(pageNum.value)\"\n        [attr.aria-current]=\"(pageNum.value === page) ? 'page' : ''\"\n        [ngClass]=\"getPaginationNumberClasses(pageNum)\">\n        {{ pageNum.label }}\n      </button>\n      <span *ngIf=\"pageNum.label === '...'\">\n        {{ pageNum.label }}\n      </span>\n    </li>\n    <li *ngFor=\"let pageNum of getMobilePages()\" class=\"nx-pagination__item nx-pagination__item--mobile\" [ngClass]=\"getPaginationItemClasses(pageNum)\">\n      <button class=\"nx-pagination--number\"\n        (click)=\"onPage(pageNum.value)\"\n        [attr.aria-current]=\"(pageNum.value === page) ? 'page' : ''\"\n        [ngClass]=\"getPaginationNumberClasses(pageNum)\">\n        {{ pageNum.label }}\n      </button>\n    </li>\n    <li class=\"nx-pagination__item nx-pagination__item-next\">\n      <button\n        [attr.aria-label]=\"paginationTexts.next\"\n        [class.is-disabled]=\"_isPaginationNextDisabled()\"\n        [tabindex]=\"_isPaginationNextDisabled() ? -1 : 0\"\n        class=\"nx-pagination__link nx-pagination__link--next\"\n        (click)=\"onNext()\">\n        <nx-icon class=\"nx-pagination__arrow\" name=\"arrow-right\"></nx-icon>\n      </button>\n    </li>\n    <li class=\"nx-pagination__item\">\n      <button\n        [attr.aria-label]=\"paginationTexts.last\"\n        [class.is-disabled]=\"_isPaginationNextDisabled()\"\n        [tabindex]=\"_isPaginationNextDisabled() ? -1 : 0\"\n        class=\"nx-pagination__link nx-pagination__link--last\"\n        (click)=\"onLast()\">\n        <nx-icon class=\"nx-pagination__arrow\" name=\"arrow-last\"></nx-icon>\n      </button>\n    </li>\n  </ul>\n</nav>\n<nav *ngIf=\"isPaginationCompactVisible()\" [attr.aria-label]=\"paginationTexts.ariaLabel\" class=\"nx-pagination-compact\" >\n  <button [attr.aria-label]=\"paginationTexts.previous\"\n    class=\"nx-pagination-compact__previous\"\n    [class.is-disabled]=\"_isPaginationPreviousDisabled()\"\n    [tabindex]=\"_isPaginationPreviousDisabled() ? -1 : 0\"\n\n    (click)=\"onPrev()\">\n    <nx-icon class=\"nx-pagination__arrow\" name=\"arrow-left\"></nx-icon>\n    <span class=\"nx-pagination-compact__direction-label\">\n      {{ paginationTexts.previous }}\n    </span>\n  </button>\n  <div class=\"nx-pagination-compact__display\">\n      <span class=\"nx-pagination-compact__current-page\">{{ page }}</span>\n      <span class=\"nx-pagination-compact__page-separator\"> {{ paginationTexts.ofLabel }} </span>\n      <span class=\"nx-pagination-compact__total-pages\">{{ totalNumberPages }}</span>\n  </div>\n  <button\n    class=\"nx-pagination-compact__next\"\n    [class.is-disabled]=\"_isPaginationNextDisabled()\"\n    [tabindex]=\"_isPaginationNextDisabled() ? -1 : 0\"\n    [attr.aria-label]=\"paginationTexts.next\"\n    (click)=\"onNext()\">\n    <span class=\"nx-pagination-compact__direction-label\"> {{ paginationTexts.next }} </span>\n    <nx-icon class=\"nx-pagination__arrow\" name=\"arrow-right\"></nx-icon>\n  </button>\n</nav>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host button:focus{outline:0}:host button::-moz-focus-inner{border:0}.nx-pagination__container{display:flex;font-size:16px;font-size:var(--pagination-font-size,16px);line-height:24px;line-height:var(--pagination-line-height,24px);font-weight:400;font-weight:var(--pagination-font-weight,400);letter-spacing:0;letter-spacing:var(--pagination-letter-spacing,0);margin:8px 0;padding:0}@media (min-width:704px){.nx-pagination__container{margin:8px 0 40px}}.nx-pagination__item{margin:0 16px;list-style:none;font-weight:600;font-weight:var(--pagination-item-font-weight,600)}.nx-pagination__item:first-child,.nx-pagination__item:last-child{margin:0}.nx-pagination__item.is-ellipsis{color:#414141;color:var(--pagination-text-color,#414141)}.nx-pagination__item--expanded-view{display:block}.nx-pagination__link{color:#006192;color:var(--pagination-link-color,#006192);padding:0;text-decoration:none;background-color:transparent;border:0;cursor:pointer}.nx-pagination__link.is-active{color:#414141;color:var(--pagination-text-color,#414141);font-weight:700;cursor:default}nx-icon.nx-pagination__arrow{font-size:24px;font-size:var(--pagination-icon-size,24px)}.nx-pagination-compact{display:flex;font-size:16px;font-size:var(--pagination-font-size,16px);line-height:24px;line-height:var(--pagination-line-height,24px);font-weight:400;font-weight:var(--pagination-font-weight,400);letter-spacing:0;letter-spacing:var(--pagination-letter-spacing,0);margin:16px 0;padding:0}@media (min-width:704px){.nx-pagination-compact{margin:40px 0}}.nx-pagination-compact__display{color:#414141;color:var(--pagination-text-color,#414141);display:flex}.nx-pagination-compact__current-page{font-weight:700;margin-right:8px}.nx-pagination-compact__total-pages{margin-left:8px}.nx-pagination-compact__next,.nx-pagination-compact__previous,.nx-pagination__arrow-first,.nx-pagination__arrow-last{color:#006192;color:var(--pagination-link-color,#006192);font-weight:600;font-weight:var(--pagination-icon-font-weight,600);text-decoration:none;text-transform:uppercase;display:flex;align-items:center;padding:0;background-color:transparent;cursor:pointer;border:0}.nx-pagination-compact__previous{margin-right:32px}.nx-pagination-compact__next{margin-left:32px}.nx-pagination-compact__direction-label{font-size:16px;font-size:var(--pagination-compact-direction-font-size,16px);line-height:24px;line-height:var(--pagination-compact-direction-line-height,24px);font-weight:600;font-weight:var(--pagination-compact-direction-font-weight,600);letter-spacing:1px;letter-spacing:var(--pagination-compact-direction-letter-spacing,1px)}.nx-pagination-compact__direction-label:first-child{margin-right:8px}.nx-pagination-compact__direction-label:last-child{margin-left:8px}.nx-pagination-compact__next.is-disabled,.nx-pagination-compact__previous.is-disabled,.nx-pagination__link.is-disabled{color:rgba(0,97,146,.4);color:var(--pagination-link-disabled-color,rgba(0,97,146,.4));cursor:not-allowed}:host-context([data-whatinput=keyboard]) .nx-pagination--number:focus,:host-context([data-whatinput=keyboard]) .nx-pagination-compact__next:focus,:host-context([data-whatinput=keyboard]) .nx-pagination-compact__previous:focus,:host-context([data-whatinput=keyboard]) .nx-pagination__link:focus{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6);border-radius:4px}@media screen and (-ms-high-contrast:active){.nx-pagination__link.is-active{border-bottom:4px solid windowText}.nx-pagination-compact__next.is-disabled,.nx-pagination-compact__previous.is-disabled,.nx-pagination__link.is-disabled{color:GrayText}:host-context([data-whatinput=keyboard]) .nx-pagination--number:focus,:host-context([data-whatinput=keyboard]) .nx-pagination-compact__next:focus,:host-context([data-whatinput=keyboard]) .nx-pagination-compact__previous:focus,:host-context([data-whatinput=keyboard]) .nx-pagination__link:focus{box-shadow:0 0 0 2px background,0 0 0 6px windowText}}.nx-pagination__link--first,.nx-pagination__link--last,.nx-pagination__link--next,.nx-pagination__link--previous{display:flex;align-items:center}.nx-pagination__item-next{margin-right:8px}.nx-pagination__item-previous{margin-left:8px}.nx-pagination-compact__direction-label,.nx-pagination__item--desktop,.nx-pagination__link--first,.nx-pagination__link--last{display:block}.nx-pagination__item--mobile{display:none}@media (max-width:703px){.nx-pagination-compact__direction-label,.nx-pagination__item--desktop,.nx-pagination__link--first,.nx-pagination__link--last{display:none}.nx-pagination__item--mobile{display:block}}"]
                }] }
    ];
    /** @nocollapse */
    NxPaginationComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [NX_PAGINATION_TEXTS,] }] },
        { type: NxPaginationUtils },
        { type: ChangeDetectorRef }
    ]; };
    NxPaginationComponent.propDecorators = {
        page: [{ type: Input, args: ['nxPage',] }],
        count: [{ type: Input, args: ['nxCount',] }],
        perPage: [{ type: Input, args: ['nxPerPage',] }],
        type: [{ type: Input, args: ['nxType',] }],
        nxGoPrev: [{ type: Output }],
        nxGoNext: [{ type: Output }],
        nxGoPage: [{ type: Output }]
    };
    return NxPaginationComponent;
}());
export { NxPaginationComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxPaginationComponent.prototype._page;
    /**
     * @type {?}
     * @private
     */
    NxPaginationComponent.prototype._count;
    /**
     * @type {?}
     * @private
     */
    NxPaginationComponent.prototype._perPage;
    /**
     * @type {?}
     * @private
     */
    NxPaginationComponent.prototype._type;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPaginationComponent.prototype.paginationTexts;
    /**
     * \@docs-private
     * @type {?}
     */
    NxPaginationComponent.prototype.totalNumberPages;
    /**
     * An event emitted when the previous page button is clicked.
     * @type {?}
     */
    NxPaginationComponent.prototype.nxGoPrev;
    /**
     * An event emitted when the next page button is clicked
     * @type {?}
     */
    NxPaginationComponent.prototype.nxGoNext;
    /**
     * An event emitted when a page number is clicked.
     * Provides the number of the page as parameter.
     * @type {?}
     */
    NxPaginationComponent.prototype.nxGoPage;
    /**
     * @type {?}
     * @private
     */
    NxPaginationComponent.prototype.paginationUtilsService;
    /**
     * @type {?}
     * @private
     */
    NxPaginationComponent.prototype._changeDetectorRef;
}
