// @allianz/ngx-ndbx/esm5/tree/node.js
/**
 * @fileoverview added by tsickle
 * Generated from: node.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CdkTreeNode, CdkTreeNodeDef } from '@angular/cdk/tree';
import { Component, Directive, Input, } from '@angular/core';
/**
 * Wrapper for the CdkTree node with NDBX design styles.
 * @template T
 */
var NxTreeNodeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NxTreeNodeComponent, _super);
    function NxTreeNodeComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NxTreeNodeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-tree-node',
                    exportAs: 'nxTreeNode',
                    inputs: ['disabled', 'tabIndex'],
                    host: {
                        '[attr.aria-expanded]': 'isExpanded',
                        '[attr.aria-level]': 'role === "treeitem" ? level : null',
                        '[attr.role]': 'role',
                        'class': 'nx-tree__node',
                        '[class.is-expanded]': 'isExpanded'
                    },
                    providers: [{ provide: CdkTreeNode, useExisting: NxTreeNodeComponent }],
                    template: "<div class=\"nx-tree__node-box\">\n  <ng-content></ng-content>\n</div>\n"
                }] }
    ];
    return NxTreeNodeComponent;
}(CdkTreeNode));
export { NxTreeNodeComponent };
/**
 * Wrapper for the CdkTree node definition with NDBX design styles.
 * @template T
 */
var NxTreeNodeDefDirective = /** @class */ (function (_super) {
    tslib_1.__extends(NxTreeNodeDefDirective, _super);
    function NxTreeNodeDefDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NxTreeNodeDefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[nxTreeNodeDef]',
                    inputs: [
                        'when: nxTreeNodeDefWhen'
                    ],
                    providers: [{ provide: CdkTreeNodeDef, useExisting: NxTreeNodeDefDirective }]
                },] }
    ];
    NxTreeNodeDefDirective.propDecorators = {
        data: [{ type: Input, args: ['nxTreeNode',] }]
    };
    return NxTreeNodeDefDirective;
}(CdkTreeNodeDef));
export { NxTreeNodeDefDirective };
if (false) {
    /** @type {?} */
    NxTreeNodeDefDirective.prototype.data;
}
