// @allianz/ngx-ndbx/esm5/tree/tree.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: tree.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CdkTree } from '@angular/cdk/tree';
import { ChangeDetectionStrategy, Component, ViewChild, ViewEncapsulation } from '@angular/core';
import { NxTreeNodeOutletDirective } from './outlet';
/**
 * Wrapper for the CdkTable with NDBX design styles.
 * @template T
 */
var NxTreeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NxTreeComponent, _super);
    function NxTreeComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NxTreeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-tree',
                    exportAs: 'nxTree',
                    template: "<ng-container nxTreeNodeOutlet></ng-container>",
                    host: {
                        'class': 'nx-tree',
                        'role': 'tree',
                    },
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [{ provide: CdkTree, useExisting: NxTreeComponent }],
                    styles: [":host.nx-tree{display:block}.nx-tree__node{display:block;overflow:hidden}"]
                }] }
    ];
    NxTreeComponent.propDecorators = {
        _nodeOutlet: [{ type: ViewChild, args: [NxTreeNodeOutletDirective, { static: true },] }]
    };
    return NxTreeComponent;
}(CdkTree));
export { NxTreeComponent };
if (false) {
    /** @type {?} */
    NxTreeComponent.prototype._nodeOutlet;
}
