// @allianz/ngx-ndbx/esm5/progress-stepper/single/single-step.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: single/single-step.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { NxProgressStepperDirective } from './../progress-stepper.component';
var NxSingleStepperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NxSingleStepperComponent, _super);
    function NxSingleStepperComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._rightLabel = 'Next step:';
        return _this;
    }
    Object.defineProperty(NxSingleStepperComponent.prototype, "rightLabel", {
        /** Sets the label on the right showing the next step. */
        get: /**
         * Sets the label on the right showing the next step.
         * @return {?}
         */
        function () { return this._rightLabel; },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) { this._rightLabel = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxSingleStepperComponent.prototype, "progress", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return (this.selectedIndex + 1) / this.count;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxSingleStepperComponent.prototype, "currentLabel", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            /** @type {?} */
            var step = this.currentStep;
            /** @type {?} */
            var label = step.stepLabel || step.label;
            return "\n    " + this.currentStepLabel + " " + (this.selectedIndex + 1) + "/" + this.count + ": " + label + "\n    ";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxSingleStepperComponent.prototype, "nextLabel", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            /** @type {?} */
            var step = this.nextStep;
            if (step) {
                /** @type {?} */
                var label = step.stepLabel || step.label;
                return "\n        " + this.rightLabel + " " + label + "\n      ";
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    NxSingleStepperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-single-stepper',
                    template: "<div class=\"nx-single-step-container\">\n  <span *ngIf=\"title\" class=\"nx-single-step__title\">{{title}}</span>\n  <span class=\"nx-progress-stepper__title\">\n    <ng-content select=\"nx-label\"></ng-content>\n  </span>\n\n  <nx-progressbar [value]=\"progress\"></nx-progressbar>\n\n  <div class=\"nx-single-step__infos\">\n    <div class=\"nx-single-step__current\">\n      {{ currentLabel }}\n    </div>\n    <div class=\"nx-single-step__next\">\n      {{ nextLabel }}\n    </div>\n  </div>\n\n  <div class=\"nx-progress-stepper__mobile-label\">\n    {{ currentLabel }}\n  </div>\n</div>\n\n<div\n  *ngFor=\"let step of _steps; let i = index\"\n  class=\"nx-progress-stepper__content\"\n  role=\"tabpanel\"\n  [id]=\"_getStepContentId(i)\"\n  [attr.aria-label]=\"currentLabel\"\n  [attr.aria-expanded]=\"selectedIndex === i\"\n>\n  <ng-container [ngTemplateOutlet]=\"step.content\"></ng-container>\n</div>\n",
                    providers: [{ provide: NxProgressStepperDirective, useExisting: NxSingleStepperComponent }],
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host{display:block}.nx-progress-stepper__content[aria-expanded=false]{height:0;overflow:hidden}.nx-progress-stepper__mobile-label{display:none;color:#414141;color:var(--progress-indicator-mobile-color,#414141);font-size:16px;font-size:var(--progress-indicator-mobile-font-size,16px);line-height:24px;line-height:var(--progress-indicator-mobile-line-height,24px);font-weight:400;font-weight:var(--progress-indicator-mobile-font-weight,400);letter-spacing:0;letter-spacing:var(--progress-indicator-mobile-letter-spacing,0)}@media (max-width:703px){.nx-progress-stepper__mobile-label{display:block;text-align:center}}.nx-progress-stepper__title{display:block}.nx-progress-stepper__title ::ng-deep nx-label{margin-bottom:24px}", ":host{display:block}.nx-single-step__infos{display:flex;font-size:16px;font-size:var(--single-stepper-step-font-size,16px);line-height:24px;line-height:var(--single-stepper-step-line-height,24px);font-weight:300;font-weight:var(--single-stepper-step-font-weight,300);letter-spacing:0;letter-spacing:var(--single-stepper-step-letter-spacing,0);color:#414141;color:var(--single-stepper-step-color,#414141);margin-top:8px}.nx-single-step__title,nx-progressbar{margin-bottom:8px;display:block}@media (max-width:703px){.nx-single-step__infos,nx-progressbar{display:none}}.nx-single-step__title{font-size:20px;font-size:var(--progress-stepper-label-font-size,20px);line-height:28px;line-height:var(--progress-stepper-label-line-height,28px);font-weight:300;font-weight:var(--progress-stepper-label-font-weight,300);letter-spacing:.2px;letter-spacing:var(--progress-stepper-label-letter-spacing,.2px);margin-bottom:24px}@media (min-width:320px){.nx-single-step__title{display:block}}.nx-single-step__next{margin-left:auto}@media (max-width:703px){.nx-single-step__next{display:none}}"]
                }] }
    ];
    NxSingleStepperComponent.propDecorators = {
        rightLabel: [{ type: Input, args: ['nextLabel',] }]
    };
    return NxSingleStepperComponent;
}(NxProgressStepperDirective));
export { NxSingleStepperComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxSingleStepperComponent.prototype._rightLabel;
}
