// @allianz/ngx-ndbx/esm2015/breadcrumb/breadcrumb.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: breadcrumb.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, QueryList, Input, ChangeDetectorRef, ChangeDetectionStrategy } from '@angular/core';
import { NxBreadcrumbItemComponent } from './breadcrumb-item.component';
import { startWith, takeUntil, filter } from 'rxjs/operators';
import { Subject } from 'rxjs';
export class NxBreadcrumbComponent {
    /**
     * @param {?} _cdr
     */
    constructor(_cdr) {
        this._cdr = _cdr;
        this._destroyed = new Subject();
        this._negative = false;
    }
    /**
     * Whether the component uses the negative styling.
     * @param {?} value
     * @return {?}
     */
    set negative(value) {
        this._negative = value;
        this._cdr.markForCheck();
    }
    /**
     * @return {?}
     */
    get negative() {
        return this._negative;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        if (this.breadcrumbItems.length === 0) {
            console.warn('A breadcrumb needs NxBreadcrumbItemComponent children wrapped in <li>!');
        }
        this.breadcrumbItems.changes
            .pipe(startWith(this.breadcrumbItems), filter((/**
         * @param {?} items
         * @return {?}
         */
        items => items.length !== 0)), takeUntil(this._destroyed))
            .subscribe((/**
         * @param {?} items
         * @return {?}
         */
        items => {
            this.breadcrumbItems.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => item.resetAriaLabel()));
            this.breadcrumbItems.last.setAsLast();
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
}
NxBreadcrumbComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'ol[nxBreadcrumb]',
                template: "<ng-content></ng-content>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[class.is-negative]': 'negative'
                },
                styles: [":host{display:flex;list-style:none;height:24px;padding:0}::ng-deep li:last-child ::ng-deep .nx-breadcrumb-item{font-weight:600;cursor:auto}::ng-deep li:last-child ::ng-deep .nx-breadcrumb-item__chevron{display:none}::ng-deep li+li ::ng-deep .nx-breadcrumb-item{margin-left:4px}:host(.is-negative) ::ng-deep .nx-breadcrumb-item{color:#fff}"]
            }] }
];
/** @nocollapse */
NxBreadcrumbComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
NxBreadcrumbComponent.propDecorators = {
    negative: [{ type: Input }],
    breadcrumbItems: [{ type: ContentChildren, args: [NxBreadcrumbItemComponent, { descendants: true },] }]
};
if (false) {
    /** @type {?} */
    NxBreadcrumbComponent.prototype._destroyed;
    /**
     * @type {?}
     * @private
     */
    NxBreadcrumbComponent.prototype._negative;
    /**
     * \@docs-private
     * @type {?}
     */
    NxBreadcrumbComponent.prototype.breadcrumbItems;
    /**
     * @type {?}
     * @private
     */
    NxBreadcrumbComponent.prototype._cdr;
}
