// @allianz/ngx-ndbx/esm5/config/nx-expert.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: nx-expert.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FORMFIELD_DEFAULT_OPTIONS } from '@allianz/ngx-ndbx/formfield';
import { LABEL_DEFAULT_OPTIONS, ERROR_DEFAULT_OPTIONS } from '@allianz/ngx-ndbx/base';
import { DATEPICKER_DEFAULT_OPTIONS } from '@allianz/ngx-ndbx/datefield';
// expert presets
/** @type {?} */
var formfieldExpertOptions = {
    appearance: 'outline',
    nxFloatLabel: 'always'
};
/** @type {?} */
var errorExpertOptions = { appearance: 'text' };
/** @type {?} */
var labelExpertOptions = { size: 'small' };
/** @type {?} */
var datepickerExpertOptions = { toggleIconTabindex: -1 };
var ɵ0 = formfieldExpertOptions, ɵ1 = labelExpertOptions, ɵ2 = datepickerExpertOptions, ɵ3 = errorExpertOptions;
var NxExpertModule = /** @class */ (function () {
    function NxExpertModule() {
    }
    NxExpertModule.decorators = [
        { type: NgModule, args: [{
                    providers: [
                        { provide: FORMFIELD_DEFAULT_OPTIONS, useValue: ɵ0 },
                        { provide: LABEL_DEFAULT_OPTIONS, useValue: ɵ1 },
                        { provide: DATEPICKER_DEFAULT_OPTIONS, useValue: ɵ2 },
                        { provide: ERROR_DEFAULT_OPTIONS, useValue: ɵ3 }
                    ]
                },] }
    ];
    return NxExpertModule;
}());
export { NxExpertModule };
export { ɵ0, ɵ1, ɵ2, ɵ3 };
