// @allianz/ngx-ndbx/fesm2015/allianz-ngx-ndbx-core.js
import { removeClasses, mapClassNames, appendClasses } from '@allianz/ngx-ndbx/utils';

/**
 * @fileoverview added by tsickle
 * Generated from: breakpoints.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const BREAKPOINT_MOBILE = '(max-width: 703px)';
/** @type {?} */
const BREAKPOINT_TABLET = '(min-width: 704px) and (max-width: 992px)';
/** @type {?} */
const BREAKPOINT_DESKTOP = '(min-width: 993px)';

/**
 * @fileoverview added by tsickle
 * Generated from: mapped-styles.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class MappedStyles {
    /**
     * @param {?} mapping
     * @param {?=} baseClasses
     * @param {?=} _elementRef
     * @param {?=} _renderer
     */
    constructor(mapping, baseClasses = [], _elementRef, _renderer) {
        this.mapping = mapping;
        this.baseClasses = baseClasses;
        this._elementRef = _elementRef;
        this._renderer = _renderer;
        this.classNames = '';
    }
    /**
     * @return {?}
     */
    get classNames() {
        return this._classNames;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set classNames(value) {
        if (this._classNames === value) {
            return;
        }
        removeClasses(this._renderer, this._elementRef, this._classNamesSanitized);
        this._classNamesSanitized = mapClassNames(value, this.baseClasses, this.mapping);
        this._classNames = value;
        appendClasses(this._renderer, this._elementRef, this._classNamesSanitized);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    MappedStyles.prototype._classNames;
    /**
     * @type {?}
     * @private
     */
    MappedStyles.prototype._classNamesSanitized;
    /**
     * @type {?}
     * @private
     */
    MappedStyles.prototype.mapping;
    /**
     * @type {?}
     * @private
     */
    MappedStyles.prototype.baseClasses;
    /**
     * @type {?}
     * @protected
     */
    MappedStyles.prototype._elementRef;
    /**
     * @type {?}
     * @protected
     */
    MappedStyles.prototype._renderer;
}

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: allianz-ngx-ndbx-core.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { BREAKPOINT_DESKTOP, BREAKPOINT_MOBILE, BREAKPOINT_TABLET, MappedStyles };
