// @allianz/ngx-ndbx/fesm2015/allianz-ngx-ndbx-circle-toggle.js
import { NxIconModule } from '@allianz/ngx-ndbx/icon';
import { EventEmitter, Component, ChangeDetectionStrategy, forwardRef, ChangeDetectorRef, ContentChildren, Output, Input, Optional, ViewChild, HostListener, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { startWith, filter, tap, takeUntil } from 'rxjs/operators';
import { Subject, merge } from 'rxjs';
import { UniqueSelectionDispatcher } from '@angular/cdk/collections';

/**
 * @fileoverview added by tsickle
 * Generated from: circle-toggle/toggle-button.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * \@docs-private
 * @abstract
 */
class ToggleButton {
}
if (false) {
    /** @type {?} */
    ToggleButton.prototype.id;
    /** @type {?} */
    ToggleButton.prototype.value;
    /** @type {?} */
    ToggleButton.prototype.checked;
    /** @type {?} */
    ToggleButton.prototype.name;
    /** @type {?} */
    ToggleButton.prototype.negative;
    /** @type {?} */
    ToggleButton.prototype.disabled;
    /** @type {?} */
    ToggleButton.prototype.tabIndex;
    /** @type {?} */
    ToggleButton.prototype.checkedChange;
    /** @type {?} */
    ToggleButton.prototype.selectionChange;
    /** @type {?} */
    ToggleButton.prototype.toggleButton;
    /**
     * @abstract
     * @param {?} event
     * @return {?}
     */
    ToggleButton.prototype.toggle = function (event) { };
    /**
     * @abstract
     * @return {?}
     */
    ToggleButton.prototype.setGroupSelection = function () { };
}

/**
 * @fileoverview added by tsickle
 * Generated from: circle-toggle-group/circle-toggle-group.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
let nextId = 0;
class NxCircleToggleGroupComponent {
    /**
     * @param {?} _changeDetectorRef
     */
    constructor(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._id = `nx-circle-toggle-group-${nextId++}`;
        this._destroyed = new Subject();
        /**
         * An event emitted when the selection changes. Outputs the value of the currently selected button.
         */
        this.valueChange = new EventEmitter();
        this._name = `toggle-group-${nextId++}`;
        this._negative = false;
        this._responsive = true;
        this.onChangeCallback = (/**
         * @param {?} value
         * @return {?}
         */
        (value) => { });
        this.onTouchedCallback = (/**
         * @return {?}
         */
        () => { });
    }
    /**
     * Id of the circle toggle group.
     *
     * If not set, the circle toggle group gets an incremented value by default.
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        this._id = value;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get id() {
        return this._id;
    }
    /**
     * Name that is used for accessibility.
     * @param {?} value
     * @return {?}
     */
    set name(value) {
        this._name = value;
        this.updateToggleButtonsNames();
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get name() {
        return this._name;
    }
    /**
     * Whether the circle toggle group is disabled.
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (this._disabled !== newValue) {
            this._disabled = newValue;
            this._changeDetectorRef.markForCheck();
        }
        if (this.buttons) {
            this.buttons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            button => button.disabled = value));
        }
    }
    /**
     * @return {?}
     */
    get disabled() {
        return this._disabled;
    }
    /**
     * Whether the circle toggle group uses the negative styling.
     * @param {?} value
     * @return {?}
     */
    set negative(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (this.negative !== newValue) {
            this._negative = newValue;
        }
        if (this.buttons) {
            this.buttons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            button => button.negative = value));
        }
    }
    /**
     * @return {?}
     */
    get negative() {
        return this._negative;
    }
    /**
     * The value of the selected circle toggle in the circle toggle group.
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        this.writeValue(value);
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * Whether the circle toggle group has a responsive behavior.
     * @param {?} value
     * @return {?}
     */
    set responsive(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (newValue !== this.responsive) {
            this._responsive = newValue;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get responsive() {
        return this._responsive;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        if (this.value !== value) {
            this._value = value;
            this.notifySelectedChild(value);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.subscribeToSelectionChanges();
        // react if a content child is deleted, added etc.
        this.buttons.changes
            .pipe(startWith(this.buttons), filter((/**
         * @param {?} toggles
         * @return {?}
         */
        toggles => toggles.length > 0)), tap((/**
         * @param {?} toggles
         * @return {?}
         */
        toggles => Promise.resolve().then((/**
         * @return {?}
         */
        () => {
            toggles.forEach((/**
             * @param {?} toggle
             * @return {?}
             */
            toggle => toggle.toggleButton.resetClasses()));
            this.buttons.first.toggleButton.setFirstButton();
            this.buttons.last.toggleButton.setLastButton();
        })))), takeUntil(this._destroyed))
            .subscribe((/**
         * @return {?}
         */
        () => this.subscribeToSelectionChanges()));
    }
    /**
     * \@docs-private
     * @return {?}
     */
    subscribeToSelectionChanges() {
        /** @type {?} */
        const changedOrDestroyed = merge(this.buttons.changes, this._destroyed);
        merge(...this.buttons.map((/**
         * @param {?} button
         * @return {?}
         */
        button => button.selectionChange)))
            .pipe(takeUntil(changedOrDestroyed))
            .subscribe((/**
         * @param {?} change
         * @return {?}
         */
        (change) => {
            this.onChangeCallback(change.value);
            this.valueChange.emit(change.value);
        }));
    }
    /**
     * \@docs-private
     * inform the selected button if the group selection is set programattically at the beginning
     * @param {?} newValue
     * @return {?}
     */
    notifySelectedChild(newValue) {
        if (this.buttons) {
            /** @type {?} */
            const selected = this.buttons.find((/**
             * @param {?} button
             * @return {?}
             */
            (button) => button.value === newValue));
            if (selected) {
                selected.setGroupSelection();
            }
        }
    }
    /**
     * \@docs-private
     * @return {?}
     */
    updateToggleButtonsNames() {
        if (this.buttons) {
            this.buttons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            button => button.name = this.name));
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get selectedButton() {
        return this.buttons ? this.buttons.find((/**
         * @param {?} button
         * @return {?}
         */
        button => button.checked)) : null;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get buttons() {
        return this._buttons;
    }
}
NxCircleToggleGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-circle-toggle-group',
                template: `<ng-content></ng-content>`,
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => NxCircleToggleGroupComponent)),
                        multi: true
                    }
                ],
                host: {
                    '[class.is-responsive]': 'responsive',
                    '[class.is-disabled]': 'disabled',
                    '[attr.aria-disabled]': 'disabled',
                    '[attr.aria-labelledby]': 'name',
                    '[attr.name]': 'name',
                    '[attr.id]': 'id',
                    '[class.nx-circle-toggle-group]': 'true',
                    'role': 'radiogroup'
                },
                styles: [":host{display:flex;flex-wrap:wrap;flex-direction:column;justify-content:space-around}@media (min-width:320px){:host{flex-direction:row}}:host(.is-responsive){flex-direction:column}@media (min-width:704px){:host(.is-responsive){flex-direction:row}}"]
            }] }
];
/** @nocollapse */
NxCircleToggleGroupComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
NxCircleToggleGroupComponent.propDecorators = {
    _buttons: [{ type: ContentChildren, args: [ToggleButton,] }],
    valueChange: [{ type: Output }],
    name: [{ type: Input }],
    disabled: [{ type: Input }],
    negative: [{ type: Input }],
    value: [{ type: Input }],
    responsive: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype._buttons;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype._id;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype._destroyed;
    /**
     * An event emitted when the selection changes. Outputs the value of the currently selected button.
     * @type {?}
     */
    NxCircleToggleGroupComponent.prototype.valueChange;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype._name;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype._disabled;
    /** @type {?} */
    NxCircleToggleGroupComponent.prototype._negative;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype._value;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype._responsive;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype.onChangeCallback;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype.onTouchedCallback;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleGroupComponent.prototype._changeDetectorRef;
}

/**
 * @fileoverview added by tsickle
 * Generated from: mobile-toggle-button/mobile-toggle-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * \@docs-private
 */
class NxMobileToggleButtonComponent {
    /**
     * @param {?} _changeDetectorRef
     */
    constructor(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this.checked = false;
        this.negative = false;
        this.disabled = false;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    setFirstButton() {
        this.isFirst = true;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * \@docs-private
     * @return {?}
     */
    setLastButton() {
        this.isLast = true;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * \@docs-private
     * @return {?}
     */
    resetClasses() {
        this.isFirst = false;
        this.isLast = false;
        this._changeDetectorRef.markForCheck();
    }
}
NxMobileToggleButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-mobile-toggle-button',
                template: "<nx-icon class=\"nx-toggle-button__mobile-check\" name=\"check\" aria-hidden=\"true\"></nx-icon>\n<!-- If there is no label defined, use the circleText instead -->\n<span class=\"nx-toggle-button__label-text\"><ng-template *ngIf=\"label; then labelContent else circleTextContent\"></ng-template></span>\n<span *ngIf=\"hint\" class=\"nx-toggle-button__label-hint\">({{hint}})</span>\n\n<ng-template #labelContent>{{label}}</ng-template>\n<ng-template #circleTextContent>{{circleText}}</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[class.is-disabled]': 'disabled',
                    '[class.is-negative]': 'negative',
                    '[class.is-first]': 'isFirst',
                    '[class.is-last]': 'isLast',
                    '[class.is-flipped]': 'checked'
                },
                styles: [":host{display:none;position:relative;transition:background-color .2s,border-color .2s,color .2s;flex-direction:row;border-width:1px 2px;border-color:#007ab3;border-color:var(--circle-toggle-border-color,#007ab3);border-style:solid;padding:11px 31px;width:100%;justify-content:center;align-items:center;cursor:pointer}:host:not(.is-flipped):not(.is-disabled):hover{background-color:#006192;background-color:var(--circle-toggle-hover-background-color,#006192);border-color:#006192;border-color:var(--circle-toggle-hover-border-color,#006192)}:host:not(.is-flipped):not(.is-disabled):hover .nx-toggle-button__label-hint,:host:not(.is-flipped):not(.is-disabled):hover .nx-toggle-button__label-text{color:#fff;color:var(--circle-toggle-hover-icon-color,#fff)}@media (max-width:703px){:host-context(.is-responsive){display:flex}}.nx-toggle-button__label-hint{color:#007ab3;color:var(--circle-toggle-icon-color,#007ab3);font-size:14px;font-size:var(--circle-toggle-hint-mobile-font-size,14px);line-height:20px;line-height:var(--circle-toggle-hint-mobile-line-height,20px);font-weight:300;font-weight:var(--circle-toggle-hint-mobile-font-weight,300);letter-spacing:.2px;letter-spacing:var(--circle-toggle-hint-mobile-letter-spacing,.2px)}.nx-toggle-button__label-text{color:#007ab3;color:var(--circle-toggle-icon-color,#007ab3);font-size:14px;font-size:var(--circle-toggle-label-mobile-font-size,14px);line-height:20px;line-height:var(--circle-toggle-label-mobile-line-height,20px);font-weight:var(--circle-toggle-label-mobile-font-weight,400);letter-spacing:.2px;letter-spacing:var(--circle-toggle-label-mobile-letter-spacing,.2px);padding-top:0;padding-right:4px;font-weight:600}.nx-toggle-button__mobile-check{font-size:24px;height:24px;display:inline-block;opacity:0;margin-right:8px;color:#fff;color:var(--circle-toggle-checkmark-color,#fff)}:host(.is-first){border-top-width:2px;border-radius:4px 4px 0 0}:host(.is-last){border-bottom-width:2px;border-radius:0 0 4px 4px}:host(.is-flipped){background-color:#007ab3;background-color:var(--circle-toggle-selected-background-color,#007ab3)}:host(.is-flipped) .nx-toggle-button__label-hint,:host(.is-flipped) .nx-toggle-button__label-text{color:#fff;color:var(--circle-toggle-selected-icon-color,#fff)}:host(.is-flipped) .nx-toggle-button__mobile-check{opacity:1}:host(.is-negative){border-color:#fff}:host(.is-negative) .nx-toggle-button__label-hint,:host(.is-negative) .nx-toggle-button__label-text,:host(.is-negative) .nx-toggle-button__mobile-check{color:#fff}:host(.is-negative):not(.is-flipped):not(.is-disabled):hover{background-color:#d9d9d9;border-color:#d9d9d9}:host(.is-negative):not(.is-flipped):not(.is-disabled):hover .nx-toggle-button__label-hint,:host(.is-negative):not(.is-flipped):not(.is-disabled):hover .nx-toggle-button__label-text{color:#006192}:host(.is-negative.is-flipped){background-color:#fff}:host(.is-negative.is-flipped) .nx-toggle-button__label-hint,:host(.is-negative.is-flipped) .nx-toggle-button__label-text,:host(.is-negative.is-flipped) .nx-toggle-button__mobile-check{color:#007ab3}:host(.is-disabled){opacity:.4;cursor:not-allowed}@media screen and (-ms-high-contrast:active){:host,:host(.is-negative){background-color:buttonFace;border-color:buttonText}:host .nx-toggle-button__label-hint,:host .nx-toggle-button__label-text,:host .nx-toggle-button__mobile-check,:host(.is-negative) .nx-toggle-button__label-hint,:host(.is-negative) .nx-toggle-button__label-text,:host(.is-negative) .nx-toggle-button__mobile-check{color:buttonText;-ms-high-contrast-adjust:none}:host(.is-flipped),:host(.is-negative.is-flipped){background-color:highlight}:host(.is-flipped) .nx-toggle-button__label-hint,:host(.is-flipped) .nx-toggle-button__label-text,:host(.is-flipped) .nx-toggle-button__mobile-check,:host(.is-negative.is-flipped) .nx-toggle-button__label-hint,:host(.is-negative.is-flipped) .nx-toggle-button__label-text,:host(.is-negative.is-flipped) .nx-toggle-button__mobile-check{color:highlightText}:host(.is-disabled){opacity:1;border-color:GrayText;color:GrayText}:host(.is-disabled).is-flipped .nx-toggle-button__mobile-check{color:GrayText}}"]
            }] }
];
/** @nocollapse */
NxMobileToggleButtonComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
NxMobileToggleButtonComponent.propDecorators = {
    checked: [{ type: Input }],
    negative: [{ type: Input }],
    disabled: [{ type: Input }],
    hint: [{ type: Input }],
    label: [{ type: Input }],
    circleText: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.checked;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.negative;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.disabled;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.hint;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.label;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.circleText;
    /**
     * \@docs-private
     * @type {?}
     */
    NxMobileToggleButtonComponent.prototype.isFirst;
    /**
     * \@docs-private
     * @type {?}
     */
    NxMobileToggleButtonComponent.prototype.isLast;
    /**
     * @type {?}
     * @private
     */
    NxMobileToggleButtonComponent.prototype._changeDetectorRef;
}

/**
 * @fileoverview added by tsickle
 * Generated from: circle-toggle/circle-toggle.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ToggleChangeEvent {
    /**
     * @param {?} button
     * @param {?} value
     */
    constructor(button, value) {
        this.button = button;
        this.value = value;
    }
}
if (false) {
    /**
     * A toggle button
     * @type {?}
     */
    ToggleChangeEvent.prototype.button;
    /**
     * The value of the toggle button that is sent with the event.
     * @type {?}
     */
    ToggleChangeEvent.prototype.value;
}
/** @type {?} */
let nextId$1 = 0;
class NxCircleToggleComponent extends ToggleButton {
    /**
     * @param {?} toggleGroup
     * @param {?} _checkedDispatcher
     * @param {?} _changeDetectorRef
     */
    constructor(/** @docs-private */ toggleGroup, _checkedDispatcher, _changeDetectorRef) {
        super();
        this.toggleGroup = toggleGroup;
        this._checkedDispatcher = _checkedDispatcher;
        this._changeDetectorRef = _changeDetectorRef;
        this._id = `toggle-button-${nextId$1++}`;
        /**
         * \@docs-private
         */
        this.inGroup = false;
        this._name = null;
        /**
         * An event that is emitted when the checked state fo the circle toggle changes.
         */
        this.checkedChange = new EventEmitter();
        /**
         * An event that is emitted when the checked state of the circle toggle changes.
         * The event object contains the circle toggle itself and its value (see ToggleChangeEvent).
         */
        this.selectionChange = new EventEmitter();
        this._checked = false;
        this._value = null;
        this._iconName = null;
        this._svg = null;
        this._svgChecked = null;
        this._circleText = null;
        this._label = null;
        this._hint = null;
        this._hover = false;
        /**
         * touched is set to true on touch devices.
         */
        this._touched = false;
        this._removeUniqueSelectionListener = (/**
         * @return {?}
         */
        () => { });
        this.onChangeCallback = (/**
         * @param {?} checked
         * @return {?}
         */
        (checked) => { });
        this.onTouchedCallback = (/**
         * @return {?}
         */
        () => { });
        if (this.toggleGroup) {
            this.name = this.toggleGroup.name;
        }
    }
    /**
     * Id of the circle toggle.
     *
     * If not set, the circle toggle gets an incremented value by default.
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        this._id = value;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get id() {
        return this._id;
    }
    /**
     * Name that is used for accessibility.
     * @param {?} value
     * @return {?}
     */
    set name(value) {
        this._name = value;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get name() {
        return this._name;
    }
    /**
     * Whether the circle toggle is checked.
     * @param {?} value
     * @return {?}
     */
    set checked(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (this.checked !== newValue) {
            this._checked = newValue;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get checked() {
        return this._checked;
    }
    /**
     * The value that is used in the model.
     * @param {?} newValue
     * @return {?}
     */
    set value(newValue) {
        this._value = newValue;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * Id of the icon that should be displayed.
     * @param {?} name
     * @return {?}
     */
    set iconName(name) {
        this._iconName = name;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get iconName() {
        return this._iconName;
    }
    /**
     * SVG that is displayed if the circle toggle is unchecked.
     * @param {?} src
     * @return {?}
     */
    set svg(src) {
        this._svg = src;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get svg() {
        return this._svg;
    }
    /**
     * SVG that is displayed if the circle toggle is checked.
     * @param {?} src
     * @return {?}
     */
    set svgChecked(src) {
        this._svgChecked = src;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get svgChecked() {
        return this._svgChecked;
    }
    /**
     * A text that is displayed inside the circle toggle.
     * @param {?} value
     * @return {?}
     */
    set circleText(value) {
        if (value !== this._circleText) {
            this._circleText = value;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get circleText() {
        return this._circleText;
    }
    /**
     * Label displayed below the circle.
     * @param {?} value
     * @return {?}
     */
    set label(value) {
        this._label = value;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get label() {
        return this._label;
    }
    /**
     * Additional hint displayed below the label.
     * @param {?} value
     * @return {?}
     */
    set hint(value) {
        this._hint = value;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get hint() {
        return this._hint;
    }
    /**
     * Whether the circle toggle uses the negative set of styling.
     * @param {?} value
     * @return {?}
     */
    set negative(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (this.negative !== newValue) {
            this._negative = newValue;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get negative() {
        return this._negative;
    }
    /**
     * Whether the circle toggle has a responsive behavior.
     * @param {?} value
     * @return {?}
     */
    set responsive(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (this.responsive !== newValue) {
            this._responsive = newValue;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get responsive() {
        return this._responsive;
    }
    /**
     * Whether the circle toggle is disabled.
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        /** @type {?} */
        const newValue = coerceBooleanProperty(value);
        if (this.disabled !== newValue) {
            this._disabled = newValue;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get disabled() {
        return this._disabled;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    onMouseEnter() {
        if (!this._touched) {
            this._hover = true;
        }
    }
    /**
     * \@docs-private
     * @return {?}
     */
    onMouseLeave() {
        if (!this._touched) {
            this._hover = false;
        }
    }
    /**
     * \@docs-private
     * @return {?}
     */
    onTouchStart() {
        this._touched = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.toggleGroup) {
            this.attachListenerForGroup();
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.toggleGroup) {
            Promise.resolve().then((/**
             * @return {?}
             */
            () => {
                this.inGroup = true;
                this.negative = this.toggleGroup.negative;
                this.disabled = this.toggleGroup.disabled;
                this.responsive = this.toggleGroup.responsive;
                this.id = this.toggleGroup.id + `-button-${nextId$1++}`;
            }));
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        // function returned by the listener
        this._removeUniqueSelectionListener();
    }
    /**
     * \@docs-private
     * @return {?}
     */
    attachListenerForGroup() {
        this._removeUniqueSelectionListener =
            this._checkedDispatcher.listen((/**
             * @param {?} groupId
             * @param {?} buttonId
             * @return {?}
             */
            (groupId, buttonId) => {
                if (this.id !== buttonId && groupId === this.toggleGroup.id) {
                    this.checked = false;
                }
            }));
    }
    /**
     * @param {?} newValue
     * @return {?}
     */
    writeValue(newValue) {
        this.checked = newValue;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * \@docs-private
     * @param {?} event
     * @return {?}
     */
    toggle(event) {
        event.preventDefault();
        event.stopPropagation();
        if ((this.toggleGroup && this.checked) || this.disabled) {
            return;
        }
        else {
            this.checked = !this.checked;
            this.onChangeCallback(this.checked);
            this.checkedChange.emit(this.checked);
            this.selectionChange.emit(new ToggleChangeEvent(this, this.value));
            if (this.toggleGroup) {
                this._checkedDispatcher.notify(this.toggleGroup.id, this.id);
            }
        }
    }
    /**
     * \@docs-private
     *
     * method called by the group if the internal value of the group is changed programatically,
     * does not trigger change emission
     * @return {?}
     */
    setGroupSelection() {
        // propagate changes only if the value in the group is different than the button checked value
        if (!this.checked) {
            this.checked = !this.checked;
            this.onChangeCallback(this.checked);
            this._checkedDispatcher.notify(this.toggleGroup.id, this.id);
        }
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get svgButton() {
        return !!this.svg && !!this.svgChecked;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get svgUrl() {
        /** @type {?} */
        let useFilledSvg = this.checked || (!this.disabled && this._hover);
        if (this.negative) {
            useFilledSvg = !useFilledSvg;
        }
        return useFilledSvg ? this.svgChecked : this.svg;
    }
    /**
     * \@docs-private
     * @return {?}
     */
    get type() {
        return this.toggleGroup ? 'radio' : 'checkbox';
    }
    /**
     * \@docs-private
     * @param {?} event
     * @return {?}
     */
    handleEnterKey(event) {
        if (!this.toggleGroup) {
            this.toggle(event);
        }
    }
}
NxCircleToggleComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-circle-toggle',
                template: "<input [id]=\"id + '-input'\"\n       class=\"nx-circle-toggle__input\"\n       [attr.aria-labelledby]=\"id + '-label'\"\n       [type]=\"type\"\n       [attr.name]=\"name\"\n       [checked]=\"checked\"\n       [disabled]= \"disabled\"\n       [attr.aria-checked]=\"checked\"\n       (keydown.enter)=\"handleEnterKey($event)\"\n       (click)=\"toggle($event)\">\n<label class=\"nx-circle-toggle__label\"\n  [for]=\"id + '-input'\"\n  [id]=\"id + '-label'\">\n    <nx-icon-toggle-button\n      [label]=\"label\"\n      [hint]=\"hint\"\n      [checked]=\"checked\"\n      [icon]=\"iconName\"\n      [negative]=\"negative\"\n      [svg]=\"svgUrl\"\n      [disabled]=\"disabled\"\n      [inGroup]=\"inGroup\"\n      [touched]=\"_touched\"\n      [circleText]=\"circleText\">\n    </nx-icon-toggle-button>\n    <!-- The mobile toggle button is needed for the responsive mobile version of the component.\n    The display swtich is handled in the css. -->\n    <nx-mobile-toggle-button\n      [label]=\"label\"\n      [hint]=\"hint\"\n      [checked]=\"checked\"\n      [negative]=\"negative\"\n      [disabled]=\"disabled\"\n      [circleText]=\"circleText\">\n    </nx-mobile-toggle-button>\n</label>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: ToggleButton,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => NxCircleToggleComponent))
                    },
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => NxCircleToggleComponent)),
                        multi: true
                    }
                ],
                host: {
                    '[class.nx-toggle-circle]': 'true',
                    '[class.in-group]': 'inGroup',
                    '[class.is-disabled]': 'disabled',
                    '[class.is-responsive]': 'responsive'
                },
                styles: [":host{display:inline-block;vertical-align:top;position:relative}:host:not(.is-responsive){opacity:.9999}:host .nx-circle-toggle__input{opacity:0;position:absolute}:host:not(.in-group){margin-right:64px}@media (max-width:703px){:host:not(.in-group){margin-right:48px}}:host-context([data-whatinput=keyboard]) .nx-circle-toggle__input:focus+.nx-circle-toggle__label ::ng-deep .nx-toggle-circle__check-focus,:host-context([data-whatinput=keyboard]) .nx-circle-toggle__input:focus+.nx-circle-toggle__label ::ng-deep .nx-toggle-circle__icon-wrapper{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6)}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]) .nx-circle-toggle__input:focus+.nx-circle-toggle__label ::ng-deep .nx-toggle-circle__check-focus,:host-context([data-whatinput=keyboard]) .nx-circle-toggle__input:focus+.nx-circle-toggle__label ::ng-deep .nx-toggle-circle__icon-wrapper{box-shadow:0 0 0 2px background,0 0 0 6px windowText}}:host-context([data-whatinput=keyboard]) .nx-circle-toggle__input:focus+.nx-circle-toggle__label ::ng-deep nx-mobile-toggle-button{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6);z-index:1}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]) .nx-circle-toggle__input:focus+.nx-circle-toggle__label ::ng-deep nx-mobile-toggle-button{box-shadow:0 0 0 2px background,0 0 0 6px windowText}}"]
            }] }
];
/** @nocollapse */
NxCircleToggleComponent.ctorParameters = () => [
    { type: NxCircleToggleGroupComponent, decorators: [{ type: Optional }] },
    { type: UniqueSelectionDispatcher },
    { type: ChangeDetectorRef }
];
NxCircleToggleComponent.propDecorators = {
    name: [{ type: Input }],
    checkedChange: [{ type: Output }],
    selectionChange: [{ type: Output }],
    checked: [{ type: Input }],
    value: [{ type: Input }],
    iconName: [{ type: Input, args: ['icon',] }],
    svg: [{ type: Input }],
    svgChecked: [{ type: Input }],
    circleText: [{ type: Input }],
    label: [{ type: Input }],
    hint: [{ type: Input }],
    negative: [{ type: Input }],
    responsive: [{ type: Input }],
    disabled: [{ type: Input }],
    toggleButton: [{ type: ViewChild, args: [NxMobileToggleButtonComponent, { static: true },] }],
    onMouseEnter: [{ type: HostListener, args: ['mouseenter',] }],
    onMouseLeave: [{ type: HostListener, args: ['mouseleave',] }],
    onTouchStart: [{ type: HostListener, args: ['touchstart',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._id;
    /**
     * \@docs-private
     * @type {?}
     */
    NxCircleToggleComponent.prototype.inGroup;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._name;
    /**
     * An event that is emitted when the checked state fo the circle toggle changes.
     * @type {?}
     */
    NxCircleToggleComponent.prototype.checkedChange;
    /**
     * An event that is emitted when the checked state of the circle toggle changes.
     * The event object contains the circle toggle itself and its value (see ToggleChangeEvent).
     * @type {?}
     */
    NxCircleToggleComponent.prototype.selectionChange;
    /** @type {?} */
    NxCircleToggleComponent.prototype._checked;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._value;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._iconName;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._svg;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._svgChecked;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._circleText;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._label;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._hint;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._negative;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._responsive;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._disabled;
    /**
     * \@docs-private
     * @type {?}
     */
    NxCircleToggleComponent.prototype.toggleButton;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._hover;
    /**
     * touched is set to true on touch devices.
     * @type {?}
     */
    NxCircleToggleComponent.prototype._touched;
    /** @type {?} */
    NxCircleToggleComponent.prototype._removeUniqueSelectionListener;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype.onChangeCallback;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype.onTouchedCallback;
    /**
     * \@docs-private
     * @type {?}
     */
    NxCircleToggleComponent.prototype.toggleGroup;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._checkedDispatcher;
    /**
     * @type {?}
     * @private
     */
    NxCircleToggleComponent.prototype._changeDetectorRef;
}

/**
 * @fileoverview added by tsickle
 * Generated from: icon-toggle-button/icon-toggle-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * \@docs-private
 */
class NxIconToggleButtonComponent {
    constructor() {
        this.inGroup = true;
    }
}
NxIconToggleButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-icon-toggle-button',
                template: "<div class=\"nx-toggle-circle__icon-wrapper\">\n  <ng-container *ngIf=\"!inGroup\">\n    <span class=\"nx-toggle-circle__check-icon\">\n      <nx-icon name=\"check\"></nx-icon>\n    </span>\n    <div class=\"nx-toggle-circle__check-focus\"></div>\n  </ng-container>\n  <span class=\"nx-toggle-circle__icon\">\n    <nx-icon size=\"m\" *ngIf=\"icon\" [name]=\"icon\"></nx-icon>\n    <img *ngIf=\"svgUrl\" class=\"nx-toggle-circle__icon-image\" [src]=\"svgUrl\" />\n    <span *ngIf=\"circleText\" class=\"nx-toggle-circle__content-text\">{{circleText}}</span>\n  </span>\n</div>\n<span [class.is-label-bold]=\"hint\" class=\"nx-toggle-circle__label-text\">{{label}}</span>\n<span *ngIf=\"hint\" class=\"nx-toggle-circle__label-hint\">{{hint}}</span>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[class.nx-circle-button]': 'true',
                    '[class.is-flipped]': 'checked',
                    '[class.has-checkmark]': '!inGroup',
                    '[class.is-disabled]': 'disabled',
                    '[class.is-negative]': 'negative',
                    '[class.is-touched]': 'touched'
                },
                styles: ["@media (max-width:703px){:host-context(.is-responsive){display:none}}:host{display:flex;flex-direction:column;padding:0;align-items:center;cursor:pointer;position:relative}:host .nx-toggle-circle__label-text{font-size:16px;font-size:var(--circle-toggle-label-font-size,16px);line-height:24px;line-height:var(--circle-toggle-label-line-height,24px);font-weight:400;font-weight:var(--circle-toggle-label-font-weight,400);letter-spacing:0;letter-spacing:var(--circle-toggle-label-letter-spacing,0);color:#414141;color:var(--circle-toggle-label-color,#414141);position:relative;display:inline-block;padding-top:12px;max-width:16rem;z-index:1}@media (max-width:703px){:host .nx-toggle-circle__label-text{font-size:14px;font-size:var(--circle-toggle-label-mobile-font-size,14px);line-height:20px;line-height:var(--circle-toggle-label-mobile-line-height,20px);font-weight:400;font-weight:var(--circle-toggle-label-mobile-font-weight,400);letter-spacing:.2px;letter-spacing:var(--circle-toggle-label-mobile-letter-spacing,.2px)}}:host .is-label-bold{font-weight:700}:host .nx-toggle-circle__icon-wrapper{height:80px;width:80px;border-radius:50%;border-color:var(--circle-toggle-border-color,#007ab3);display:flex;align-items:center;justify-content:center;transition:background-color .2s,border-color .2s,box-shadow .2s;border:2px solid #007ab3}@media (max-width:703px){:host .is-label-bold{font-weight:400}:host .nx-toggle-circle__icon-wrapper{height:64px;width:64px}:host .nx-toggle-circle__icon-wrapper .nx-toggle-circle__icon nx-icon{font-size:40px}}:host .nx-toggle-circle__check-icon{background-color:#3da556;background-color:var(--circle-toggle-checkmark-background-color,#3da556);color:#fff;color:var(--circle-toggle-checkmark-color,#fff);border-color:var(--circle-toggle-checkmark-border-color,#fff);border-radius:50%;font-size:24px;width:32px;height:32px;position:absolute;top:-8px;right:calc(50% - 48px);transition:opacity .3s;z-index:1;opacity:0;display:flex;justify-content:center;align-items:center;border:2px solid #fff}:host .nx-toggle-circle__check-focus{border-radius:50%;width:28px;height:28px;position:absolute;top:-6px;right:calc(50% - 48px + 2px);z-index:-1;opacity:0;transition:box-shadow .2s}:host .nx-toggle-circle__content-text{font-size:24px;font-size:var(--circle-toggle-content-text-font-size,24px);line-height:32px;line-height:var(--circle-toggle-content-text-line-height,32px);font-weight:400;font-weight:var(--circle-toggle-content-text-font-weight,400);letter-spacing:.2px;letter-spacing:var(--circle-toggle-content-text-letter-spacing,.2px);color:#006192;color:var(--circle-toggle-content-text-color,#006192)}@media (max-width:703px){:host .nx-toggle-circle__content-text{font-size:20px;font-size:var(--circle-toggle-content-text-mobile-font-size,20px);line-height:28px;line-height:var(--circle-toggle-content-text-mobile-line-height,28px);font-weight:400;font-weight:var(--circle-toggle-content-text-mobile-font-weight,400);letter-spacing:.2px;letter-spacing:var(--circle-toggle-content-text-mobile-letter-spacing,.2px)}:host .nx-toggle-circle__label-hint .nx-toggle-circle__icon-wrapper{width:64px;height:64px}:host .nx-toggle-circle__label-hint .nx-toggle-circle__icon nx-icon{font-size:32px}:host .nx-toggle-circle__label-hint .nx-toggle-circle__check-icon{right:calc(50% - 40px)}:host .nx-toggle-circle__label-hint .nx-toggle-circle__icon-image{width:64px;height:64px}}:host .nx-toggle-circle__label-hint{font-size:16px;font-size:var(--circle-toggle-hint-font-size,16px);line-height:24px;line-height:var(--circle-toggle-hint-line-height,24px);font-weight:400;font-weight:var(--circle-toggle-hint-font-weight,400);letter-spacing:0;letter-spacing:var(--circle-toggle-hint-letter-spacing,0);color:#414141;color:var(--circle-toggle-hint-color,#414141);display:inline-block;transition:color .2s;z-index:1}@media (max-width:703px){:host .nx-toggle-circle__label-hint{font-size:14px;font-size:var(--circle-toggle-hint-mobile-font-size,14px);line-height:20px;line-height:var(--circle-toggle-hint-mobile-line-height,20px);font-weight:300;font-weight:var(--circle-toggle-hint-mobile-font-weight,300);letter-spacing:.2px;letter-spacing:var(--circle-toggle-hint-mobile-letter-spacing,.2px)}}:host .nx-toggle-circle__icon{display:flex;justify-content:center;align-content:center}:host .nx-toggle-circle__icon nx-icon{color:#007ab3;color:var(--circle-toggle-icon-color,#007ab3)}:host .nx-toggle-circle__icon-image{pointer-events:none;position:relative;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}:host(.is-flipped) .nx-toggle-circle__icon-wrapper{background-color:#007ab3;background-color:var(--circle-toggle-selected-background-color,#007ab3);border-color:#007ab3;border-color:var(--circle-toggle-selected-border-color,#007ab3)}:host(.is-flipped) .nx-toggle-circle__icon nx-icon{color:#fff;color:var(--circle-toggle-selected-icon-color,#fff)}:host(.is-flipped) .nx-toggle-circle__content-text{color:#fff;color:var(--circle-toggle-selected-content-text-color,#fff)}:host(.is-flipped) .nx-toggle-circle__check-focus,:host(.is-flipped) .nx-toggle-circle__check-icon{opacity:1}:host:not(.is-disabled):not(.is-negative):not(.is-touched):hover .nx-toggle-circle__icon-wrapper{border-color:#006192;border-color:var(--circle-toggle-hover-border-color,#006192);background-color:#006192;background-color:var(--circle-toggle-hover-background-color,#006192)}:host:not(.is-disabled):not(.is-negative):not(.is-touched):hover:not(.is-flipped) .nx-toggle-circle__icon-wrapper{border-color:#007ab3;border-color:var(--circle-toggle-selected-border-color,#007ab3);background-color:#007ab3;background-color:var(--circle-toggle-selected-background-color,#007ab3)}:host:not(.is-disabled):not(.is-negative):not(.is-touched):hover .nx-toggle-circle__content-text{color:#fff;color:var(--circle-toggle-hover-icon-color,#fff)}:host:not(.is-disabled):not(.is-negative):not(.is-touched):hover .nx-toggle-circle__icon nx-icon{color:#fff;color:var(--circle-toggle-hover-icon-color,#fff)}:host:hover.is-flipped:not(.has-checkmark){cursor:auto}:host(.is-disabled){opacity:.4;cursor:not-allowed}:host(.is-negative) .nx-toggle-circle__icon-wrapper{border-color:#fff;background-color:transparent}:host(.is-negative) .nx-toggle-circle__content-text,:host(.is-negative) .nx-toggle-circle__icon nx-icon,:host(.is-negative) .nx-toggle-circle__label-hint,:host(.is-negative) .nx-toggle-circle__label-text{color:#fff}:host(.is-negative):not(.is-disabled):not(.is-touched):hover .nx-toggle-circle__icon-wrapper,:host(.is-negative.is-flipped) .nx-toggle-circle__icon-wrapper{border-color:#fff;background-color:#fff}:host(.is-negative):not(.is-disabled):not(.is-touched):hover .nx-toggle-circle__icon nx-icon,:host(.is-negative.is-flipped) .nx-toggle-circle__icon nx-icon{color:#007ab3}:host(.is-negative):not(.is-disabled):not(.is-touched):hover .nx-toggle-circle__content-text,:host(.is-negative.is-flipped) .nx-toggle-circle__content-text{color:#006192}@media screen and (-ms-high-contrast:active){.nx-toggle-circle__icon-wrapper{border-color:buttonText;background-color:buttonFace}.nx-toggle-circle__content-text{-ms-high-contrast-adjust:none}.nx-toggle-circle__label-hint,.nx-toggle-circle__label-text{color:buttonText}:host(.is-negative):not(.is-flipped):not(.is-disabled):not(.is-touched) .nx-toggle-circle__content-text,:host(.is-negative):not(.is-flipped):not(.is-disabled):not(.is-touched) .nx-toggle-circle__icon nx-icon,:host:not(.is-flipped) .nx-toggle-circle__content-text,:host:not(.is-flipped) .nx-toggle-circle__icon nx-icon{color:buttonText}:host(.is-flipped) .nx-toggle-circle__icon-wrapper,:host(.is-negative.is-flipped) .nx-toggle-circle__icon-wrapper{background:highlight;border-color:highlight;color:highlightText}:host(.is-flipped) .nx-toggle-circle__content-text,:host(.is-flipped) .nx-toggle-circle__icon nx-icon,:host(.is-negative.is-flipped) .nx-toggle-circle__content-text,:host(.is-negative.is-flipped) .nx-toggle-circle__icon nx-icon{color:highlightText}:host(.is-flipped) .nx-toggle-circle__check-icon,:host(.is-negative.is-flipped) .nx-toggle-circle__check-icon{border-color:windowText;color:windowText}:host.is-negative:not(.is-disabled):not(.is-touched):hover.is-flipped .nx-toggle-circle__content-text,:host.is-negative:not(.is-disabled):not(.is-touched):hover.is-flipped .nx-toggle-circle__icon nx-icon,:host:not(.is-disabled):not(.is-touched):hover.is-flipped .nx-toggle-circle__content-text,:host:not(.is-disabled):not(.is-touched):hover.is-flipped .nx-toggle-circle__icon nx-icon{color:highlightText}:host.is-negative:not(.is-disabled):not(.is-touched):hover:not(.is-flipped) .nx-toggle-circle__content-text,:host.is-negative:not(.is-disabled):not(.is-touched):hover:not(.is-flipped) .nx-toggle-circle__icon nx-icon,:host:not(.is-disabled):not(.is-touched):hover:not(.is-flipped) .nx-toggle-circle__content-text,:host:not(.is-disabled):not(.is-touched):hover:not(.is-flipped) .nx-toggle-circle__icon nx-icon{color:buttonText}:host(.is-disabled),:host(.is-disabled.is-flipped),:host(.is-disabled.is-flipped):hover{opacity:1}:host(.is-disabled) .nx-toggle-circle__icon-wrapper,:host(.is-disabled.is-flipped) .nx-toggle-circle__icon-wrapper,:host(.is-disabled.is-flipped):hover .nx-toggle-circle__icon-wrapper{border-color:GrayText;background:buttonFace;color:GrayText}:host(.is-disabled) .nx-toggle-circle__icon-wrapper .nx-toggle-circle__content-text,:host(.is-disabled) .nx-toggle-circle__icon-wrapper .nx-toggle-circle__icon nx-icon,:host(.is-disabled.is-flipped) .nx-toggle-circle__icon-wrapper .nx-toggle-circle__content-text,:host(.is-disabled.is-flipped) .nx-toggle-circle__icon-wrapper .nx-toggle-circle__icon nx-icon,:host(.is-disabled.is-flipped):hover .nx-toggle-circle__icon-wrapper .nx-toggle-circle__content-text,:host(.is-disabled.is-flipped):hover .nx-toggle-circle__icon-wrapper .nx-toggle-circle__icon nx-icon{color:GrayText}:host(.is-disabled) .nx-toggle-circle__check-icon,:host(.is-disabled.is-flipped) .nx-toggle-circle__check-icon,:host(.is-disabled.is-flipped):hover .nx-toggle-circle__check-icon{border-color:GrayText;color:GrayText}:host(.is-disabled) .nx-toggle-circle__label-hint,:host(.is-disabled) .nx-toggle-circle__label-text,:host(.is-disabled.is-flipped) .nx-toggle-circle__label-hint,:host(.is-disabled.is-flipped) .nx-toggle-circle__label-text,:host(.is-disabled.is-flipped):hover .nx-toggle-circle__label-hint,:host(.is-disabled.is-flipped):hover .nx-toggle-circle__label-text{color:GrayText}}"]
            }] }
];
NxIconToggleButtonComponent.propDecorators = {
    inGroup: [{ type: Input }],
    hint: [{ type: Input }],
    label: [{ type: Input }],
    checked: [{ type: Input }],
    icon: [{ type: Input }],
    svgUrl: [{ type: Input, args: ['svg',] }],
    disabled: [{ type: Input }],
    negative: [{ type: Input }],
    touched: [{ type: Input }],
    circleText: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.inGroup;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.hint;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.label;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.checked;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.icon;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.svgUrl;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.disabled;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.negative;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.touched;
    /** @type {?} */
    NxIconToggleButtonComponent.prototype.circleText;
}

/**
 * @fileoverview added by tsickle
 * Generated from: circle-toggle.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NxCircleToggleModule {
}
NxCircleToggleModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    NxCircleToggleGroupComponent,
                    NxCircleToggleComponent,
                    NxIconToggleButtonComponent,
                    NxMobileToggleButtonComponent
                ],
                exports: [
                    NxCircleToggleGroupComponent,
                    NxIconToggleButtonComponent,
                    NxCircleToggleComponent,
                    NxMobileToggleButtonComponent
                ],
                imports: [
                    CommonModule,
                    NxIconModule
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: allianz-ngx-ndbx-circle-toggle.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NxCircleToggleComponent, NxCircleToggleGroupComponent, NxCircleToggleModule, NxIconToggleButtonComponent, NxMobileToggleButtonComponent, ToggleChangeEvent, ToggleButton as a };
