// @allianz/ngx-ndbx/esm5/natural-language-form/natural-language-form.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: natural-language-form/natural-language-form.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChildren, HostListener, Input, QueryList, } from '@angular/core';
import { merge, Subject, Subscription } from 'rxjs';
import { delay, throttleTime } from 'rxjs/operators';
import { NxWordComponent } from './word.component';
/** @type {?} */
var DEFAULT_SIZE = 'large';
var NxNaturalLanguageFormComponent = /** @class */ (function () {
    function NxNaturalLanguageFormComponent(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._negative = false;
        /**
         * \@docs-private
         */
        this.resizeEvent$ = new Subject();
        /**
         * \@docs-private
         */
        this.updatePopoversSubscription = Subscription.EMPTY;
        /**
         * \@docs-private
         */
        this._size = DEFAULT_SIZE;
    }
    Object.defineProperty(NxNaturalLanguageFormComponent.prototype, "negative", {
        get: /**
         * @return {?}
         */
        function () {
            return this._negative;
        },
        /** Whether the negative set of stylings should be used. */
        set: /**
         * Whether the negative set of stylings should be used.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._negative = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxNaturalLanguageFormComponent.prototype, "size", {
        get: /**
         * @return {?}
         */
        function () {
            return this._size;
        },
        /**
         * Sets the size of the NLF. Default value: large
         */
        set: /**
         * Sets the size of the NLF. Default value: large
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._size = value;
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /** @docs-private */
    /**
     * \@docs-private
     * @param {?} event
     * @return {?}
     */
    NxNaturalLanguageFormComponent.prototype.onResize = /**
     * \@docs-private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.resizeEvent$.next();
    };
    /**
     * @return {?}
     */
    NxNaturalLanguageFormComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Collect all words and listen for changes so we can update any open error popover
        // which would otherwise get wrongly positioned.
        /** @type {?} */
        var subjects = this._words.map((/**
         * @param {?} word
         * @return {?}
         */
        function (word) { return word.inputChanges; }));
        /** @type {?} */
        var source = merge.apply(void 0, tslib_1.__spread(subjects));
        this.updatePopoversSubscription = source.subscribe((/**
         * @return {?}
         */
        function () {
            _this.updatePositionPopovers();
        }));
        this.resizeObservable = this.resizeEvent$.pipe(throttleTime(500), delay(100));
        this.resizeObservable.subscribe((/**
         * @return {?}
         */
        function () { return _this.resizeWords(); }));
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxNaturalLanguageFormComponent.prototype.resizeWords = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this._words.forEach((/**
         * @param {?} word
         * @return {?}
         */
        function (word) {
            word.updateCurrentTextWidth();
        }));
    };
    /**
     * @return {?}
     */
    NxNaturalLanguageFormComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.updatePopoversSubscription.unsubscribe();
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxNaturalLanguageFormComponent.prototype.updatePositionPopovers = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        if (this._words) {
            this._words.forEach((/**
             * @param {?} word
             * @return {?}
             */
            function (word) {
                word.repositionError();
            }));
        }
    };
    NxNaturalLanguageFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-natural-language-form',
                    template: "<div class=\"nx-natural-language-form__wrapper\">\n    <ng-content></ng-content>\n</div>",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.is-negative]': 'negative',
                        '[class.nx-natural-language-form--small]': 'size === "small"',
                        '[class.nx-natural-language-form--large]': 'size === "large"',
                    },
                    styles: [":host{display:block;font-size:40px;font-size:var(--natural-language-form-large-font-size,40px);line-height:52px;line-height:var(--natural-language-form-large-line-height,52px);font-weight:300;font-weight:var(--natural-language-form-large-font-weight,300);letter-spacing:.3px;letter-spacing:var(--natural-language-form-large-letter-spacing,.3px);color:#414141;color:var(--natural-language-form-text-color,#414141);padding-top:72px;padding-bottom:80px}:host ::ng-deep .c-input{font-size:inherit;line-height:inherit}:host ::ng-deep nx-word+nx-word{margin-left:2px}:host.is-negative{color:#fff}@media (max-width:991px){:host{font-size:26px;font-size:var(--natural-language-form-small-font-size,26px);line-height:36px;line-height:var(--natural-language-form-small-line-height,36px);font-weight:300;font-weight:var(--natural-language-form-small-font-weight,300);letter-spacing:.3px;letter-spacing:var(--natural-language-form-small-letter-spacing,.3px)}}.nx-natural-language-form__wrapper{display:flex;flex-wrap:wrap}.nx-natural-language-form__wrapper>::ng-deep *{margin-bottom:8px}:host(.nx-natural-language-form--small){font-size:26px;font-size:var(--natural-language-form-small-font-size,26px);line-height:36px;line-height:var(--natural-language-form-small-line-height,36px);font-weight:300;font-weight:var(--natural-language-form-small-font-weight,300);letter-spacing:.3px;letter-spacing:var(--natural-language-form-small-letter-spacing,.3px)}:host(.nx-natural-language-form--small) ::ng-deep .c-input{font-size:inherit;line-height:inherit}"]
                }] }
    ];
    /** @nocollapse */
    NxNaturalLanguageFormComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxNaturalLanguageFormComponent.propDecorators = {
        _words: [{ type: ContentChildren, args: [NxWordComponent,] }],
        negative: [{ type: Input, args: ['nxNegative',] }],
        size: [{ type: Input }],
        onResize: [{ type: HostListener, args: ['window:orientationchange', ['$event'],] }, { type: HostListener, args: ['window:resize', ['$event'],] }]
    };
    return NxNaturalLanguageFormComponent;
}());
export { NxNaturalLanguageFormComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxNaturalLanguageFormComponent.prototype._negative;
    /**
     * \@docs-private
     * @type {?}
     */
    NxNaturalLanguageFormComponent.prototype.resizeEvent$;
    /**
     * \@docs-private
     * @type {?}
     */
    NxNaturalLanguageFormComponent.prototype.resizeObservable;
    /** @type {?} */
    NxNaturalLanguageFormComponent.prototype._words;
    /**
     * \@docs-private
     * @type {?}
     */
    NxNaturalLanguageFormComponent.prototype.updatePopoversSubscription;
    /**
     * \@docs-private
     * @type {?}
     * @private
     */
    NxNaturalLanguageFormComponent.prototype._size;
    /**
     * @type {?}
     * @private
     */
    NxNaturalLanguageFormComponent.prototype._changeDetectorRef;
}
