// @allianz/ngx-ndbx/esm5/base/label/label.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: label/label.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, Optional, Inject, InjectionToken } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Subject } from 'rxjs';
/** @type {?} */
var nextId = 0;
/**
 * @record
 */
export function LabelDefaultOptions() { }
if (false) {
    /**
     * Sets the default appearance. (optional)
     * @type {?|undefined}
     */
    LabelDefaultOptions.prototype.size;
}
/** @type {?} */
var DEFAULT_SIZE = 'large';
/** @type {?} */
export var LABEL_DEFAULT_OPTIONS = new InjectionToken('LABEL_DEFAULT_OPTIONS');
var NxLabelComponent = /** @class */ (function () {
    function NxLabelComponent(_defaultOptions) {
        this._defaultOptions = _defaultOptions;
        this._stateChanges = new Subject();
        this._disabled = false;
        this._negative = false;
        this._id = "nx-label-" + nextId++;
        this._size = (this._defaultOptions && this._defaultOptions.size) ? this._defaultOptions.size : DEFAULT_SIZE;
    }
    Object.defineProperty(NxLabelComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        /** Sets the label to disabled */
        set: /**
         * Sets the label to disabled
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = coerceBooleanProperty(value);
            this._stateChanges.next();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxLabelComponent.prototype, "negative", {
        get: /**
         * @return {?}
         */
        function () {
            return this._negative;
        },
        /** Sets the label to disabled */
        set: /**
         * Sets the label to disabled
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._negative = coerceBooleanProperty(value);
            this._stateChanges.next();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxLabelComponent.prototype, "id", {
        get: /**
         * @return {?}
         */
        function () {
            return this._id;
        },
        /** Sets the Id of the label */
        set: /**
         * Sets the Id of the label
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this._id !== value) {
                this._id = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxLabelComponent.prototype, "size", {
        get: /**
         * @return {?}
         */
        function () {
            return this._size;
        },
        /**
         * **NDBX Expert option**
         * Sets the appearance of the label */
        set: /**
         * **NDBX Expert option**
         * Sets the appearance of the label
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._size = value;
            this._stateChanges.next();
        },
        enumerable: true,
        configurable: true
    });
    NxLabelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-label',
                    template: "<label class=\"nx-label__content\">\n  <ng-content></ng-content>\n</label>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[attr.disabled]': 'disabled',
                        '[attr.aria-labelledby]': 'id || null',
                        '[class.nx-label--negative]': 'negative',
                        '[class.nx-label--large]': 'size === "large"',
                        '[class.nx-label--small]': 'size === "small"',
                        '[attr.id]': 'id',
                    },
                    styles: [":host{display:block;color:#414141;color:var(--text-01,#414141)}:host.nx-label--small{font-size:16px;font-size:var(--base-label-small-font-size,16px);line-height:24px;line-height:var(--base-label-small-line-height,24px);font-weight:600;font-weight:var(--base-label-small-font-weight,600);letter-spacing:0;letter-spacing:var(--base-label-small-letter-spacing,0)}:host.nx-label--large{font-size:20px;font-size:var(--base-label-large-font-size,20px);line-height:28px;line-height:var(--base-label-large-line-height,28px);font-weight:300;font-weight:var(--base-label-large-font-weight,300);letter-spacing:.2px;letter-spacing:var(--base-label-large-letter-spacing,.2px)}:host[disabled=true] label{cursor:not-allowed}:host.nx-label--negative{color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    NxLabelComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [LABEL_DEFAULT_OPTIONS,] }] }
    ]; };
    NxLabelComponent.propDecorators = {
        disabled: [{ type: Input }],
        negative: [{ type: Input }],
        id: [{ type: Input }],
        size: [{ type: Input }]
    };
    return NxLabelComponent;
}());
export { NxLabelComponent };
if (false) {
    /** @type {?} */
    NxLabelComponent.prototype._stateChanges;
    /**
     * @type {?}
     * @private
     */
    NxLabelComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    NxLabelComponent.prototype._negative;
    /**
     * @type {?}
     * @private
     */
    NxLabelComponent.prototype._size;
    /**
     * @type {?}
     * @private
     */
    NxLabelComponent.prototype._id;
    /**
     * @type {?}
     * @private
     */
    NxLabelComponent.prototype._defaultOptions;
}
