// @allianz/ngx-ndbx/esm2015/table/expandable/expandable-table.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: expandable/expandable-table.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChildren, QueryList, Directive } from '@angular/core';
import { NxExpandableTableRowComponent } from './expandable-table-row.component';
import { flatMap, startWith, map, distinctUntilChanged } from 'rxjs/operators';
import { combineLatest, BehaviorSubject } from 'rxjs';
/**
 * The nxExpandableTable adds row expansion functionality to the nxTable.
 * It provides the logic to expand/collapse all expandable rows simultanously.
 */
export class NxExpandableTableDirective {
    constructor() {
        this.expanded = new BehaviorSubject(false);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.rows.changes.pipe(startWith(this.rows), flatMap((/**
         * @param {?} rows
         * @return {?}
         */
        (rows) => {
            return combineLatest(rows.map((/**
             * @param {?} row
             * @return {?}
             */
            row => row.expanded)));
        })), map((/**
         * @param {?} values
         * @return {?}
         */
        (values) => values.reduce((/**
         * @param {?} a
         * @param {?} x
         * @return {?}
         */
        (a, x) => a && x), true))), distinctUntilChanged())
            .subscribe((/**
         * @param {?} rowsOpen
         * @return {?}
         */
        (rowsOpen) => {
            this.expanded.next(rowsOpen);
        }));
    }
    /**
     * Toggles the expandable rows. If all rows are open it will close them, otherwise it will open them.
     * @return {?}
     */
    toggle() {
        if (this.expanded.value) {
            this.close();
        }
        else {
            this.expand();
        }
    }
    /**
     * Expands all expandable rows.
     * @return {?}
     */
    expand() {
        this.rows.forEach((/**
         * @param {?} row
         * @return {?}
         */
        row => row.expanded.next(true)));
    }
    /**
     * Closes all expandable rows.
     * @return {?}
     */
    close() {
        this.rows.forEach((/**
         * @param {?} row
         * @return {?}
         */
        row => row.expanded.next(false)));
    }
}
NxExpandableTableDirective.decorators = [
    { type: Directive, args: [{
                selector: 'table[nxExpandableTable]',
                exportAs: 'nxExpandableTable'
            },] }
];
NxExpandableTableDirective.propDecorators = {
    rows: [{ type: ContentChildren, args: [NxExpandableTableRowComponent, { descendants: true },] }]
};
if (false) {
    /** @type {?} */
    NxExpandableTableDirective.prototype.rows;
    /** @type {?} */
    NxExpandableTableDirective.prototype.expanded;
}
