// @allianz/ngx-ndbx/esm5/table/expandable/expandable-table-cell.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: expandable/expandable-table-cell.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { animate, state, style, transition, trigger } from '@angular/animations';
import { ChangeDetectionStrategy, Component, ChangeDetectorRef } from '@angular/core';
import { NxExpandableTableRowComponent } from './expandable-table-row.component';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
/**
 * This is an expandable table cell.
 * Only works in conjunction with the `[nxExpandableTableRow]` component.
 */
var NxExpandableTableCellComponent = /** @class */ (function () {
    function NxExpandableTableCellComponent(_row, _changeDetectorRef) {
        var _this = this;
        this._changeDetectorRef = _changeDetectorRef;
        this._destroyed = new Subject();
        this._open = false;
        _row.expanded.pipe(takeUntil(this._destroyed))
            .subscribe((/**
         * @param {?} open
         * @return {?}
         */
        function (open) {
            _this._open = open;
            _this._changeDetectorRef.markForCheck();
        }));
    }
    /**
     * @return {?}
     */
    NxExpandableTableCellComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._destroyed.next();
        this._destroyed.complete();
    };
    NxExpandableTableCellComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'td[nxExpandableTableCell]',
                    template: "<div [@openClose]=\"_open ? 'open' : 'closed'\" class=\"nx-expandable-table-cell__content\">\n  <div class=\"nx-expandable-table-cell__wrapper\">\n    <ng-content></ng-content>\n  </div>\n</div>\n",
                    host: {
                        'class': 'nx-expandable-table-cell'
                    },
                    animations: [
                        trigger('openClose', [
                            state('open', style({
                                height: '*',
                                visibility: 'visible'
                            })),
                            state('closed, void', style({
                                height: '0',
                                visibility: 'hidden' // visibility and height to toggle instead of display: none to prevent cell width jumping
                            })),
                            transition('closed <=> open, void => closed', [
                                animate('225ms cubic-bezier(0.4,0.0,0.2,1)')
                            ])
                        ])
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    NxExpandableTableCellComponent.ctorParameters = function () { return [
        { type: NxExpandableTableRowComponent },
        { type: ChangeDetectorRef }
    ]; };
    return NxExpandableTableCellComponent;
}());
export { NxExpandableTableCellComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxExpandableTableCellComponent.prototype._destroyed;
    /** @type {?} */
    NxExpandableTableCellComponent.prototype._open;
    /**
     * @type {?}
     * @private
     */
    NxExpandableTableCellComponent.prototype._changeDetectorRef;
}
