// @allianz/ngx-ndbx/esm5/table/expandable/expandable-table.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: expandable/expandable-table.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChildren, QueryList, Directive } from '@angular/core';
import { NxExpandableTableRowComponent } from './expandable-table-row.component';
import { flatMap, startWith, map, distinctUntilChanged } from 'rxjs/operators';
import { combineLatest, BehaviorSubject } from 'rxjs';
/**
 * The nxExpandableTable adds row expansion functionality to the nxTable.
 * It provides the logic to expand/collapse all expandable rows simultanously.
 */
var NxExpandableTableDirective = /** @class */ (function () {
    function NxExpandableTableDirective() {
        this.expanded = new BehaviorSubject(false);
    }
    /**
     * @return {?}
     */
    NxExpandableTableDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.rows.changes.pipe(startWith(this.rows), flatMap((/**
         * @param {?} rows
         * @return {?}
         */
        function (rows) {
            return combineLatest(rows.map((/**
             * @param {?} row
             * @return {?}
             */
            function (row) { return row.expanded; })));
        })), map((/**
         * @param {?} values
         * @return {?}
         */
        function (values) {
            return values.reduce((/**
             * @param {?} a
             * @param {?} x
             * @return {?}
             */
            function (a, x) { return a && x; }), true);
        })), distinctUntilChanged())
            .subscribe((/**
         * @param {?} rowsOpen
         * @return {?}
         */
        function (rowsOpen) {
            _this.expanded.next(rowsOpen);
        }));
    };
    /**
     * Toggles the expandable rows. If all rows are open it will close them, otherwise it will open them.
     */
    /**
     * Toggles the expandable rows. If all rows are open it will close them, otherwise it will open them.
     * @return {?}
     */
    NxExpandableTableDirective.prototype.toggle = /**
     * Toggles the expandable rows. If all rows are open it will close them, otherwise it will open them.
     * @return {?}
     */
    function () {
        if (this.expanded.value) {
            this.close();
        }
        else {
            this.expand();
        }
    };
    /**
     * Expands all expandable rows.
     */
    /**
     * Expands all expandable rows.
     * @return {?}
     */
    NxExpandableTableDirective.prototype.expand = /**
     * Expands all expandable rows.
     * @return {?}
     */
    function () {
        this.rows.forEach((/**
         * @param {?} row
         * @return {?}
         */
        function (row) { return row.expanded.next(true); }));
    };
    /**
     * Closes all expandable rows.
     */
    /**
     * Closes all expandable rows.
     * @return {?}
     */
    NxExpandableTableDirective.prototype.close = /**
     * Closes all expandable rows.
     * @return {?}
     */
    function () {
        this.rows.forEach((/**
         * @param {?} row
         * @return {?}
         */
        function (row) { return row.expanded.next(false); }));
    };
    NxExpandableTableDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'table[nxExpandableTable]',
                    exportAs: 'nxExpandableTable'
                },] }
    ];
    NxExpandableTableDirective.propDecorators = {
        rows: [{ type: ContentChildren, args: [NxExpandableTableRowComponent, { descendants: true },] }]
    };
    return NxExpandableTableDirective;
}());
export { NxExpandableTableDirective };
if (false) {
    /** @type {?} */
    NxExpandableTableDirective.prototype.rows;
    /** @type {?} */
    NxExpandableTableDirective.prototype.expanded;
}
