// @allianz/ngx-ndbx/esm5/table/expandable/toggle-button.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: expandable/toggle-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
/**
 * @record
 */
export function NxExpandable() { }
if (false) {
    /**
     * Indicates if this expandable instance is open or not.
     * @type {?}
     */
    NxExpandable.prototype.expanded;
    /**
     * Toggles the open state.
     * @type {?}
     */
    NxExpandable.prototype.toggle;
    /**
     * Expands this component.
     * @type {?}
     */
    NxExpandable.prototype.expand;
    /**
     * Closes this component.
     * @type {?}
     */
    NxExpandable.prototype.close;
}
/**
 * This toggle button can be clicked and toggles expandable elements.
 * e.g. it can toggle a expandable table row or anything else that implements the `NxExpandable` interface.
 */
var NxToggleButtonComponent = /** @class */ (function () {
    function NxToggleButtonComponent(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._expanded = false;
        this._ariaLabel = '';
    }
    Object.defineProperty(NxToggleButtonComponent.prototype, "target", {
        /**
         * This is the expandable target that will be toggled when the user clicks the button.
         */
        set: /**
         * This is the expandable target that will be toggled when the user clicks the button.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var _this = this;
            this._target = value;
            if (this._subscription) {
                this._subscription.unsubscribe();
            }
            if (this._target) {
                this._subscription = this._target.expanded.subscribe((/**
                 * @param {?} expanded
                 * @return {?}
                 */
                function (expanded) {
                    _this._expanded = expanded;
                    _this._changeDetectorRef.markForCheck();
                }));
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxToggleButtonComponent.prototype, "ariaLabel", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._ariaLabel = value;
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxToggleButtonComponent.prototype._onClick = /**
     * @return {?}
     */
    function () {
        if (this._target) {
            this._target.toggle();
        }
    };
    /**
     * @return {?}
     */
    NxToggleButtonComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._subscription) {
            this._subscription.unsubscribe();
        }
    };
    NxToggleButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-toggle-button',
                    template: "<button (click)=\"_onClick()\"\n    class=\"nx-toggle-button\"\n    [attr.aria-label]=\"_ariaLabel\"\n    [attr.aria-expanded]=\"_expanded\"\n    [class.is-expanded]=\"_expanded\">\n  <nx-icon name=\"chevron-down\"\n      aria-hidden=\"true\"\n      size=\"s\"\n      class=\"nx-toggle-button__icon\"></nx-icon>\n</button>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host{display:block}.nx-toggle-button{display:block;cursor:pointer;-webkit-appearance:none;-moz-appearance:none;appearance:none;border:none;outline:0;background:0 0;padding:0;line-height:24px;height:24px}.nx-toggle-button .nx-toggle-button__icon{transform:rotateZ(0);transition:transform .2s;color:#414141;color:var(--expandable-table-arrow-icon-color,#414141)}.nx-toggle-button.is-expanded .nx-toggle-button__icon{transform:rotateZ(-180deg)}.nx-toggle-button::-moz-focus-inner{border:0}:host-context([data-whatinput=keyboard]) .nx-toggle-button:focus{border-radius:4px;box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6)}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]) .nx-toggle-button:focus{box-shadow:0 0 0 2px background,0 0 0 6px windowText}}"]
                }] }
    ];
    /** @nocollapse */
    NxToggleButtonComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxToggleButtonComponent.propDecorators = {
        target: [{ type: Input }],
        ariaLabel: [{ type: Input }]
    };
    return NxToggleButtonComponent;
}());
export { NxToggleButtonComponent };
if (false) {
    /** @type {?} */
    NxToggleButtonComponent.prototype._expanded;
    /** @type {?} */
    NxToggleButtonComponent.prototype._target;
    /** @type {?} */
    NxToggleButtonComponent.prototype._ariaLabel;
    /**
     * @type {?}
     * @private
     */
    NxToggleButtonComponent.prototype._subscription;
    /**
     * @type {?}
     * @private
     */
    NxToggleButtonComponent.prototype._changeDetectorRef;
}
