// @allianz/ngx-ndbx/esm2015/image/figure.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: image/figure.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
/** @type {?} */
const DEFAULT_SIZE = 'auto';
export class NxFigureComponent {
    /**
     * Sets the type of the visual appearance of the image.
     * The default value  is 'auto'.
     * @param {?} value
     * @return {?}
     */
    set classNames(value) {
        if (this._classNames === value) {
            return;
        }
        this._classNames = value;
        /** @type {?} */
        const sizeRegex = /^(auto|1by1|1dot8by1|1dot2by1|1by1dot1|2dot6by1)$/;
        const [size = null] = this._classNames.match(sizeRegex) || [DEFAULT_SIZE];
        this.size = (/** @type {?} */ (size));
        this.rounded = !!this._classNames.match(/rounded/);
    }
    /**
     * @return {?}
     */
    get classNames() {
        return this._classNames;
    }
}
NxFigureComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'figure[nxFigure]',
                template: '<ng-content></ng-content>',
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[class.nx-image--auto]': 'size === "auto" && !rounded',
                    '[class.nx-image--1by1]': 'size === "1by1"',
                    '[class.nx-image--1dot8by1]': 'size === "1dot8by1"',
                    '[class.nx-image--1dot2by1]': 'size === "1dot2by1"',
                    '[class.nx-image--1by1dot1]': 'size === "1by1dot1"',
                    '[class.nx-image--2dot6by1]': 'size === "2dot6by1"',
                    '[class.nx-image--rounded]': 'rounded'
                },
                styles: [":host{display:block;position:relative;line-height:0}:host ::ng-deep img{display:block;font-family:\"object-fit: cover;\";-o-object-fit:cover;object-fit:cover;height:100%;width:100%}:host(.nx-image--auto) ::ng-deep img{height:auto}:host(:not(.nx-image--auto)) ::ng-deep img{position:absolute;top:0;left:0}:host(.nx-image--rounded){padding-top:100%}:host(.nx-image--rounded) ::ng-deep img{border-radius:50%;bottom:0;left:0;position:absolute;right:0;top:0}:host(.nx-image--1by1){padding-top:100%}:host(.nx-image--1dot8by1){padding-top:55.55556%}:host(.nx-image--1dot2by1){padding-top:83.33333%}:host(.nx-image--1by1dot1){padding-top:110%}:host(.nx-image--2dot6by1){padding-top:38.46154%}:host(.nx-image--1by1dot4){padding-top:140%}"]
            }] }
];
NxFigureComponent.propDecorators = {
    classNames: [{ type: Input, args: ['nxFigure',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxFigureComponent.prototype._classNames;
    /**
     * \@docs-private
     * @type {?}
     */
    NxFigureComponent.prototype.rounded;
    /**
     * \@docs-private
     * @type {?}
     */
    NxFigureComponent.prototype.size;
}
