// @allianz/ngx-ndbx/esm2015/tree/node.js
/**
 * @fileoverview added by tsickle
 * Generated from: node.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CdkTreeNode, CdkTreeNodeDef } from '@angular/cdk/tree';
import { Component, Directive, Input, } from '@angular/core';
/**
 * Wrapper for the CdkTree node with NDBX design styles.
 * @template T
 */
export class NxTreeNodeComponent extends CdkTreeNode {
}
NxTreeNodeComponent.decorators = [
    { type: Component, args: [{
                selector: 'nx-tree-node',
                exportAs: 'nxTreeNode',
                inputs: ['disabled', 'tabIndex'],
                host: {
                    '[attr.aria-expanded]': 'isExpanded',
                    '[attr.aria-level]': 'role === "treeitem" ? level : null',
                    '[attr.role]': 'role',
                    'class': 'nx-tree__node',
                    '[class.is-expanded]': 'isExpanded'
                },
                providers: [{ provide: CdkTreeNode, useExisting: NxTreeNodeComponent }],
                template: "<div class=\"nx-tree__node-box\">\n  <ng-content></ng-content>\n</div>\n"
            }] }
];
/**
 * Wrapper for the CdkTree node definition with NDBX design styles.
 * @template T
 */
export class NxTreeNodeDefDirective extends CdkTreeNodeDef {
}
NxTreeNodeDefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[nxTreeNodeDef]',
                inputs: [
                    'when: nxTreeNodeDefWhen'
                ],
                providers: [{ provide: CdkTreeNodeDef, useExisting: NxTreeNodeDefDirective }]
            },] }
];
NxTreeNodeDefDirective.propDecorators = {
    data: [{ type: Input, args: ['nxTreeNode',] }]
};
if (false) {
    /** @type {?} */
    NxTreeNodeDefDirective.prototype.data;
}
