// @allianz/ngx-ndbx/esm5/progress-stepper/multi/multi-step-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: multi/multi-step-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
/**
 * \@docs-private
 */
var NxMultiStepItemComponent = /** @class */ (function () {
    function NxMultiStepItemComponent(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._direction = 'horizontal';
    }
    Object.defineProperty(NxMultiStepItemComponent.prototype, "direction", {
        /** The direction of the step */
        get: /**
         * The direction of the step
         * @return {?}
         */
        function () {
            return this._direction;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._direction = value;
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "selected", {
        /** Sets the selected step. */
        get: /**
         * Sets the selected step.
         * @return {?}
         */
        function () {
            return this._selected;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._selected = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "active", {
        /** Sets the active step. */
        get: /**
         * Sets the active step.
         * @return {?}
         */
        function () {
            return this._active;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._active = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "last", {
        /** Sets the last step. */
        get: /**
         * Sets the last step.
         * @return {?}
         */
        function () {
            return this._last;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._last = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxMultiStepItemComponent.prototype, "completed", {
        /** Sets the step completed. */
        get: /**
         * Sets the step completed.
         * @return {?}
         */
        function () {
            return this._completed;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._completed = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    NxMultiStepItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-multi-step-item',
                    template: "<div class=\"label-box\">\n  <div class=\"dot\">\n    <div *ngIf=\"!completed && selected\" class=\"small-dot\"></div>\n    <nx-icon *ngIf=\"completed && !disabled\" name=\"check\"></nx-icon>\n  </div>\n  <div *ngIf=\"direction === 'vertical'\" class=\"label\" [attr.label]=\"label\">{{ label }}</div>\n</div>\n\n<div *ngIf=\"!last\" class=\"line\"></div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.nx-multi-step-item--vertical]': 'direction === "vertical"',
                        '[class.is-selected]': 'selected',
                        '[class.is-completed]': 'completed',
                        '[class.is-active]': 'active',
                        '[class.is-disabled]': 'disabled',
                        '[class.is-last]': 'last',
                        '[attr.aria-disabled]': 'disabled ? "true" : null'
                    },
                    styles: [":host{flex:1 1 100%;display:flex;align-items:center;justify-content:center;position:relative;padding-bottom:8px;cursor:pointer}:host.is-disabled{cursor:not-allowed}:host.is-disabled .dot{border-color:#c2c2c2;border-color:var(--multi-stepper-dot-disabled-color,#c2c2c2)}:host.is-completed:not(.is-disabled) .dot{background-color:#007ab3;background-color:var(--multi-stepper-dot-touched-color,#007ab3)}:host.nx-multi-step-item--vertical{flex:0 0 auto;flex-direction:column;align-items:flex-start;padding:0}:host.nx-multi-step-item--vertical .label-box{display:flex;align-items:center}:host.nx-multi-step-item--vertical .label{position:static;transform:none;margin-left:16px;white-space:nowrap;text-overflow:ellipsis;text-align:left;min-width:0;font-size:16px;font-size:var(--multi-stepper-step-font-size,16px);line-height:24px;line-height:var(--multi-stepper-step-line-height,24px);font-weight:400;font-weight:var(--multi-stepper-step-font-weight,400);letter-spacing:0;letter-spacing:var(--multi-stepper-step-letter-spacing,0)}:host.nx-multi-step-item--vertical .label:before{content:attr(label);height:0;overflow:hidden;display:block}:host.nx-multi-step-item--vertical.is-disabled .label{color:#c2c2c2;color:var(--multi-stepper-step-disabled-color,#c2c2c2)}:host.nx-multi-step-item--vertical.is-selected .label{font-weight:700}@media screen and (-ms-high-contrast:active){:host .dot{border-color:buttonText}:host.is-selected .dot{box-shadow:0 0 0 4px highlight}:host.is-disabled .label{color:GrayText}:host.is-disabled .dot{border-color:GrayText}:host.nx-multi-step-item--vertical .label{-ms-high-contrast-adjust:none;background-color:buttonFace;border-radius:4px;box-shadow:0 0 0 2px buttonFace;color:buttonText}:host.nx-multi-step-item--vertical.is-selected .label{background-color:highlight;border-radius:4px;box-shadow:0 0 0 2px highlight;color:highlightText}}:host.nx-multi-step-item--vertical .line{border-top:0;border-left-color:var(--multi-stepper-line-color,#c2c2c2);height:16px;flex:0 0 auto;margin-left:6px;border-left:4px dotted #c2c2c2}:host.nx-multi-step-item--vertical.is-active .line{border-left-style:solid;border-left-color:#007ab3;border-left-color:var(--multi-stepper-line-filled-color,#007ab3);margin-top:-5px;margin-bottom:-5px;height:26px}@media screen and (-ms-high-contrast:active){:host.nx-multi-step-item--vertical.is-active .line{border-color:windowText}}:host.nx-multi-step-item--vertical.is-last{flex:0 0 auto}:host:not(.nx-multi-step-item--vertical) .line{border-top-color:var(--multi-stepper-line-color,#c2c2c2);margin:0 4px;position:absolute;left:calc(16px + 4px);width:calc(100% - (16px + (4px * 2)));transform:translateX(50%);top:calc(16px / 2 - 4px / 2);border-top:4px dotted #c2c2c2}:host:not(.nx-multi-step-item--vertical).is-active .line{border-top-color:#007ab3;border-top-color:var(--multi-stepper-line-filled-color,#007ab3);border-top-style:solid;margin:0;left:calc(16px - 1px);width:calc(100% - (16px - 2px))}:host:not(:first-child)::before{content:''}:host.nx-multi-step-item--active::after{border-top-color:#007ab3;border-top-color:var(--multi-stepper-line-filled-color,#007ab3);border-top-style:solid;border-image-source:none;transition:border-style .1s ease-in;margin:0;left:calc(4px * 2);width:calc(100% - 4px)}.label-box{outline:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;border:none;padding:0;margin:0;background:0 0}:host-context([data-whatinput=keyboard]) .label-box:focus{outline:0}:host-context([data-whatinput=keyboard]) .label-box:focus::-moz-focus-inner{border:0}:host-context([data-whatinput=keyboard]) .label-box:focus .dot{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6)}.label{position:absolute;top:calc(8px + 16px);text-align:center;color:#414141;color:var(--multi-stepper-step-color,#414141);left:0;width:100%}.dot{width:16px;height:16px;border-color:var(--multi-stepper-dot-touched-color,#007ab3);border-radius:50%;color:#fff;flex:0 0 auto;position:relative;border:2px solid #007ab3}.dot nx-icon{font-size:16px;position:absolute;top:50%;left:50%;transform:translate(-50%,-50%)}.small-dot{position:absolute;top:50%;left:50%;transform:translate(-50%,-50%);width:8px;height:8px;border-radius:50%;background-color:#007ab3;background-color:var(--multi-stepper-dot-touched-color,#007ab3)}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]) .label-box:focus .dot{box-shadow:0 0 0 2px background,0 0 0 6px windowText}.small-dot{background-color:windowText}}"]
                }] }
    ];
    /** @nocollapse */
    NxMultiStepItemComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxMultiStepItemComponent.propDecorators = {
        direction: [{ type: Input }],
        label: [{ type: Input }],
        selected: [{ type: Input }],
        active: [{ type: Input }],
        disabled: [{ type: Input }],
        last: [{ type: Input }],
        completed: [{ type: Input }]
    };
    return NxMultiStepItemComponent;
}());
export { NxMultiStepItemComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._direction;
    /**
     * The label of the step.
     * @type {?}
     */
    NxMultiStepItemComponent.prototype.label;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._selected;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._active;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._last;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._completed;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepItemComponent.prototype._changeDetectorRef;
}
