// @allianz/ngx-ndbx/esm5/progress-stepper/multi/multi-step.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: multi/multi-step.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Input, QueryList, ContentChildren } from '@angular/core';
import { NxProgressStepperDirective } from './../progress-stepper.component';
import { NxMultiStepperGroupComponent } from './multi-step-group.component';
var NxMultiStepperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NxMultiStepperComponent, _super);
    function NxMultiStepperComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._direction = 'horizontal';
        return _this;
    }
    Object.defineProperty(NxMultiStepperComponent.prototype, "direction", {
        /** Sets the direction of the multi stepper. */
        get: /**
         * Sets the direction of the multi stepper.
         * @return {?}
         */
        function () {
            return this._direction;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._direction = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NxMultiStepperComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.groups.length) {
            this._steps.reset(this._stepsInGroups);
        }
        this.groups.changes.subscribe((/**
         * @return {?}
         */
        function () {
            _this._steps.reset(_this._stepsInGroups);
        }));
    };
    /**
     * @param {?} index
     * @return {?}
     */
    NxMultiStepperComponent.prototype._stepDisabled = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        /** @type {?} */
        var steps = this.steps.toArray();
        if (this.linear && index >= 0) {
            return steps.slice(0, index).some((/**
             * @param {?} step
             * @return {?}
             */
            function (step) {
                /** @type {?} */
                var control = step.stepControl;
                /** @type {?} */
                var isIncomplete = control
                    ? control.invalid || control.pending || !step.interacted
                    : !step.completed;
                return isIncomplete && !step.optional;
            }));
        }
        return false;
    };
    Object.defineProperty(NxMultiStepperComponent.prototype, "_stepsInGroups", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            if (this.groups.length) {
                return this.groups
                    .reduce((/**
                 * @param {?} steps
                 * @param {?} group
                 * @return {?}
                 */
                function (steps, group) { return steps.concat(group.steps.toArray()); }), []);
            }
            return [];
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} step
     * @return {?}
     */
    NxMultiStepperComponent.prototype._getIndex = /**
     * @param {?} step
     * @return {?}
     */
    function (step) {
        return this.steps.toArray().indexOf(step);
    };
    /**
     * @param {?} groupIndex
     * @return {?}
     */
    NxMultiStepperComponent.prototype._isGroupActive = /**
     * @param {?} groupIndex
     * @return {?}
     */
    function (groupIndex) {
        return this.selectedIndex >= this.getGroupItemIndex(groupIndex);
    };
    /**
     * @param {?} groupIndex
     * @return {?}
     */
    NxMultiStepperComponent.prototype._isGroupDisabled = /**
     * @param {?} groupIndex
     * @return {?}
     */
    function (groupIndex) {
        return this._stepDisabled(this.getGroupItemIndex(groupIndex));
    };
    /**
     * @private
     * @param {?} groupIndex
     * @return {?}
     */
    NxMultiStepperComponent.prototype.getGroupItemIndex = /**
     * @private
     * @param {?} groupIndex
     * @return {?}
     */
    function (groupIndex) {
        /** @type {?} */
        var count = 0;
        /** @type {?} */
        var groups = this.groups.toArray();
        for (var i = 0; i < groupIndex; i++) {
            count += groups[i].steps.length;
        }
        return count;
    };
    NxMultiStepperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-multi-stepper',
                    template: "<span *ngIf=\"title && direction === 'horizontal'\" class=\"nx-single-step__title\">{{title}}</span>\n<span class=\"nx-progress-stepper__title\" *ngIf=\"direction === 'horizontal'\">\n  <ng-content select=\"nx-label\"></ng-content>\n</span>\n\n<div class=\"nx-multi-stepper__items\">\n  <ng-container *ngIf=\"groups.length === 0 || direction === 'horizontal'; else groupsTemplate\">\n    <ng-container *ngFor=\"let step of _steps;let i = index; let isLast = last; let isFirst = first\">\n\n      <nx-multi-step-item\n        (click)=\"!_stepDisabled(i) && step.select()\"\n        [last]=\"isLast\"\n        [label]=\"step.stepLabel || step.label\"\n        [completed]=\"step.completed\"\n        [selected]=\"selectedIndex === i\"\n        [active]=\"selectedIndex > i\"\n        [disabled]=\"_stepDisabled(i)\"\n        [direction]=\"direction\">\n      </nx-multi-step-item>\n    </ng-container>\n  </ng-container>\n</div>\n\n\n<div class=\"labels-container\"\n     *ngIf=\"direction === 'horizontal'\">\n  <div class=\"label\" [attr.label]=\"step.label\" *ngFor=\"let step of _steps;let i = index; \"\n       [ngClass]=\"{'is-disabled':_stepDisabled(i), 'is-selected': (selectedIndex === i)}\"\n       (click)=\"!_stepDisabled(i) && step.select()\">\n    <span>{{ step.label }}</span>\n  </div>\n</div>\n\n<div class=\"nx-progress-stepper__mobile-label\">\n  {{currentLabel}}\n</div>\n\n<div *ngFor=\"let step of _steps; let i = index\"\n      class=\"nx-progress-stepper__content\" role=\"tabpanel\"\n      [id]=\"_getStepContentId(i)\"\n      [attr.aria-label]=\"currentLabel\"\n      [attr.aria-expanded]=\"selectedIndex === i\">\n  <ng-container [ngTemplateOutlet]=\"step.content\"></ng-container>\n</div>\n\n<ng-template #groupsTemplate>\n  <ng-container *ngFor=\"let group of groups; let i = index;\">\n    <div class=\"nx-multi-stepper__group\">\n      <div class=\"nx-multi-stepper__group-label\"\n          [class.is-active]=\"_isGroupActive(i)\"\n          [class.is-disabled]=\"_isGroupDisabled(i)\">\n        {{ group.label }}\n      </div>\n\n      <div class=\"nx-multi-stepper__group-items\">\n        <ng-container *ngFor=\"let step of group.steps; let isLast = last;\">\n          <nx-multi-step-item\n              (click)=\"!_stepDisabled(_getIndex(step)) && step.select()\"\n              [last]=\"isLast\"\n              [label]=\"step.stepLabel || step.label\"\n              [completed]=\"step.completed\"\n              [selected]=\"selectedIndex === _getIndex(step)\"\n              [active]=\"selectedIndex > _getIndex(step)\"\n              [disabled]=\"_stepDisabled(_getIndex(step))\"\n              [direction]=\"direction\">\n          </nx-multi-step-item>\n        </ng-container>\n      </div>\n    </div>\n  </ng-container>\n</ng-template>\n",
                    providers: [
                        {
                            provide: NxProgressStepperDirective,
                            useExisting: NxMultiStepperComponent
                        }
                    ],
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.nx-multi-stepper--vertical]': 'direction === "vertical"'
                    },
                    styles: [":host{display:block}.nx-progress-stepper__content[aria-expanded=false]{height:0;overflow:hidden}.nx-progress-stepper__mobile-label{display:none;color:#414141;color:var(--progress-indicator-mobile-color,#414141);font-size:16px;font-size:var(--progress-indicator-mobile-font-size,16px);line-height:24px;line-height:var(--progress-indicator-mobile-line-height,24px);font-weight:400;font-weight:var(--progress-indicator-mobile-font-weight,400);letter-spacing:0;letter-spacing:var(--progress-indicator-mobile-letter-spacing,0)}@media (max-width:703px){.nx-progress-stepper__mobile-label{display:block;text-align:center}}.nx-progress-stepper__title{display:block}.nx-progress-stepper__title ::ng-deep nx-label{margin-bottom:24px}", ":host{width:100%}@media (min-width:704px){:host.nx-multi-stepper--vertical{display:flex}}:host.nx-multi-stepper--vertical .nx-progress-stepper__content[aria-expanded=false]{width:0}:host.nx-multi-stepper--vertical .nx-multi-stepper__items{align-items:flex-start;flex-direction:column;margin:0 56px 0 0}:host.nx-multi-stepper--vertical .nx-multi-stepper__group-items{flex-direction:column}:host.nx-multi-stepper--vertical .nx-multi-stepper__group-label{transform:none;border-left:4px solid transparent;margin-bottom:16px;margin-left:6px;padding-left:22px}:host.nx-multi-stepper--vertical .nx-multi-stepper__group:nth-child(n+2) .nx-multi-stepper__group-label{padding-top:16px;margin-top:16px;border-left-style:dotted;border-left-color:#c2c2c2;border-left-color:var(--multi-stepper-line-color,#c2c2c2)}:host.nx-multi-stepper--vertical .nx-multi-stepper__group:nth-child(n+2) .nx-multi-stepper__group-label.is-active{padding-top:36px;padding-bottom:20px;margin-top:-4px;margin-bottom:-4px;border-left-style:solid;border-left-color:#007ab3;border-left-color:var(--multi-stepper-line-filled-color,#007ab3)}:host .labels-container{width:100%;flex:1 1 100%;display:flex;margin:0 auto;vertical-align:top;align-self:flex-start;align-items:flex-start}:host .labels-container .label{flex:1 1 100%;display:flex;align-items:center;justify-content:center;position:relative;font-size:16px;font-size:var(--multi-stepper-step-font-size,16px);line-height:24px;line-height:var(--multi-stepper-step-line-height,24px);font-weight:400;font-weight:var(--multi-stepper-step-font-weight,400);letter-spacing:0;letter-spacing:var(--multi-stepper-step-letter-spacing,0);color:#414141;color:var(--multi-stepper-step-color,#414141);cursor:pointer;text-align:center}:host .labels-container .label.is-disabled{color:#c2c2c2;color:var(--multi-stepper-step-disabled-color,#c2c2c2);cursor:not-allowed}@media screen and (-ms-high-contrast:active){:host .labels-container .label{color:buttonText}:host .labels-container .label.is-disabled{color:GrayText}:host .labels-container .label span{-ms-high-contrast-adjust:none;background-color:buttonFace;border-radius:4px;box-shadow:0 0 0 2px buttonFace;color:buttonText}:host .labels-container .label.is-selected span{background-color:highlight;border-radius:4px;box-shadow:0 0 0 2px highlight;color:highlightText}}.nx-multi-stepper__items{display:flex;margin:0 auto;align-items:center}@media (max-width:703px){:host .labels-container{display:none}.nx-multi-stepper__items{display:none}}.nx-multi-stepper__group-items{display:flex}.nx-multi-stepper__group{display:flex;flex-direction:column;flex:1 1 auto}.nx-multi-stepper__group-label{font-size:20px;font-size:var(--multi-stepper-group-font-size,20px);line-height:28px;line-height:var(--multi-stepper-group-line-height,28px);font-weight:400;font-weight:var(--multi-stepper-group-font-weight,400);letter-spacing:.2px;letter-spacing:var(--multi-stepper-group-letter-spacing,.2px);color:#414141;color:var(--multi-stepper-group-color,#414141)}.nx-multi-stepper__group-label.is-disabled{color:#c2c2c2;color:var(--multi-stepper-group-disabled-color,#c2c2c2)}"]
                }] }
    ];
    NxMultiStepperComponent.propDecorators = {
        groups: [{ type: ContentChildren, args: [NxMultiStepperGroupComponent, { descendants: true },] }],
        direction: [{ type: Input }]
    };
    return NxMultiStepperComponent;
}(NxProgressStepperDirective));
export { NxMultiStepperComponent };
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxMultiStepperComponent.prototype.groups;
    /**
     * @type {?}
     * @private
     */
    NxMultiStepperComponent.prototype._direction;
}
