// @allianz/ngx-ndbx/esm2015/menu/menu-button.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: menu-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef, Directive } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
/**
 * \@title MenuButton
 */
export class NxMenuButtonComponent {
    /**
     * @param {?} _changeDetectorRef
     */
    constructor(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._expandable = false;
        this._expanded = false;
        this._type = 'root';
    }
    /**
     * Whether this menu button is expandable or not. Will add a caret icon.
     * @param {?} value
     * @return {?}
     */
    set expandable(value) {
        this._expandable = coerceBooleanProperty(value);
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get expandable() {
        return this._expandable;
    }
    /**
     * Whether this menu button is expanded or not.
     * Only works in combination with the `expandable` option set to `true`.
     * @param {?} value
     * @return {?}
     */
    set expanded(value) {
        this._expanded = coerceBooleanProperty(value);
        this._changeDetectorRef.markForCheck();
    }
    /**
     * @return {?}
     */
    get expanded() {
        return this._expanded;
    }
    /**
     * The type of this menu button.
     * Can be `primary` or `secondary`, defaults to `primary`.
     * @param {?} value
     * @return {?}
     */
    set type(value) {
        if (value === 'root' || value === 'nested') {
            this._type = value;
            this._changeDetectorRef.markForCheck();
        }
    }
    /**
     * @return {?}
     */
    get type() {
        return this._type;
    }
}
NxMenuButtonComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: '[nxMenuButton]',
                template: "<ng-content select=\"[nxMenuButtonIcon]\"></ng-content>\n<span class=\"nx-menu-button__label\">\n  <ng-content></ng-content>\n</span>\n<nx-icon *ngIf=\"expandable\" class=\"nx-menu-button__expand-icon\" name=\"chevron-down\"></nx-icon>\n",
                host: {
                    '[class.is-expanded]': 'expandable && expanded',
                    '[class.is-expandable]': 'expandable',
                    '[class.nx-menu-button--nested]': 'type === "nested"',
                    '[class.nx-menu-button--root]': 'type === "root"',
                    'class': 'nx-menu-button'
                },
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{display:flex;align-items:center;background:#fff;background:var(--menu-button-background-color,#fff);color:#414141;color:var(--menu-button-text-color,#414141);-webkit-appearance:none;-moz-appearance:none;appearance:none;outline:0;box-shadow:none;border:none;margin:0;cursor:pointer;line-height:24px;font-weight:400;font-size:18px;width:100%;text-decoration:none;padding:24px}:host:hover{background:#ececec;background:var(--menu-button-hover-background-color,#ececec)}:host.is-expanded{font-weight:700;background:#007ab3;background:var(--menu-button-expanded-background-color,#007ab3);color:#fff;color:var(--menu-button-expanded-text-color,#fff);border-color:#007ab3;border-color:var(--menu-button-expanded-border-color,#007ab3)}:host.is-expanded .nx-menu-button__expand-icon{transform:rotateZ(180deg)}:host.nx-menu-button--nested{background:#f5f5f5;background:var(--menu-button-nested-background-color,#f5f5f5)}:host.nx-menu-button--nested.is-expanded{background:#c1ebfb;background:var(--menu-button-nested-expanded-background-color,#c1ebfb);border-color:#c1ebfb;border-color:var(--menu-button-nested-expanded-border-color,#c1ebfb);color:#414141;color:var(--menu-button-nested-expanded-color,#414141)}:host::-moz-focus-inner{border:0}:host-context([data-whatinput=keyboard]):focus{box-shadow:inset 0 0 0 4px #009de6,inset 0 0 0 6px #fff;box-shadow:var(--focus-inset-box-shadow,inset 0 0 0 4px #009de6,inset 0 0 0 6px #fff);border-radius:8px;border:none}.nx-menu-button__expand-icon{display:flex;justify-content:center;transition:.2s;transform:rotateZ(0);width:16px;height:24px;margin-right:8px;flex:0 0 auto;font-size:24px}.nx-menu-button__label{overflow:hidden;white-space:nowrap;text-overflow:ellipsis;flex:1 1 auto;text-align:left;line-height:1;padding:4px 0}::ng-deep .nx-menu-button__icon{margin-right:16px;height:24px;font-weight:400;flex:0 0 auto}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]):focus{box-shadow:inset 0 0 0 4px windowText,inset 0 0 0 6px background}:host,:host.is-expanded,:host.nx-menu-button--nested,:host.nx-menu-button--nested.is-expanded{-ms-high-contrast-adjust:none;color:buttonText;background-color:buttonFace}:host.nx-menu-button--nested.is-expanded:hover,:host:hover{background-color:highlight;color:highlightText;border-bottom-color:windowText}}"]
            }] }
];
/** @nocollapse */
NxMenuButtonComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
NxMenuButtonComponent.propDecorators = {
    expandable: [{ type: Input }],
    expanded: [{ type: Input }],
    type: [{ type: Input, args: ['nxType',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxMenuButtonComponent.prototype._expandable;
    /**
     * @type {?}
     * @private
     */
    NxMenuButtonComponent.prototype._expanded;
    /**
     * @type {?}
     * @private
     */
    NxMenuButtonComponent.prototype._type;
    /**
     * @type {?}
     * @private
     */
    NxMenuButtonComponent.prototype._changeDetectorRef;
}
/**
 * Icon that can be added to a [nxButton].
 */
export class NxMenuButtonIconDirective {
}
NxMenuButtonIconDirective.decorators = [
    { type: Directive, args: [{
                selector: '[nxMenuButtonIcon]',
                host: {
                    class: 'nx-menu-button__icon'
                }
            },] }
];
