// @allianz/ngx-ndbx/esm5/action/action.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: action.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef, ContentChild } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { NxActionIconDirective } from './action-icon.directive';
var NxActionComponent = /** @class */ (function () {
    function NxActionComponent(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this._selected = false;
        this._expandable = false;
        this._expanded = false;
    }
    Object.defineProperty(NxActionComponent.prototype, "selected", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selected;
        },
        /** Whether this action is selected or not.  */
        set: /**
         * Whether this action is selected or not.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._selected = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxActionComponent.prototype, "expandable", {
        get: /**
         * @return {?}
         */
        function () {
            return this._expandable;
        },
        /** Whether this action is expandable or not. Will add a caret icon. */
        set: /**
         * Whether this action is expandable or not. Will add a caret icon.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._expandable = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxActionComponent.prototype, "expanded", {
        get: /**
         * @return {?}
         */
        function () {
            return this._expanded;
        },
        /** Whether this action is expanded or not.
          Only works in combination with the `expandable` option set to `true`.
         */
        set: /**
         * Whether this action is expanded or not.
         * Only works in combination with the `expandable` option set to `true`.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._expanded = coerceBooleanProperty(value);
            this._changeDetectorRef.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    NxActionComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: '[nxAction]',
                    template: "<div class=\"nx-action__wrapper\">\n  <ng-container *ngIf=\"expandable\">\n    <nx-icon class=\"nx-action__expand-icon\" name=\"chevron-right\"></nx-icon>\n  </ng-container>\n  <span class=\"nx-action-icon__container\" *ngIf=\"_iconChild\">\n    <ng-content select=\"[nxActionIcon]\"></ng-content>\n  </span>\n  <span class=\"nx-action__label\">\n    <ng-content></ng-content>\n  </span>\n</div>\n",
                    host: {
                        '[class.is-expanded]': 'expanded',
                        '[class.is-selected]': 'selected',
                        '[class.is-expandable]': 'expandable',
                        'class': 'nx-action'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host(.nx-action){display:block;font-size:16px;font-size:var(--action-font-size,16px);line-height:24px;line-height:var(--action-line-height,24px);font-weight:400;font-weight:var(--action-font-weight,400);letter-spacing:0;letter-spacing:var(--action-letter-spacing,0);-webkit-appearance:none;-moz-appearance:none;appearance:none;outline:0;box-shadow:none;border:none;margin:0;cursor:pointer;padding:8px 16px 8px 0;width:100%;text-decoration:none;color:#414141;color:var(--action-color,#414141);background-color:#f5f5f5;background-color:var(--action-background-color,#f5f5f5)}:host(.nx-action):hover{color:#414141;color:var(--action-hover-color,#414141);background-color:#ececec;background-color:var(--action-hover-background-color,#ececec)}:host(.nx-action).is-expanded{font-weight:700}:host(.nx-action).is-expanded .nx-action__expand-icon{transform:rotateZ(90deg)}:host(.nx-action).is-expandable .nx-action__wrapper::before{flex:0 1 32px;max-width:8px;min-width:4px}:host(.nx-action).is-selected{color:#fff;color:var(--action-selected-color,#fff);background-color:#007ab3;background-color:var(--action-selected-background-color,#007ab3);font-weight:700}:host(.nx-action)[nxTreeNodePadding] .nx-action__wrapper::before{flex-shrink:0}:host(.nx-action)::-moz-focus-inner{border:0}:host-context([data-whatinput=keyboard]):focus{box-shadow:inset 0 0 0 4px #009de6,inset 0 0 0 6px #fff;box-shadow:var(--focus-inset-box-shadow,inset 0 0 0 4px #009de6,inset 0 0 0 6px #fff);border-radius:8px}.nx-action__expand-icon{display:flex;justify-content:center;transition:.2s;transform:rotateZ(0);margin-right:8px;flex:0 0 auto;font-size:16px;font-size:var(--action-expand-icon-size,16px)}.nx-action__wrapper{display:flex;align-items:center;position:relative;overflow:hidden}.nx-action__wrapper::before{content:'';display:block;flex:0 1 32px;min-width:16px;height:1em}.nx-action-icon__container{font-size:20px;font-size:var(--action-icon-size,20px);padding-right:16px;flex:0 0 auto;position:relative;justify-content:center;display:flex}::ng-deep .nx-action__icon{font-weight:400}::ng-deep .nx-action__icon.nx-icon--s{font-size:inherit}.nx-action__label{overflow:hidden;white-space:nowrap;text-overflow:ellipsis;flex:1 1 50%;text-align:left}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]):focus{box-shadow:inset 0 0 0 4px windowText,inset 0 0 0 6px background}:host(.nx-action).is-selected,:host(.nx-action):hover{-ms-high-contrast-adjust:none;background-color:highlight;color:highlightText;font-weight:400}}"]
                }] }
    ];
    /** @nocollapse */
    NxActionComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxActionComponent.propDecorators = {
        _iconChild: [{ type: ContentChild, args: [NxActionIconDirective, { static: false },] }],
        selected: [{ type: Input }],
        expandable: [{ type: Input }],
        expanded: [{ type: Input }]
    };
    return NxActionComponent;
}());
export { NxActionComponent };
if (false) {
    /** @type {?} */
    NxActionComponent.prototype._iconChild;
    /**
     * @type {?}
     * @private
     */
    NxActionComponent.prototype._selected;
    /**
     * @type {?}
     * @private
     */
    NxActionComponent.prototype._expandable;
    /**
     * @type {?}
     * @private
     */
    NxActionComponent.prototype._expanded;
    /**
     * @type {?}
     * @private
     */
    NxActionComponent.prototype._changeDetectorRef;
}
