// @allianz/ngx-ndbx/esm2015/table/sort-header/sort-header.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: sort-header/sort-header.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Optional, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { NxSortDirective } from './sort.directive';
import { ENTER, SPACE } from '@angular/cdk/keycodes';
import { NxSortHeaderIntl } from './sort-header-intl';
export class NxSortHeaderComponent {
    /**
     * @param {?} _sort
     * @param {?} _intl
     * @param {?} _changeDetectorRef
     */
    constructor(_sort, _intl, _changeDetectorRef) {
        this._sort = _sort;
        this._intl = _intl;
        this._changeDetectorRef = _changeDetectorRef;
        this._intlSubscription = this._intl.changes.subscribe((/**
         * @return {?}
         */
        () => this._changeDetectorRef.markForCheck()));
    }
    /**
     * Sets the key of this sort header.
     * @param {?} value
     * @return {?}
     */
    set key(value) {
        if (this._key !== value) {
            this._key = value;
        }
    }
    /**
     * @return {?}
     */
    get key() {
        return this._key;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._parentChangeSubscription = this._sort._stateChanges.subscribe((/**
         * @return {?}
         */
        () => {
            this._changeDetectorRef.markForCheck();
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._parentChangeSubscription) {
            this._parentChangeSubscription.unsubscribe();
        }
        this._intlSubscription.unsubscribe();
    }
    /**
     * @return {?}
     */
    _handleClick() {
        this._sort.sort(this._key);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    _onKeydown($event) {
        if ($event && ($event.keyCode === ENTER || $event.keyCode === SPACE)) {
            this._sort.sort(this._key);
            // prevent page from scrolling down
            if ($event.keyCode === SPACE) {
                $event.preventDefault();
            }
        }
    }
    /**
     * @return {?}
     */
    _isSortedAscending() {
        return (this._sort.active === this._key && this._sort.direction === 'asc');
    }
    /**
     * @return {?}
     */
    _isSortedDescending() {
        return (this._sort.active === this._key && this._sort.direction === 'desc');
    }
    /**
     * @return {?}
     */
    _getAriaLabel() {
        if (this._sort.active === this._key && this._sort.direction === 'asc') {
            return `${this._intl.sortDescendingAriaLabel}`;
        }
        return `${this._intl.sortAscendingAriaLabel}`;
    }
}
NxSortHeaderComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'th[nxSortHeaderCell]',
                exportAs: 'nxSortHeaderComponent',
                template: "<div class=\"nx-sort-header__focus-container\" role=\"button\" tabindex=\"0\">\n  <div class=\"nx-sort-header__icons-container\"\n      [attr.aria-label]=\"_getAriaLabel()\">\n    <nx-icon name=\"chevron-up-small\"\n        class=\"nx-sort-header__icon-up\"\n        [class.nx-sort-header__hidden-icon]=\"_isSortedDescending()\">\n    </nx-icon>\n    <nx-icon name=\"chevron-down-small\"\n        class=\"nx-sort-header__icon-down\"\n        [class.nx-sort-header__hidden-icon]=\"_isSortedAscending()\">\n    </nx-icon>\n  </div>\n\n  <ng-content></ng-content>\n</div>\n",
                host: {
                    '(click)': '_handleClick()',
                    '(keydown)': '_onKeydown($event)'
                },
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{cursor:pointer}.nx-sort-header__icons-container{display:inline-block;margin-right:8px;vertical-align:bottom;width:24px;position:relative;height:24px}.nx-sort-header__icons-container nx-icon{font-size:20px;position:absolute;font-weight:600;width:24px}.nx-sort-header__icons-container .nx-sort-header__icon-up{top:-3px}.nx-sort-header__icons-container .nx-sort-header__icon-down{top:5px}.nx-sort-header__icons-container .nx-sort-header__hidden-icon{visibility:hidden}.nx-sort-header__focus-container{display:flex}.nx-sort-header__focus-container:focus{outline:0}:host-context([data-whatinput=keyboard]) .nx-sort-header__focus-container:focus{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6);border-radius:4px}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]) .nx-sort-header__focus-container:focus{box-shadow:0 0 0 2px background,0 0 0 6px windowText}.nx-sort-header__focus-container{color:buttonText}}"]
            }] }
];
/** @nocollapse */
NxSortHeaderComponent.ctorParameters = () => [
    { type: NxSortDirective, decorators: [{ type: Optional }] },
    { type: NxSortHeaderIntl },
    { type: ChangeDetectorRef }
];
NxSortHeaderComponent.propDecorators = {
    key: [{ type: Input, args: ['nxSortHeaderCell',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxSortHeaderComponent.prototype._parentChangeSubscription;
    /**
     * @type {?}
     * @private
     */
    NxSortHeaderComponent.prototype._intlSubscription;
    /**
     * @type {?}
     * @private
     */
    NxSortHeaderComponent.prototype._key;
    /** @type {?} */
    NxSortHeaderComponent.prototype._sort;
    /** @type {?} */
    NxSortHeaderComponent.prototype._intl;
    /**
     * @type {?}
     * @private
     */
    NxSortHeaderComponent.prototype._changeDetectorRef;
}
