// @allianz/ngx-ndbx/esm5/radio-toggle/radio-toggle-button-base.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: radio-toggle/radio-toggle-button-base.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef, Input, ViewChild } from '@angular/core';
import { removeClasses } from '@allianz/ngx-ndbx/utils';
/** @type {?} */
var nextId = 0;
/**
 * \@docs-private
 */
var NxRadioToggleButtonBaseComponent = /** @class */ (function () {
    function NxRadioToggleButtonBaseComponent(renderer) {
        this.renderer = renderer;
        /**
         * \@docs-private
         */
        this.toggleButtonId = 'nx-radio-toggle-button-' + nextId++;
    }
    /** @docs-private */
    /**
     * \@docs-private
     * @param {?} event
     * @return {?}
     */
    NxRadioToggleButtonBaseComponent.prototype.triggerInputClick = /**
     * \@docs-private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
        this.toggleInput.nativeElement.click();
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxRadioToggleButtonBaseComponent.prototype.setFirstButton = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this.renderer.addClass(this.labelContainer.nativeElement, 'nx-radio-toggle__label-container--first');
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxRadioToggleButtonBaseComponent.prototype.setLastButton = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this.renderer.addClass(this.labelContainer.nativeElement, 'nx-radio-toggle__label-container--last');
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxRadioToggleButtonBaseComponent.prototype.resetClasses = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        removeClasses(this.renderer, this.labelContainer, 'nx-radio-toggle__label-container--first nx-radio-toggle__label-container--last');
    };
    NxRadioToggleButtonBaseComponent.propDecorators = {
        value: [{ type: Input, args: ['nxValue',] }],
        labelContainer: [{ type: ViewChild, args: ['toggleButtonLabelContainer', { static: true },] }],
        toggleInput: [{ type: ViewChild, args: ['toggleInput', { static: true },] }]
    };
    return NxRadioToggleButtonBaseComponent;
}());
export { NxRadioToggleButtonBaseComponent };
if (false) {
    /**
     * \@docs-private
     * @type {?}
     */
    NxRadioToggleButtonBaseComponent.prototype.toggleButtonId;
    /**
     * \@docs-private
     * @type {?}
     */
    NxRadioToggleButtonBaseComponent.prototype.value;
    /**
     * \@docs-private
     * @type {?}
     */
    NxRadioToggleButtonBaseComponent.prototype.labelContainer;
    /**
     * \@docs-private
     * @type {?}
     */
    NxRadioToggleButtonBaseComponent.prototype.toggleInput;
    /**
     * \@docs-private
     * @type {?}
     * @protected
     */
    NxRadioToggleButtonBaseComponent.prototype.renderer;
}
