// @allianz/ngx-ndbx/esm5/progress-stepper/progress-stepper.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: progress-stepper.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DOCUMENT } from '@angular/common';
import { CdkStep, CdkStepper } from '@angular/cdk/stepper';
import { ChangeDetectionStrategy, Component, ContentChildren, Directive, forwardRef, Inject, Input, QueryList, SkipSelf, ElementRef, ChangeDetectorRef } from '@angular/core';
import { takeUntil, takeWhile } from 'rxjs/operators';
import { ErrorStateMatcher } from '@allianz/ngx-ndbx/utils';
import { Subject } from 'rxjs';
import { Directionality } from '@angular/cdk/bidi';
// tslint:disable:use-input-property-decorator
// We need to reference steps in stepper and stepper in steps. To prevent circular depenedency errors
// Provide both components in a single file. Otherwise we would have to introduce interface/abstract classes
// shared between both implementations.
var NxStepComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NxStepComponent, _super);
    function NxStepComponent(stepper, _errorStateMatcher) {
        var _this = _super.call(this, stepper) || this;
        _this.stepper = stepper;
        _this._errorStateMatcher = _errorStateMatcher;
        _this._destroyed = new Subject();
        _this._interacted = false;
        return _this;
    }
    Object.defineProperty(NxStepComponent.prototype, "stepControl", {
        /** The top level abstract control of the step. */
        get: /**
         * The top level abstract control of the step.
         * @return {?}
         */
        function () {
            return this._stepControl;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var _this = this;
            this._stepControl = value;
            // If a step control changes its state, the stepper needs to update.
            if (this._stepControl) {
                this._stepControl.statusChanges
                    .pipe(takeUntil(this._destroyed), takeWhile((/**
                 * @return {?}
                 */
                function () { return _this._stepControl === value; })))
                    .subscribe((/**
                 * @return {?}
                 */
                function () {
                    _this.stepper._stateChanged();
                }));
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxStepComponent.prototype, "interacted", {
        /** Whether the user has seen the expanded step content or not. */
        get: /**
         * Whether the user has seen the expanded step content or not.
         * @return {?}
         */
        function () {
            return this._interacted;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._interacted = value;
            if (this.stepper) {
                this.stepper._stateChanged();
            }
        },
        enumerable: true,
        configurable: true
    });
    /** Custom error state matcher that checks for validity of the step form. */
    /**
     * Custom error state matcher that checks for validity of the step form.
     * @param {?} control
     * @param {?} form
     * @return {?}
     */
    NxStepComponent.prototype.isErrorState = /**
     * Custom error state matcher that checks for validity of the step form.
     * @param {?} control
     * @param {?} form
     * @return {?}
     */
    function (control, form) {
        /** @type {?} */
        var originalErrorState = this._errorStateMatcher.isErrorState(control, form);
        // Checks for the validity of a step form that is not submitted or touched,
        // e.g when the user directly clicks the "next" button or directly on the step
        /** @type {?} */
        var customErrorState = !!(control && control.invalid && this.interacted);
        return originalErrorState || customErrorState;
    };
    /**
     * @return {?}
     */
    NxStepComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._destroyed.next(true);
        this._destroyed.complete();
    };
    NxStepComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-step',
                    template: "<ng-template><ng-content></ng-content></ng-template>\n",
                    exportAs: 'nxStep',
                    providers: [{ provide: ErrorStateMatcher, useExisting: NxStepComponent }],
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    NxStepComponent.ctorParameters = function () { return [
        { type: NxProgressStepperDirective, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return NxProgressStepperDirective; })),] }] },
        { type: ErrorStateMatcher, decorators: [{ type: SkipSelf }] }
    ]; };
    return NxStepComponent;
}(CdkStep));
export { NxStepComponent };
if (false) {
    /** @type {?} */
    NxStepComponent.prototype._destroyed;
    /**
     * @type {?}
     * @private
     */
    NxStepComponent.prototype._stepControl;
    /**
     * @type {?}
     * @private
     */
    NxStepComponent.prototype._interacted;
    /**
     * @type {?}
     * @private
     */
    NxStepComponent.prototype.stepper;
    /**
     * @type {?}
     * @private
     */
    NxStepComponent.prototype._errorStateMatcher;
}
/**
 * \@docs-private
 */
var NxProgressStepperDirective = /** @class */ (function (_super) {
    tslib_1.__extends(NxProgressStepperDirective, _super);
    function NxProgressStepperDirective(_cdRef, _dir, _elementRef, _document) {
        var _this = _super.call(this, _dir, _cdRef, _elementRef, _document) || this;
        _this._cdRef = _cdRef;
        _this._stepHeader = new QueryList();
        return _this;
    }
    /**
     * @return {?}
     */
    NxProgressStepperDirective.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Mark the component for change detection whenever the content children query changes
        this._steps.changes
            .pipe(takeUntil(this._destroyed))
            .subscribe((/**
         * @return {?}
         */
        function () {
            _this._stateChanged();
        }));
        // If a step control changes its state, the stepper needs to update.
        this._steps.filter((/**
         * @param {?} s
         * @return {?}
         */
        function (s) { return !!s.stepControl; }))
            .forEach((/**
         * @param {?} step
         * @return {?}
         */
        function (step) {
            step.stepControl.statusChanges
                .pipe(takeUntil(step._destroyed))
                .subscribe((/**
             * @return {?}
             */
            function () {
                _this._stateChanged();
            }));
        }));
        // we need to defer change detection that the nxStepperNext and NxStepperPrevious buttons
        // can run change detection again, because they are initialized before the stepper content
        Promise.resolve().then((/**
         * @return {?}
         */
        function () {
            _this._cdRef.detectChanges();
        }));
    };
    Object.defineProperty(NxProgressStepperDirective.prototype, "hasPrevious", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return (this.selectedIndex - 1) >= 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxProgressStepperDirective.prototype, "hasNext", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return (this.selectedIndex + 1) < this.count;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxProgressStepperDirective.prototype, "count", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return this._steps ? this._steps.length : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxProgressStepperDirective.prototype, "currentStep", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            if (this.count === 0 || this.selectedIndex === -1) {
                return null;
            }
            /** @type {?} */
            var steps = this._steps.toArray();
            return steps[this.selectedIndex];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxProgressStepperDirective.prototype, "nextStep", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            if (this.selectedIndex + 1 >= this._steps.length) {
                return null;
            }
            return this._steps.toArray()[this.selectedIndex + 1];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxProgressStepperDirective.prototype, "steps", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            return this._steps;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NxProgressStepperDirective.prototype, "currentLabel", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () {
            /** @type {?} */
            var step = this.currentStep;
            /** @type {?} */
            var label = step.stepLabel || step.label;
            return this.currentStepLabel + " " + (this.selectedIndex + 1) + "/" + this.count + ": " + label;
        },
        enumerable: true,
        configurable: true
    });
    NxProgressStepperDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[nxProgressStepper]',
                    exportAs: 'nxProgressStepper',
                },] }
    ];
    /** @nocollapse */
    NxProgressStepperDirective.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: Directionality },
        { type: ElementRef },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
    ]; };
    NxProgressStepperDirective.propDecorators = {
        _steps: [{ type: ContentChildren, args: [NxStepComponent, { descendants: true },] }],
        currentStepLabel: [{ type: Input }],
        title: [{ type: Input }]
    };
    return NxProgressStepperDirective;
}(CdkStepper));
export { NxProgressStepperDirective };
if (false) {
    /** @type {?} */
    NxProgressStepperDirective.prototype._steps;
    /** @type {?} */
    NxProgressStepperDirective.prototype._stepHeader;
    /**
     * Sets the label on the left side showing the current step label. Used for mobile viewports.
     * @type {?}
     */
    NxProgressStepperDirective.prototype.currentStepLabel;
    /**
     * Sets the title to be rendered above the progress bar.
     * \@deletion-target 10.0.0
     * @deprecated Use `nx-label` instead
     * @type {?}
     */
    NxProgressStepperDirective.prototype.title;
    /**
     * @type {?}
     * @private
     */
    NxProgressStepperDirective.prototype._cdRef;
}
