// @allianz/ngx-ndbx/esm5/circle-toggle/mobile-toggle-button/mobile-toggle-button.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: circle-toggle/mobile-toggle-button/mobile-toggle-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
/**
 * \@docs-private
 */
var NxMobileToggleButtonComponent = /** @class */ (function () {
    function NxMobileToggleButtonComponent(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this.checked = false;
        this.negative = false;
        this.disabled = false;
    }
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxMobileToggleButtonComponent.prototype.setFirstButton = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this.isFirst = true;
        this._changeDetectorRef.markForCheck();
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxMobileToggleButtonComponent.prototype.setLastButton = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this.isLast = true;
        this._changeDetectorRef.markForCheck();
    };
    /** @docs-private */
    /**
     * \@docs-private
     * @return {?}
     */
    NxMobileToggleButtonComponent.prototype.resetClasses = /**
     * \@docs-private
     * @return {?}
     */
    function () {
        this.isFirst = false;
        this.isLast = false;
        this._changeDetectorRef.markForCheck();
    };
    NxMobileToggleButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nx-mobile-toggle-button',
                    template: "<nx-icon class=\"nx-toggle-button__mobile-check\" name=\"check\" aria-hidden=\"true\"></nx-icon>\n<!-- If there is no label defined, use the circleText instead -->\n<span class=\"nx-toggle-button__label-text\"><ng-template *ngIf=\"label; then labelContent else circleTextContent\"></ng-template></span>\n<span *ngIf=\"hint\" class=\"nx-toggle-button__label-hint\">({{hint}})</span>\n\n<ng-template #labelContent>{{label}}</ng-template>\n<ng-template #circleTextContent>{{circleText}}</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.is-disabled]': 'disabled',
                        '[class.is-negative]': 'negative',
                        '[class.is-first]': 'isFirst',
                        '[class.is-last]': 'isLast',
                        '[class.is-flipped]': 'checked'
                    },
                    styles: [":host{display:none;position:relative;transition:background-color .2s,border-color .2s,color .2s;flex-direction:row;border-width:1px 2px;border-color:#007ab3;border-color:var(--circle-toggle-border-color,#007ab3);border-style:solid;padding:11px 31px;width:100%;justify-content:center;align-items:center;cursor:pointer}:host:not(.is-flipped):not(.is-disabled):hover{background-color:#006192;background-color:var(--circle-toggle-hover-background-color,#006192);border-color:#006192;border-color:var(--circle-toggle-hover-border-color,#006192)}:host:not(.is-flipped):not(.is-disabled):hover .nx-toggle-button__label-hint,:host:not(.is-flipped):not(.is-disabled):hover .nx-toggle-button__label-text{color:#fff;color:var(--circle-toggle-hover-icon-color,#fff)}@media (max-width:703px){:host-context(.is-responsive){display:flex}}.nx-toggle-button__label-hint{color:#007ab3;color:var(--circle-toggle-icon-color,#007ab3);font-size:14px;font-size:var(--circle-toggle-hint-mobile-font-size,14px);line-height:20px;line-height:var(--circle-toggle-hint-mobile-line-height,20px);font-weight:300;font-weight:var(--circle-toggle-hint-mobile-font-weight,300);letter-spacing:.2px;letter-spacing:var(--circle-toggle-hint-mobile-letter-spacing,.2px)}.nx-toggle-button__label-text{color:#007ab3;color:var(--circle-toggle-icon-color,#007ab3);font-size:14px;font-size:var(--circle-toggle-label-mobile-font-size,14px);line-height:20px;line-height:var(--circle-toggle-label-mobile-line-height,20px);font-weight:var(--circle-toggle-label-mobile-font-weight,400);letter-spacing:.2px;letter-spacing:var(--circle-toggle-label-mobile-letter-spacing,.2px);padding-top:0;padding-right:4px;font-weight:600}.nx-toggle-button__mobile-check{font-size:24px;height:24px;display:inline-block;opacity:0;margin-right:8px;color:#fff;color:var(--circle-toggle-checkmark-color,#fff)}:host(.is-first){border-top-width:2px;border-radius:4px 4px 0 0}:host(.is-last){border-bottom-width:2px;border-radius:0 0 4px 4px}:host(.is-flipped){background-color:#007ab3;background-color:var(--circle-toggle-selected-background-color,#007ab3)}:host(.is-flipped) .nx-toggle-button__label-hint,:host(.is-flipped) .nx-toggle-button__label-text{color:#fff;color:var(--circle-toggle-selected-icon-color,#fff)}:host(.is-flipped) .nx-toggle-button__mobile-check{opacity:1}:host(.is-negative){border-color:#fff}:host(.is-negative) .nx-toggle-button__label-hint,:host(.is-negative) .nx-toggle-button__label-text,:host(.is-negative) .nx-toggle-button__mobile-check{color:#fff}:host(.is-negative):not(.is-flipped):not(.is-disabled):hover{background-color:#d9d9d9;border-color:#d9d9d9}:host(.is-negative):not(.is-flipped):not(.is-disabled):hover .nx-toggle-button__label-hint,:host(.is-negative):not(.is-flipped):not(.is-disabled):hover .nx-toggle-button__label-text{color:#006192}:host(.is-negative.is-flipped){background-color:#fff}:host(.is-negative.is-flipped) .nx-toggle-button__label-hint,:host(.is-negative.is-flipped) .nx-toggle-button__label-text,:host(.is-negative.is-flipped) .nx-toggle-button__mobile-check{color:#007ab3}:host(.is-disabled){opacity:.4;cursor:not-allowed}@media screen and (-ms-high-contrast:active){:host,:host(.is-negative){background-color:buttonFace;border-color:buttonText}:host .nx-toggle-button__label-hint,:host .nx-toggle-button__label-text,:host .nx-toggle-button__mobile-check,:host(.is-negative) .nx-toggle-button__label-hint,:host(.is-negative) .nx-toggle-button__label-text,:host(.is-negative) .nx-toggle-button__mobile-check{color:buttonText;-ms-high-contrast-adjust:none}:host(.is-flipped),:host(.is-negative.is-flipped){background-color:highlight}:host(.is-flipped) .nx-toggle-button__label-hint,:host(.is-flipped) .nx-toggle-button__label-text,:host(.is-flipped) .nx-toggle-button__mobile-check,:host(.is-negative.is-flipped) .nx-toggle-button__label-hint,:host(.is-negative.is-flipped) .nx-toggle-button__label-text,:host(.is-negative.is-flipped) .nx-toggle-button__mobile-check{color:highlightText}:host(.is-disabled){opacity:1;border-color:GrayText;color:GrayText}:host(.is-disabled).is-flipped .nx-toggle-button__mobile-check{color:GrayText}}"]
                }] }
    ];
    /** @nocollapse */
    NxMobileToggleButtonComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NxMobileToggleButtonComponent.propDecorators = {
        checked: [{ type: Input }],
        negative: [{ type: Input }],
        disabled: [{ type: Input }],
        hint: [{ type: Input }],
        label: [{ type: Input }],
        circleText: [{ type: Input }]
    };
    return NxMobileToggleButtonComponent;
}());
export { NxMobileToggleButtonComponent };
if (false) {
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.checked;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.negative;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.disabled;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.hint;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.label;
    /** @type {?} */
    NxMobileToggleButtonComponent.prototype.circleText;
    /**
     * \@docs-private
     * @type {?}
     */
    NxMobileToggleButtonComponent.prototype.isFirst;
    /**
     * \@docs-private
     * @type {?}
     */
    NxMobileToggleButtonComponent.prototype.isLast;
    /**
     * @type {?}
     * @private
     */
    NxMobileToggleButtonComponent.prototype._changeDetectorRef;
}
