// @allianz/ngx-ndbx/fesm5/allianz-ngx-ndbx-headline.js
import { Component, Input, NgModule } from '@angular/core';
import { __read } from 'tslib';

/**
 * @fileoverview added by tsickle
 * Generated from: headline.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var DEFAULT_TYPE = 'section';
var NxHeadlineComponent = /** @class */ (function () {
    function NxHeadlineComponent() {
        /**
         * \@docs-private
         */
        this.type = DEFAULT_TYPE;
        /**
         * \@docs-private
         */
        this.negative = false;
    }
    Object.defineProperty(NxHeadlineComponent.prototype, "classNames", {
        get: /**
         * @return {?}
         */
        function () {
            return this._classNames;
        },
        /** Changes the type of the headline which affects the visual appearance.
         * You can combine a HeadlineType and 'negative'. */
        set: /**
         * Changes the type of the headline which affects the visual appearance.
         * You can combine a HeadlineType and 'negative'.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this._classNames === value) {
                return;
            }
            this._classNames = value;
            /** @type {?} */
            var typeRegex = /page-bold-caps|page|section|subsection-large|subsection-medium|subsection-small|subsection-xsmall/;
            var _a = __read(this._classNames.match(typeRegex) || [DEFAULT_TYPE], 1), _b = _a[0], type = _b === void 0 ? null : _b;
            this.type = (/** @type {?} */ (type));
            this.negative = !!this._classNames.match(/negative/);
        },
        enumerable: true,
        configurable: true
    });
    NxHeadlineComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: '[nxHeadline]',
                    template: "<ng-content></ng-content>",
                    host: {
                        '[class.nx-heading--page]': 'type === "page"',
                        '[class.nx-heading--page-bold-caps]': 'type === "page-bold-caps"',
                        '[class.nx-heading--section]': 'type === "section"',
                        '[class.nx-heading--subsection-large]': 'type === "subsection-large"',
                        '[class.nx-heading--subsection-medium]': 'type === "subsection-medium"',
                        '[class.nx-heading--subsection-small]': 'type === "subsection-small"',
                        '[class.nx-heading--subsection-xsmall]': 'type === "subsection-xsmall"',
                        '[class.nx-heading--negative]': 'negative',
                    },
                    styles: [":host{margin:0}:host ::ng-deep .nx-link,:host ::ng-deep nx-link a,:host a{color:#006192;color:var(--link-color,#006192);font-size:inherit;font-weight:600;line-height:inherit;letter-spacing:inherit;text-decoration:none}:host(.nx-heading--page){font-size:54px;font-size:var(--heading-01-font-size,54px);line-height:60px;line-height:var(--heading-01-line-height,60px);font-weight:300;font-weight:var(--heading-01-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-01-letter-spacing,.3px);margin-bottom:40px}@media (max-width:703px){:host(.nx-heading--page){font-size:40px;font-size:var(--heading-01-mobile-font-size,40px);line-height:48px;line-height:var(--heading-01-mobile-line-height,48px);font-weight:300;font-weight:var(--heading-01-mobile-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-01-mobile-letter-spacing,.3px)}}:host(.nx-heading--section){font-size:40px;font-size:var(--heading-02-font-size,40px);line-height:48px;line-height:var(--heading-02-line-height,48px);font-weight:300;font-weight:var(--heading-02-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-02-letter-spacing,.3px);margin-bottom:48px}@media (max-width:703px){:host(.nx-heading--section){font-size:35px;font-size:var(--heading-02-mobile-font-size,35px);line-height:40px;line-height:var(--heading-02-mobile-line-height,40px);font-weight:300;font-weight:var(--heading-02-mobile-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-02-mobile-letter-spacing,.3px)}}:host(.nx-heading--subsection-large){font-size:30px;font-size:var(--heading-03-font-size,30px);line-height:36px;line-height:var(--heading-03-line-height,36px);font-weight:300;font-weight:var(--heading-03-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-03-letter-spacing,.3px);margin-bottom:32px}@media (max-width:703px){:host(.nx-heading--subsection-large){font-size:26px;font-size:var(--heading-03-mobile-font-size,26px);line-height:32px;line-height:var(--heading-03-mobile-line-height,32px);font-weight:300;font-weight:var(--heading-03-mobile-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-03-mobile-letter-spacing,.3px)}}:host(.nx-heading--subsection-medium){font-size:26px;font-size:var(--heading-04-font-size,26px);line-height:32px;line-height:var(--heading-04-line-height,32px);font-weight:300;font-weight:var(--heading-04-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-04-letter-spacing,.3px);margin-bottom:40px}@media (max-width:703px){:host(.nx-heading--subsection-medium){font-size:24px;font-size:var(--heading-04-mobile-font-size,24px);line-height:32px;line-height:var(--heading-04-mobile-line-height,32px);font-weight:600;font-weight:var(--heading-04-mobile-font-weight,600);letter-spacing:.2px;letter-spacing:var(--heading-04-mobile-letter-spacing,.2px)}}:host(.nx-heading--subsection-small){font-size:24px;font-size:var(--heading-05-font-size,24px);line-height:32px;line-height:var(--heading-05-line-height,32px);font-weight:600;font-weight:var(--heading-05-font-weight,600);letter-spacing:.2px;letter-spacing:var(--heading-05-letter-spacing,.2px);margin-bottom:32px}@media (max-width:703px){:host(.nx-heading--subsection-small){font-size:20px;font-size:var(--heading-05-mobile-font-size,20px);line-height:28px;line-height:var(--heading-05-mobile-line-height,28px);font-weight:300;font-weight:var(--heading-05-mobile-font-weight,300);letter-spacing:.2px;letter-spacing:var(--heading-05-mobile-letter-spacing,.2px)}}:host(.nx-heading--subsection-xsmall){font-size:20px;font-size:var(--heading-06-font-size,20px);line-height:28px;line-height:var(--heading-06-line-height,28px);font-weight:300;font-weight:var(--heading-06-font-weight,300);letter-spacing:.2px;letter-spacing:var(--heading-06-letter-spacing,.2px);margin-bottom:24px}@media (max-width:703px){:host(.nx-heading--subsection-xsmall){font-size:18px;font-size:var(--heading-06-mobile-font-size,18px);line-height:28px;line-height:var(--heading-06-mobile-line-height,28px);font-weight:300;font-weight:var(--heading-06-mobile-font-weight,300);letter-spacing:.2px;letter-spacing:var(--heading-06-mobile-letter-spacing,.2px)}}:host(.nx-heading--negative){color:#fff}:host(.nx-heading--negative) ::ng-deep a{color:#fff}:host(.nx-heading--page-bold-caps){font-size:54px;font-size:var(--heading-01-font-size,54px);line-height:60px;line-height:var(--heading-01-line-height,60px);font-weight:var(--heading-01-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-01-letter-spacing,.3px);font-weight:700;text-transform:uppercase;margin-bottom:8px}@media (max-width:703px){:host(.nx-heading--page-bold-caps){font-size:40px;font-size:var(--heading-01-mobile-font-size,40px);line-height:48px;line-height:var(--heading-01-mobile-line-height,48px);font-weight:var(--heading-01-mobile-font-weight,300);letter-spacing:.3px;letter-spacing:var(--heading-01-mobile-letter-spacing,.3px);font-weight:700;text-transform:uppercase}}"]
                }] }
    ];
    NxHeadlineComponent.propDecorators = {
        classNames: [{ type: Input, args: ['nxHeadline',] }]
    };
    return NxHeadlineComponent;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxHeadlineComponent.prototype._classNames;
    /**
     * \@docs-private
     * @type {?}
     */
    NxHeadlineComponent.prototype.type;
    /**
     * \@docs-private
     * @type {?}
     */
    NxHeadlineComponent.prototype.negative;
}

/**
 * @fileoverview added by tsickle
 * Generated from: headline.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var NxHeadlineModule = /** @class */ (function () {
    function NxHeadlineModule() {
    }
    NxHeadlineModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [NxHeadlineComponent],
                    exports: [NxHeadlineComponent]
                },] }
    ];
    return NxHeadlineModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: allianz-ngx-ndbx-headline.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NxHeadlineComponent, NxHeadlineModule };
