// @allianz/ngx-ndbx/bundles/allianz-ngx-ndbx-button.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@allianz/ngx-ndbx/button', ['exports', '@angular/core'], factory) :
    (global = global || self, factory((global.allianz = global.allianz || {}, global.allianz['ngx-ndbx'] = global.allianz['ngx-ndbx'] || {}, global.allianz['ngx-ndbx'].button = {}), global.ng.core));
}(this, (function (exports, core) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: button-base.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var DEFAULT_SIZE = 'medium';
    /** @type {?} */
    var DEFAULT_TYPE = 'primary';
    /**
     * \@docs-private
     */
    var NxButtonBase = /** @class */ (function () {
        function NxButtonBase(_changeDetectorRef, _elementRef) {
            this._changeDetectorRef = _changeDetectorRef;
            this._elementRef = _elementRef;
            /**
             * \@docs-private
             */
            this.type = DEFAULT_TYPE;
            /**
             * \@docs-private
             */
            this.size = DEFAULT_SIZE;
            this.negative = false;
            this.block = false;
        }
        Object.defineProperty(NxButtonBase.prototype, "isPrimary", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.type === 'primary'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isSecondary", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.type === 'secondary'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isTertiary", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.type === 'tertiary'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isCta", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.type === 'cta'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isEmphasis", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.type === 'emphasis'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isLarge", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.size === 'large'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isMedium", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.size === 'medium'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isSmallMedium", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.size === 'small-medium'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isSmall", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.size === 'small'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isBlock", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.block; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "isNegative", {
            /** @docs-private */
            get: /**
             * \@docs-private
             * @return {?}
             */
            function () { return this.negative; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "classNames", {
            get: /**
             * @return {?}
             */
            function () {
                return this._classNames;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (this._classNames === value) {
                    return;
                }
                this._classNames = value;
                var _a = __read(this._classNames.match(/primary|secondary|tertiary|cta|emphasis/) || [DEFAULT_TYPE], 1), _b = _a[0], type = _b === void 0 ? null : _b;
                this.type = (/** @type {?} */ (type));
                var _c = __read(this._classNames.match(/small-medium|small|medium|large/) || [DEFAULT_SIZE], 1), _d = _c[0], size = _d === void 0 ? null : _d;
                this.size = (/** @type {?} */ (size));
                this.negative = /negative/.test(this._classNames);
                this.block = /block/.test(this._classNames);
                this._changeDetectorRef.markForCheck();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NxButtonBase.prototype, "elementRef", {
            /**
             * @docs-private
             * getter used for the modal component as a quickfix
             * since button got changed from directive to component the reference used by the ngOpenModelOnClick directive
             * for nxButtons is a reference to component instance instead of an element reference. As a workaround we need a
             * way to reach the elementRef of the component until the modal gets refactored.
             */
            get: /**
             * \@docs-private
             * getter used for the modal component as a quickfix
             * since button got changed from directive to component the reference used by the ngOpenModelOnClick directive
             * for nxButtons is a reference to component instance instead of an element reference. As a workaround we need a
             * way to reach the elementRef of the component until the modal gets refactored.
             * @return {?}
             */
            function () {
                return this._elementRef;
            },
            enumerable: true,
            configurable: true
        });
        NxButtonBase.propDecorators = {
            isPrimary: [{ type: core.HostBinding, args: ['class.nx-button--primary',] }],
            isSecondary: [{ type: core.HostBinding, args: ['class.nx-button--secondary',] }],
            isTertiary: [{ type: core.HostBinding, args: ['class.nx-button--tertiary',] }],
            isCta: [{ type: core.HostBinding, args: ['class.nx-button--cta',] }],
            isEmphasis: [{ type: core.HostBinding, args: ['class.nx-button--emphasis',] }],
            isLarge: [{ type: core.HostBinding, args: ['class.nx-button--large',] }],
            isMedium: [{ type: core.HostBinding, args: ['class.nx-button--medium',] }],
            isSmallMedium: [{ type: core.HostBinding, args: ['class.nx-button--small-medium',] }],
            isSmall: [{ type: core.HostBinding, args: ['class.nx-button--small',] }],
            isBlock: [{ type: core.HostBinding, args: ['class.nx-button--block',] }],
            isNegative: [{ type: core.HostBinding, args: ['class.nx-button--negative',] }]
        };
        return NxButtonBase;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        NxButtonBase.prototype._classNames;
        /**
         * \@docs-private
         * @type {?}
         */
        NxButtonBase.prototype.type;
        /**
         * \@docs-private
         * @type {?}
         */
        NxButtonBase.prototype.size;
        /** @type {?} */
        NxButtonBase.prototype.negative;
        /** @type {?} */
        NxButtonBase.prototype.block;
        /**
         * @type {?}
         * @private
         */
        NxButtonBase.prototype._changeDetectorRef;
        /**
         * @type {?}
         * @private
         */
        NxButtonBase.prototype._elementRef;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: button.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NxButtonComponent = /** @class */ (function (_super) {
        __extends(NxButtonComponent, _super);
        function NxButtonComponent(changeDetectorRef, elementRef) {
            return _super.call(this, changeDetectorRef, elementRef) || this;
        }
        NxButtonComponent.decorators = [
            { type: core.Component, args: [{
                        template: "<span class=\"nx-button__content-wrapper\">\n  <ng-content></ng-content>\n</span>\n",
                        // tslint:disable-next-line:component-selector
                        selector: 'button[nxButton]',
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        inputs: ['classNames:nxButton'],
                        styles: [":host{font-size:16px;font-size:var(--button-medium-font-size,16px);line-height:24px;line-height:var(--button-medium-line-height,24px);font-weight:600;font-weight:var(--button-medium-font-weight,600);letter-spacing:1px;letter-spacing:var(--button-medium-letter-spacing,1px);min-height:48px;min-height:var(--button-medium-height,48px);text-transform:uppercase;text-transform:var(--button-text-transform,uppercase);border-width:var(--button-border-width,2px);border-radius:4px;border-radius:var(--button-border-radius,4px);cursor:pointer;margin:0 0 24px;padding:10px 30px;padding:calc(12px - var(--button-border-width,2px)) calc(32px - var(--button-border-width,2px));transition:background-color .2s,border-color .2s,color .2s,box-shadow .2s;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;width:auto;-webkit-hyphens:none;-ms-hyphens:none;hyphens:none;vertical-align:middle;border:2px solid transparent}:host:disabled{cursor:not-allowed}:host:focus{outline:0}:host::-moz-focus-inner{border:0}:host-context([data-whatinput=keyboard]).nx-button--negative:focus,:host-context([data-whatinput=keyboard]):focus{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6)}:host(.nx-button--primary){color:#fff;color:var(--button-primary-text-color,#fff);background-color:#007ab3;background-color:var(--button-primary-background-color,#007ab3);border-color:transparent;border-color:var(--button-primary-border-color,transparent)}:host(.nx-button--primary):hover:not(:disabled):not(:active){background-color:#006192;background-color:var(--button-primary-hover-background-color,#006192);color:#fff;color:var(--button-primary-hover-text-color,#fff);border-color:transparent;border-color:var(--button-primary-hover-border-color,transparent)}:host(.nx-button--primary):active{background-color:#003781;background-color:var(--button-primary-active-background-color,#003781);color:#fff;color:var(--button-primary-active-text-color,#fff);border-color:transparent;border-color:var(--button-primary-active-border-color,transparent)}:host(.nx-button--primary):disabled{background-color:rgba(0,122,179,.4);background-color:var(--button-primary-disabled-background-color,rgba(0,122,179,.4));color:rgba(255,255,255,.4);color:var(--button-primary-disabled-text-color,rgba(255,255,255,.4));border-color:transparent;border-color:var(--button-primary-disabled-border-color,transparent)}:host(.nx-button--primary.nx-button--negative){background-color:#fff;border-color:#fff;color:#006192}:host(.nx-button--primary.nx-button--negative):hover:not(:disabled):not(:active){background-color:#d9d9d9;border-color:#d9d9d9;color:#006192}:host(.nx-button--primary.nx-button--negative):active{background-color:#c2c2c2;border-color:#c2c2c2;color:#006192}:host(.nx-button--primary.nx-button--negative):disabled{background-color:rgba(255,255,255,.4);border-color:transparent;color:rgba(65,65,65,.3)}:host(.nx-button--secondary){color:#006192;color:var(--button-secondary-text-color,#006192);background-color:transparent;background-color:var(--button-secondary-background-color,transparent);border-color:#007ab3;border-color:var(--button-secondary-border-color,#007ab3)}:host(.nx-button--secondary):hover:not(:disabled):not(:active){background-color:#006192;background-color:var(--button-secondary-hover-background-color,#006192);color:#fff;color:var(--button-secondary-hover-text-color,#fff);border-color:transparent;border-color:var(--button-secondary-hover-border-color,transparent)}:host(.nx-button--secondary):active{background-color:#003781;background-color:var(--button-secondary-active-background-color,#003781);color:#fff;color:var(--button-secondary-active-text-color,#fff);border-color:transparent;border-color:var(--button-secondary-active-border-color,transparent)}:host(.nx-button--secondary):disabled{background-color:transparent;background-color:var(--button-secondary-disabled-background-color,transparent);color:rgba(0,97,146,.4);color:var(--button-secondary-disabled-text-color,rgba(0,97,146,.4));border-color:rgba(0,122,179,.4);border-color:var(--button-secondary-disabled-border-color,rgba(0,122,179,.4))}:host(.nx-button--secondary.nx-button--negative){color:#fff;border-color:#fff}:host(.nx-button--secondary.nx-button--negative):hover:not(:disabled):not(:active){background-color:#d9d9d9;border-color:#d9d9d9;color:#006192;color:var(--button-secondary-text-color,#006192)}:host(.nx-button--secondary.nx-button--negative):active{background-color:#c2c2c2;border-color:#c2c2c2;color:#006192;color:var(--button-secondary-text-color,#006192)}:host(.nx-button--secondary.nx-button--negative):disabled{background-color:transparent;border-color:rgba(255,255,255,.4);color:rgba(255,255,255,.4)}:host(.nx-button--tertiary){color:#006192;color:var(--button-tertiary-text-color,#006192);background-color:transparent;background-color:var(--button-tertiary-background-color,transparent);border-color:transparent;border-color:var(--button-tertiary-border-color,transparent)}:host(.nx-button--tertiary):hover:not(:disabled):not(:active){background-color:#ececec;background-color:var(--button-tertiary-hover-background-color,#ececec);color:#006192;color:var(--button-tertiary-hover-text-color,#006192);border-color:transparent;border-color:var(--button-tertiary-hover-border-color,transparent)}:host(.nx-button--tertiary):active{background-color:#d9d9d9;background-color:var(--button-tertiary-active-background-color,#d9d9d9);color:#006192;color:var(--button-tertiary-active-text-color,#006192);border-color:transparent;border-color:var(--button-tertiary-active-border-color,transparent)}:host(.nx-button--tertiary):disabled{background-color:transparent;background-color:var(--button-tertiary-disabled-background-color,transparent);color:rgba(0,97,146,.4);color:var(--button-tertiary-disabled-text-color,rgba(0,97,146,.4));border-color:transparent;border-color:var(--button-tertiary-disabled-border-color,transparent)}:host(.nx-button--tertiary.nx-button--negative){color:#fff}:host(.nx-button--tertiary.nx-button--negative):hover{background-color:#d9d9d9;color:#006192;color:var(--button-secondary-text-color,#006192)}:host(.nx-button--tertiary.nx-button--negative):active{background-color:#c2c2c2;color:#006192;color:var(--button-secondary-text-color,#006192)}:host(.nx-button--tertiary.nx-button--negative):disabled{background-color:transparent;color:rgba(255,255,255,.4)}:host(.nx-button--large){font-size:20px;font-size:var(--button-large-font-size,20px);line-height:24px;line-height:var(--button-large-line-height,24px);font-weight:600;font-weight:var(--button-large-font-weight,600);letter-spacing:1px;letter-spacing:var(--button-large-letter-spacing,1px);min-height:72px;min-height:var(--button-large-height,72px);padding:22px 46px;padding:calc(24px - var(--button-border-width,2px)) calc(48px - var(--button-border-width,2px))}@media (max-width:703px){:host{display:block;width:100%}:host(.nx-button--large){padding:22px 30px;padding:calc(24px - var(--button-border-width,2px)) calc(32px - var(--button-border-width,2px))}}:host(.nx-button--small-medium){display:inline-block;font-size:16px;font-size:var(--button-small-medium-font-size,16px);line-height:24px;line-height:var(--button-small-medium-line-height,24px);font-weight:600;font-weight:var(--button-small-medium-font-weight,600);letter-spacing:1px;letter-spacing:var(--button-small-medium-letter-spacing,1px);padding:6px 22px;padding:calc(8px - var(--button-border-width,2px)) calc(24px - var(--button-border-width,2px));text-transform:none;width:auto;min-height:40px;min-height:var(--button-small-medium-height,40px)}:host(.nx-button--small){font-size:14px;font-size:var(--button-small-font-size,14px);line-height:16px;line-height:var(--button-small-line-height,16px);font-weight:600;font-weight:var(--button-small-font-weight,600);letter-spacing:1px;letter-spacing:var(--button-small-letter-spacing,1px);min-height:32px;min-height:var(--button-small-height,32px);display:inline-block;padding:6px 14px;padding:calc(8px - var(--button-border-width,2px)) calc(16px - var(--button-border-width,2px));text-transform:none;width:auto}:host(.nx-button--cta){color:#fff;color:var(--button-cta-text-color,#fff);background-color:#3da556;background-color:var(--button-cta-background-color,#3da556);border-color:transparent;border-color:var(--button-cta-border-color,transparent)}:host(.nx-button--cta):hover:not(:disabled):not(:active){background-color:#348357;background-color:var(--button-cta-hover-background-color,#348357);color:#fff;color:var(--button-cta-hover-text-color,#fff);border-color:transparent;border-color:var(--button-cta-hover-border-color,transparent)}:host(.nx-button--cta):active{background-color:#276442;background-color:var(--button-cta-active-background-color,#276442);color:#fff;color:var(--button-cta-active-text-color,#fff);border-color:transparent;border-color:var(--button-cta-active-border-color,transparent)}:host(.nx-button--cta):disabled{background-color:rgba(61,165,86,.4);background-color:var(--button-cta-disabled-background-color,rgba(61,165,86,.4));color:rgba(255,255,255,.4);color:var(--button-cta-disabled-text-color,rgba(255,255,255,.4));border-color:transparent;border-color:var(--button-cta-disabled-border-color,transparent)}:host(.nx-button--emphasis){color:#fff;color:var(--button-emphasis-text-color,#fff);background-color:#f86200;background-color:var(--button-emphasis-background-color,#f86200);border-color:transparent;border-color:var(--button-emphasis-border-color,transparent)}:host(.nx-button--emphasis):hover:not(:disabled):not(:active){background-color:#e15200;background-color:var(--button-emphasis-hover-background-color,#e15200);color:#fff;color:var(--button-emphasis-hover-text-color,#fff);border-color:transparent;border-color:var(--button-emphasis-hover-border-color,transparent)}:host(.nx-button--emphasis):active{background-color:#c15500;background-color:var(--button-emphasis-active-background-color,#c15500);color:#fff;color:var(--button-emphasis-active-text-color,#fff);border-color:transparent;border-color:var(--button-emphasis-active-border-color,transparent)}:host(.nx-button--emphasis):disabled{background-color:rgba(248,98,0,.4);background-color:var(--button-emphasis-disabled-background-color,rgba(248,98,0,.4));color:rgba(255,255,255,.4);color:var(--button-emphasis-disabled-text-color,rgba(255,255,255,.4));border-color:transparent;border-color:var(--button-emphasis-disabled-border-color,transparent)}@media (min-width:704px){:host(.nx-button--block){display:block;width:100%}}:host .nx-button__content-wrapper{display:inline-flex;align-items:center;justify-content:center;vertical-align:bottom}:host([nxiconbutton]),:host([nxsidebartoggle]){padding:0;flex:0 0 auto}:host([nxiconbutton]).nx-button--small,:host([nxsidebartoggle]).nx-button--small{width:32px;width:var(--button-small-height,32px);font-size:16px;font-size:var(--icon-button-small-icon-size,16px)}:host([nxiconbutton]).nx-button--small-medium,:host([nxsidebartoggle]).nx-button--small-medium{width:40px;width:var(--button-small-medium-height,40px);font-size:24px;font-size:var(--icon-button-small-medium-icon-size,24px)}:host([nxiconbutton]).nx-button--medium,:host([nxsidebartoggle]).nx-button--medium{width:48px;width:var(--button-medium-height,48px);font-size:24px;font-size:var(--icon-button-medium-icon-size,24px)}:host([nxiconbutton]).nx-button--large,:host([nxsidebartoggle]).nx-button--large{width:72px;width:var(--button-large-height,72px);font-size:24px;font-size:var(--icon-button-large-icon-size,24px)}:host([nxbutton]).nx-button--small ::ng-deep nx-icon{font-size:16px;font-size:var(--button-small-icon-size,16px)}:host([nxbutton]).nx-button--small-medium ::ng-deep nx-icon{font-size:20px;font-size:var(--button-small-medium-icon-size,20px)}:host([nxbutton]).nx-button--medium ::ng-deep nx-icon{font-size:24px;font-size:var(--button-medium-icon-size,24px)}:host([nxbutton]).nx-button--large ::ng-deep nx-icon{font-size:24px;font-size:var(--button-large-icon-size,24px)}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]).nx-button--negative:focus,:host-context([data-whatinput=keyboard]):focus{box-shadow:0 0 0 2px background,0 0 0 6px windowText}:host{background-color:buttonFace}:host .nx-button__content-wrapper{-ms-high-contrast-adjust:none}:host:disabled{color:GrayText;border-color:GrayText}}"]
                    }] }
        ];
        /** @nocollapse */
        NxButtonComponent.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef },
            { type: core.ElementRef }
        ]; };
        return NxButtonComponent;
    }(NxButtonBase));

    /**
     * @fileoverview added by tsickle
     * Generated from: icon-button.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NxIconButtonComponent = /** @class */ (function (_super) {
        __extends(NxIconButtonComponent, _super);
        function NxIconButtonComponent(changeDetectorRef, elementRef) {
            return _super.call(this, changeDetectorRef, elementRef) || this;
        }
        NxIconButtonComponent.decorators = [
            { type: core.Component, args: [{
                        // tslint:disable-next-line:component-selector
                        selector: 'button[nxIconButton]',
                        template: "<span class=\"nx-button__content-wrapper\">\n  <ng-content></ng-content>\n</span>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        inputs: ['classNames:nxIconButton'],
                        styles: [":host{font-size:16px;font-size:var(--button-medium-font-size,16px);line-height:24px;line-height:var(--button-medium-line-height,24px);font-weight:600;font-weight:var(--button-medium-font-weight,600);letter-spacing:1px;letter-spacing:var(--button-medium-letter-spacing,1px);min-height:48px;min-height:var(--button-medium-height,48px);text-transform:uppercase;text-transform:var(--button-text-transform,uppercase);border-width:var(--button-border-width,2px);border-radius:4px;border-radius:var(--button-border-radius,4px);cursor:pointer;margin:0 0 24px;padding:10px 30px;padding:calc(12px - var(--button-border-width,2px)) calc(32px - var(--button-border-width,2px));transition:background-color .2s,border-color .2s,color .2s,box-shadow .2s;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;width:auto;-webkit-hyphens:none;-ms-hyphens:none;hyphens:none;vertical-align:middle;border:2px solid transparent}:host:disabled{cursor:not-allowed}:host:focus{outline:0}:host::-moz-focus-inner{border:0}:host-context([data-whatinput=keyboard]).nx-button--negative:focus,:host-context([data-whatinput=keyboard]):focus{box-shadow:0 0 0 2px #fff,0 0 0 6px #009de6;box-shadow:var(--focus-box-shadow,0 0 0 2px #fff,0 0 0 6px #009de6)}:host(.nx-button--primary){color:#fff;color:var(--button-primary-text-color,#fff);background-color:#007ab3;background-color:var(--button-primary-background-color,#007ab3);border-color:transparent;border-color:var(--button-primary-border-color,transparent)}:host(.nx-button--primary):hover:not(:disabled):not(:active){background-color:#006192;background-color:var(--button-primary-hover-background-color,#006192);color:#fff;color:var(--button-primary-hover-text-color,#fff);border-color:transparent;border-color:var(--button-primary-hover-border-color,transparent)}:host(.nx-button--primary):active{background-color:#003781;background-color:var(--button-primary-active-background-color,#003781);color:#fff;color:var(--button-primary-active-text-color,#fff);border-color:transparent;border-color:var(--button-primary-active-border-color,transparent)}:host(.nx-button--primary):disabled{background-color:rgba(0,122,179,.4);background-color:var(--button-primary-disabled-background-color,rgba(0,122,179,.4));color:rgba(255,255,255,.4);color:var(--button-primary-disabled-text-color,rgba(255,255,255,.4));border-color:transparent;border-color:var(--button-primary-disabled-border-color,transparent)}:host(.nx-button--primary.nx-button--negative){background-color:#fff;border-color:#fff;color:#006192}:host(.nx-button--primary.nx-button--negative):hover:not(:disabled):not(:active){background-color:#d9d9d9;border-color:#d9d9d9;color:#006192}:host(.nx-button--primary.nx-button--negative):active{background-color:#c2c2c2;border-color:#c2c2c2;color:#006192}:host(.nx-button--primary.nx-button--negative):disabled{background-color:rgba(255,255,255,.4);border-color:transparent;color:rgba(65,65,65,.3)}:host(.nx-button--secondary){color:#006192;color:var(--button-secondary-text-color,#006192);background-color:transparent;background-color:var(--button-secondary-background-color,transparent);border-color:#007ab3;border-color:var(--button-secondary-border-color,#007ab3)}:host(.nx-button--secondary):hover:not(:disabled):not(:active){background-color:#006192;background-color:var(--button-secondary-hover-background-color,#006192);color:#fff;color:var(--button-secondary-hover-text-color,#fff);border-color:transparent;border-color:var(--button-secondary-hover-border-color,transparent)}:host(.nx-button--secondary):active{background-color:#003781;background-color:var(--button-secondary-active-background-color,#003781);color:#fff;color:var(--button-secondary-active-text-color,#fff);border-color:transparent;border-color:var(--button-secondary-active-border-color,transparent)}:host(.nx-button--secondary):disabled{background-color:transparent;background-color:var(--button-secondary-disabled-background-color,transparent);color:rgba(0,97,146,.4);color:var(--button-secondary-disabled-text-color,rgba(0,97,146,.4));border-color:rgba(0,122,179,.4);border-color:var(--button-secondary-disabled-border-color,rgba(0,122,179,.4))}:host(.nx-button--secondary.nx-button--negative){color:#fff;border-color:#fff}:host(.nx-button--secondary.nx-button--negative):hover:not(:disabled):not(:active){background-color:#d9d9d9;border-color:#d9d9d9;color:#006192;color:var(--button-secondary-text-color,#006192)}:host(.nx-button--secondary.nx-button--negative):active{background-color:#c2c2c2;border-color:#c2c2c2;color:#006192;color:var(--button-secondary-text-color,#006192)}:host(.nx-button--secondary.nx-button--negative):disabled{background-color:transparent;border-color:rgba(255,255,255,.4);color:rgba(255,255,255,.4)}:host(.nx-button--tertiary){color:#006192;color:var(--button-tertiary-text-color,#006192);background-color:transparent;background-color:var(--button-tertiary-background-color,transparent);border-color:transparent;border-color:var(--button-tertiary-border-color,transparent)}:host(.nx-button--tertiary):hover:not(:disabled):not(:active){background-color:#ececec;background-color:var(--button-tertiary-hover-background-color,#ececec);color:#006192;color:var(--button-tertiary-hover-text-color,#006192);border-color:transparent;border-color:var(--button-tertiary-hover-border-color,transparent)}:host(.nx-button--tertiary):active{background-color:#d9d9d9;background-color:var(--button-tertiary-active-background-color,#d9d9d9);color:#006192;color:var(--button-tertiary-active-text-color,#006192);border-color:transparent;border-color:var(--button-tertiary-active-border-color,transparent)}:host(.nx-button--tertiary):disabled{background-color:transparent;background-color:var(--button-tertiary-disabled-background-color,transparent);color:rgba(0,97,146,.4);color:var(--button-tertiary-disabled-text-color,rgba(0,97,146,.4));border-color:transparent;border-color:var(--button-tertiary-disabled-border-color,transparent)}:host(.nx-button--tertiary.nx-button--negative){color:#fff}:host(.nx-button--tertiary.nx-button--negative):hover{background-color:#d9d9d9;color:#006192;color:var(--button-secondary-text-color,#006192)}:host(.nx-button--tertiary.nx-button--negative):active{background-color:#c2c2c2;color:#006192;color:var(--button-secondary-text-color,#006192)}:host(.nx-button--tertiary.nx-button--negative):disabled{background-color:transparent;color:rgba(255,255,255,.4)}:host(.nx-button--large){font-size:20px;font-size:var(--button-large-font-size,20px);line-height:24px;line-height:var(--button-large-line-height,24px);font-weight:600;font-weight:var(--button-large-font-weight,600);letter-spacing:1px;letter-spacing:var(--button-large-letter-spacing,1px);min-height:72px;min-height:var(--button-large-height,72px);padding:22px 46px;padding:calc(24px - var(--button-border-width,2px)) calc(48px - var(--button-border-width,2px))}@media (max-width:703px){:host{display:block;width:100%}:host(.nx-button--large){padding:22px 30px;padding:calc(24px - var(--button-border-width,2px)) calc(32px - var(--button-border-width,2px))}}:host(.nx-button--small-medium){display:inline-block;font-size:16px;font-size:var(--button-small-medium-font-size,16px);line-height:24px;line-height:var(--button-small-medium-line-height,24px);font-weight:600;font-weight:var(--button-small-medium-font-weight,600);letter-spacing:1px;letter-spacing:var(--button-small-medium-letter-spacing,1px);padding:6px 22px;padding:calc(8px - var(--button-border-width,2px)) calc(24px - var(--button-border-width,2px));text-transform:none;width:auto;min-height:40px;min-height:var(--button-small-medium-height,40px)}:host(.nx-button--small){font-size:14px;font-size:var(--button-small-font-size,14px);line-height:16px;line-height:var(--button-small-line-height,16px);font-weight:600;font-weight:var(--button-small-font-weight,600);letter-spacing:1px;letter-spacing:var(--button-small-letter-spacing,1px);min-height:32px;min-height:var(--button-small-height,32px);display:inline-block;padding:6px 14px;padding:calc(8px - var(--button-border-width,2px)) calc(16px - var(--button-border-width,2px));text-transform:none;width:auto}:host(.nx-button--cta){color:#fff;color:var(--button-cta-text-color,#fff);background-color:#3da556;background-color:var(--button-cta-background-color,#3da556);border-color:transparent;border-color:var(--button-cta-border-color,transparent)}:host(.nx-button--cta):hover:not(:disabled):not(:active){background-color:#348357;background-color:var(--button-cta-hover-background-color,#348357);color:#fff;color:var(--button-cta-hover-text-color,#fff);border-color:transparent;border-color:var(--button-cta-hover-border-color,transparent)}:host(.nx-button--cta):active{background-color:#276442;background-color:var(--button-cta-active-background-color,#276442);color:#fff;color:var(--button-cta-active-text-color,#fff);border-color:transparent;border-color:var(--button-cta-active-border-color,transparent)}:host(.nx-button--cta):disabled{background-color:rgba(61,165,86,.4);background-color:var(--button-cta-disabled-background-color,rgba(61,165,86,.4));color:rgba(255,255,255,.4);color:var(--button-cta-disabled-text-color,rgba(255,255,255,.4));border-color:transparent;border-color:var(--button-cta-disabled-border-color,transparent)}:host(.nx-button--emphasis){color:#fff;color:var(--button-emphasis-text-color,#fff);background-color:#f86200;background-color:var(--button-emphasis-background-color,#f86200);border-color:transparent;border-color:var(--button-emphasis-border-color,transparent)}:host(.nx-button--emphasis):hover:not(:disabled):not(:active){background-color:#e15200;background-color:var(--button-emphasis-hover-background-color,#e15200);color:#fff;color:var(--button-emphasis-hover-text-color,#fff);border-color:transparent;border-color:var(--button-emphasis-hover-border-color,transparent)}:host(.nx-button--emphasis):active{background-color:#c15500;background-color:var(--button-emphasis-active-background-color,#c15500);color:#fff;color:var(--button-emphasis-active-text-color,#fff);border-color:transparent;border-color:var(--button-emphasis-active-border-color,transparent)}:host(.nx-button--emphasis):disabled{background-color:rgba(248,98,0,.4);background-color:var(--button-emphasis-disabled-background-color,rgba(248,98,0,.4));color:rgba(255,255,255,.4);color:var(--button-emphasis-disabled-text-color,rgba(255,255,255,.4));border-color:transparent;border-color:var(--button-emphasis-disabled-border-color,transparent)}@media (min-width:704px){:host(.nx-button--block){display:block;width:100%}}:host .nx-button__content-wrapper{display:inline-flex;align-items:center;justify-content:center;vertical-align:bottom}:host([nxiconbutton]),:host([nxsidebartoggle]){padding:0;flex:0 0 auto}:host([nxiconbutton]).nx-button--small,:host([nxsidebartoggle]).nx-button--small{width:32px;width:var(--button-small-height,32px);font-size:16px;font-size:var(--icon-button-small-icon-size,16px)}:host([nxiconbutton]).nx-button--small-medium,:host([nxsidebartoggle]).nx-button--small-medium{width:40px;width:var(--button-small-medium-height,40px);font-size:24px;font-size:var(--icon-button-small-medium-icon-size,24px)}:host([nxiconbutton]).nx-button--medium,:host([nxsidebartoggle]).nx-button--medium{width:48px;width:var(--button-medium-height,48px);font-size:24px;font-size:var(--icon-button-medium-icon-size,24px)}:host([nxiconbutton]).nx-button--large,:host([nxsidebartoggle]).nx-button--large{width:72px;width:var(--button-large-height,72px);font-size:24px;font-size:var(--icon-button-large-icon-size,24px)}:host([nxbutton]).nx-button--small ::ng-deep nx-icon{font-size:16px;font-size:var(--button-small-icon-size,16px)}:host([nxbutton]).nx-button--small-medium ::ng-deep nx-icon{font-size:20px;font-size:var(--button-small-medium-icon-size,20px)}:host([nxbutton]).nx-button--medium ::ng-deep nx-icon{font-size:24px;font-size:var(--button-medium-icon-size,24px)}:host([nxbutton]).nx-button--large ::ng-deep nx-icon{font-size:24px;font-size:var(--button-large-icon-size,24px)}@media screen and (-ms-high-contrast:active){:host-context([data-whatinput=keyboard]).nx-button--negative:focus,:host-context([data-whatinput=keyboard]):focus{box-shadow:0 0 0 2px background,0 0 0 6px windowText}:host{background-color:buttonFace}:host .nx-button__content-wrapper{-ms-high-contrast-adjust:none}:host:disabled{color:GrayText;border-color:GrayText}}"]
                    }] }
        ];
        /** @nocollapse */
        NxIconButtonComponent.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef },
            { type: core.ElementRef }
        ]; };
        return NxIconButtonComponent;
    }(NxButtonBase));

    /**
     * @fileoverview added by tsickle
     * Generated from: button.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NxButtonModule = /** @class */ (function () {
        function NxButtonModule() {
        }
        NxButtonModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [NxButtonComponent, NxIconButtonComponent],
                        exports: [NxButtonComponent, NxIconButtonComponent]
                    },] }
        ];
        return NxButtonModule;
    }());

    exports.NxButtonBase = NxButtonBase;
    exports.NxButtonComponent = NxButtonComponent;
    exports.NxButtonModule = NxButtonModule;
    exports.NxIconButtonComponent = NxIconButtonComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
