// @allianz/ngx-ndbx/esm2015/datefield/date-validators.js
/**
 * @fileoverview added by tsickle
 * Generated from: date-validators.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Provides a set of date validators that can be used by form controls.
 * \@dynamic
 */
export class NxDateValidators {
    /**
     * The form control validator for whether the input parses.
     * @template D
     * @param {?} dateAdapter
     * @param {?} dateFormats
     * @param {?} input
     * @param {?} strict
     * @param {?=} customParseFormat
     * @return {?}
     */
    static parse(dateAdapter, dateFormats, input, strict, customParseFormat = null) {
        return (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const parsedValue = dateAdapter.parse(input.value, customParseFormat || dateFormats.parse.dateInput, strict);
            /** @type {?} */
            const valid = !parsedValue || dateAdapter.isValid(parsedValue);
            return valid ? null : { 'nxDatefieldParse': { 'text': input } };
        });
    }
    /**
     * The form control validator for the min date.
     * @template D
     * @param {?} dateAdapter
     * @param {?} min
     * @return {?}
     */
    static min(dateAdapter, min) {
        return (/**
         * @param {?} control
         * @return {?}
         */
        (control) => {
            /** @type {?} */
            const controlValue = getValidDateOrNull(dateAdapter, dateAdapter.deserialize(control.value));
            return (!min || !controlValue ||
                dateAdapter.compareDate(min, controlValue) <= 0) ?
                null : { 'nxDatefieldMin': { 'min': min, 'actual': controlValue } };
        });
    }
    /**
     * The form control validator for the max date.
     * @template D
     * @param {?} dateAdapter
     * @param {?} max
     * @return {?}
     */
    static max(dateAdapter, max) {
        return (/**
         * @param {?} control
         * @return {?}
         */
        (control) => {
            /** @type {?} */
            const controlValue = getValidDateOrNull(dateAdapter, dateAdapter.deserialize(control.value));
            return (!max || !controlValue ||
                dateAdapter.compareDate(max, controlValue) >= 0) ?
                null : { 'nxDatefieldMax': { 'max': max, 'actual': controlValue } };
        });
    }
    /**
     * The form control validator for the date filter.
     * @template D
     * @param {?} dateAdapter
     * @param {?} dateFilter
     * @return {?}
     */
    static filter(dateAdapter, dateFilter) {
        return (/**
         * @param {?} control
         * @return {?}
         */
        (control) => {
            /** @type {?} */
            const controlValue = getValidDateOrNull(dateAdapter, dateAdapter.deserialize(control.value));
            return !dateFilter || !controlValue || dateFilter(controlValue) ?
                null : { 'nxDatefieldFilter': true };
        });
    }
}
/**
 * @template D
 * @param {?} dateAdapter
 * @param {?} obj
 * @return {?}
 */
function getValidDateOrNull(dateAdapter, obj) {
    return (dateAdapter.isDateInstance(obj) && dateAdapter.isValid(obj)) ? obj : null;
}
