// @allianz/ngx-ndbx/esm5/grid/layout.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: layout.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
var NxLayoutComponent = /** @class */ (function () {
    function NxLayoutComponent() {
        this._classNames = '';
        /**
         * \@docs-private
         */
        this.grid = true;
    }
    Object.defineProperty(NxLayoutComponent.prototype, "classNames", {
        get: /**
         * @return {?}
         */
        function () {
            return this._classNames;
        },
        /**
         * Type of layout.
         *
         * Values: grid | grid nogutters | grid maxwidth. Default value: grid.
         */
        set: /**
         * Type of layout.
         *
         * Values: grid | grid nogutters | grid maxwidth. Default value: grid.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this._classNames === value) {
                return;
            }
            this._classNames = value;
            this.grid = !!this._classNames.match(/grid/);
            this.noGutters = !!this._classNames.match(/nogutters/);
            this.maxWidth = !!this._classNames.match(/maxwidth/);
        },
        enumerable: true,
        configurable: true
    });
    NxLayoutComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: '[nxLayout]',
                    template: '<ng-content></ng-content>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.nx-grid]': 'grid',
                        '[class.nx-grid--no-gutters]': 'noGutters',
                        '[class.nx-grid--max-width]': 'maxWidth'
                    },
                    styles: [":host(.nx-grid){position:relative;margin-left:auto;margin-right:auto;padding-right:16px;padding-left:16px}@media (min-width:704px){:host(.nx-grid){padding-right:32px;padding-left:32px}}:host(.nx-grid--no-gutters) ::ng-deep{padding-left:0;padding-right:0}:host(.nx-grid--no-gutters) ::ng-deep .nx-grid__row{margin-left:0;margin-right:0}:host(.nx-grid--no-gutters) ::ng-deep .nx-grid__row>[class*=nx-grid__column]{padding-left:0;padding-right:0}:host(.nx-grid--max-width){max-width:100%;width:1184px;width:var(--grid-max-width,1184px)}"]
                }] }
    ];
    NxLayoutComponent.propDecorators = {
        classNames: [{ type: Input, args: ['nxLayout',] }]
    };
    return NxLayoutComponent;
}());
export { NxLayoutComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NxLayoutComponent.prototype._classNames;
    /**
     * \@docs-private
     * @type {?}
     */
    NxLayoutComponent.prototype.grid;
    /**
     * \@docs-private
     * @type {?}
     */
    NxLayoutComponent.prototype.noGutters;
    /**
     * \@docs-private
     * @type {?}
     */
    NxLayoutComponent.prototype.maxWidth;
}
