// @allianz/ngx-ndbx/esm5/datefield/adapter/date-adapter.js
/**
 * @fileoverview added by tsickle
 * Generated from: adapter/date-adapter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
/**
 * \@docs-private
 * @abstract
 * @template D
 */
var /**
 * \@docs-private
 * @abstract
 * @template D
 */
NxDateAdapter = /** @class */ (function () {
    function NxDateAdapter() {
        this.locale = 'de-DE';
        this._localeChanges = new Subject();
    }
    Object.defineProperty(NxDateAdapter.prototype, "localeChanges", {
        get: /**
         * @return {?}
         */
        function () { return this._localeChanges; },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @return {?}
     */
    NxDateAdapter.prototype.deserialize = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value == null || this.isDateInstance(value) && this.isValid(value)) {
            return value;
        }
        return this.invalid();
    };
    /**
     * @param {?} first
     * @param {?} second
     * @return {?}
     */
    NxDateAdapter.prototype.sameDate = /**
     * @param {?} first
     * @param {?} second
     * @return {?}
     */
    function (first, second) {
        if (first && second) {
            /** @type {?} */
            var firstValid = this.isValid(first);
            /** @type {?} */
            var secondValid = this.isValid(second);
            if (firstValid && secondValid) {
                return !this.compareDate(first, second);
            }
            return firstValid === secondValid;
        }
        return first === second;
    };
    /**
     * @param {?} first
     * @param {?} second
     * @return {?}
     */
    NxDateAdapter.prototype.compareDate = /**
     * @param {?} first
     * @param {?} second
     * @return {?}
     */
    function (first, second) {
        return this.getYear(first) - this.getYear(second) ||
            this.getMonth(first) - this.getMonth(second) ||
            this.getDate(first) - this.getDate(second);
    };
    /**
     * @param {?} locale
     * @return {?}
     */
    NxDateAdapter.prototype.setLocale = /**
     * @param {?} locale
     * @return {?}
     */
    function (locale) {
        this.locale = locale;
        this._localeChanges.next(this.locale);
    };
    /**
     * @return {?}
     */
    NxDateAdapter.prototype.getLocale = /**
     * @return {?}
     */
    function () {
        return this.locale;
    };
    /**
     * @param {?} date
     * @param {?=} min
     * @param {?=} max
     * @return {?}
     */
    NxDateAdapter.prototype.clampDate = /**
     * @param {?} date
     * @param {?=} min
     * @param {?=} max
     * @return {?}
     */
    function (date, min, max) {
        if (min && this.compareDate(date, min) < 0) {
            return min;
        }
        if (max && this.compareDate(date, max) > 0) {
            return max;
        }
        return date;
    };
    return NxDateAdapter;
}());
/**
 * \@docs-private
 * @abstract
 * @template D
 */
export { NxDateAdapter };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    NxDateAdapter.prototype.locale;
    /**
     * @type {?}
     * @protected
     */
    NxDateAdapter.prototype._localeChanges;
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.toIso8601 = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} displayFormat
     * @return {?}
     */
    NxDateAdapter.prototype.format = function (date, displayFormat) { };
    /**
     * @abstract
     * @param {?} value
     * @param {?} format
     * @param {?} strict
     * @return {?}
     */
    NxDateAdapter.prototype.parse = function (value, format, strict) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.isValid = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.clone = function (date) { };
    /**
     * @abstract
     * @param {?} obj
     * @return {?}
     */
    NxDateAdapter.prototype.isDateInstance = function (obj) { };
    /**
     * @abstract
     * @return {?}
     */
    NxDateAdapter.prototype.invalid = function () { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.getYear = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.getMonth = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.getDate = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.getYearName = function (date) { };
    /**
     * @abstract
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.createDate = function (year, month, date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.getNumDaysInMonth = function (date) { };
    /**
     * @abstract
     * @return {?}
     */
    NxDateAdapter.prototype.getDateNames = function () { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    NxDateAdapter.prototype.getDayOfWeek = function (date) { };
    /**
     * @abstract
     * @return {?}
     */
    NxDateAdapter.prototype.getFirstDayOfWeek = function () { };
    /**
     * @abstract
     * @param {?} style
     * @return {?}
     */
    NxDateAdapter.prototype.getMonthNames = function (style) { };
    /**
     * @abstract
     * @return {?}
     */
    NxDateAdapter.prototype.today = function () { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} months
     * @return {?}
     */
    NxDateAdapter.prototype.addCalendarMonths = function (date, months) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} years
     * @return {?}
     */
    NxDateAdapter.prototype.addCalendarYears = function (date, years) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    NxDateAdapter.prototype.addCalendarDays = function (date, days) { };
    /**
     * @abstract
     * @param {?} style
     * @return {?}
     */
    NxDateAdapter.prototype.getDayOfWeekNames = function (style) { };
}
